<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Mini-course CoDA - Exploratory analysis of compositional data (part II)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Mini-course CoDA</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html">
 <span class="menu-text">About</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">Exploratory analysis of compositional data (part II)</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation docked overflow-auto">
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  4_eda.qmd
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./9_practical.html" class="sidebar-item-text sidebar-link">Practical applications of learned concepts in R</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#exploratory-analysis-of-a-microbial-data-set" id="toc-exploratory-analysis-of-a-microbial-data-set" class="nav-link active" data-scroll-target="#exploratory-analysis-of-a-microbial-data-set">Exploratory analysis of a microbial data set</a>
  <ul class="collapse">
  <li><a href="#load-packages-and-data-data" id="toc-load-packages-and-data-data" class="nav-link" data-scroll-target="#load-packages-and-data-data">Load packages and data data</a></li>
  <li><a href="#exploration-of-the-phyloseq-object" id="toc-exploration-of-the-phyloseq-object" class="nav-link" data-scroll-target="#exploration-of-the-phyloseq-object">Exploration of the phyloseq object</a></li>
  <li><a href="#add-sample-variables" id="toc-add-sample-variables" class="nav-link" data-scroll-target="#add-sample-variables">Add sample variables</a></li>
  <li><a href="#sequencing-depth-library-size" id="toc-sequencing-depth-library-size" class="nav-link" data-scroll-target="#sequencing-depth-library-size">Sequencing depth / library size</a></li>
  <li><a href="#sparsity" id="toc-sparsity" class="nav-link" data-scroll-target="#sparsity">Sparsity</a></li>
  <li><a href="#visualize-microbial-composition" id="toc-visualize-microbial-composition" class="nav-link" data-scroll-target="#visualize-microbial-composition">Visualize microbial composition</a></li>
  <li><a href="#zero-replacement" id="toc-zero-replacement" class="nav-link" data-scroll-target="#zero-replacement">Zero replacement</a></li>
  <li><a href="#mean---variance-relationship" id="toc-mean---variance-relationship" class="nav-link" data-scroll-target="#mean---variance-relationship">Mean - variance relationship</a></li>
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization">Normalization</a></li>
  <li><a href="#alpha-diversity" id="toc-alpha-diversity" class="nav-link" data-scroll-target="#alpha-diversity">Alpha diversity</a></li>
  <li><a href="#low-dimensional-representation" id="toc-low-dimensional-representation" class="nav-link" data-scroll-target="#low-dimensional-representation">Low-dimensional representation</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title d-none d-lg-block">Exploratory analysis of compositional data (part II)</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="exploratory-analysis-of-a-microbial-data-set" class="level2">
<h2 class="anchored" data-anchor-id="exploratory-analysis-of-a-microbial-data-set">Exploratory analysis of a microbial data set</h2>
<p>Here we use the “88 soils” data set <span class="citation" data-cites="lauber2009pyrosequencing">(<a href="#ref-lauber2009pyrosequencing" role="doc-biblioref">Lauber et al. 2009</a>)</span> containing bacterial communities in 88 soils from across North and South America.</p>
<section id="load-packages-and-data-data" class="level3">
<h3 class="anchored" data-anchor-id="load-packages-and-data-data">Load packages and data data</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vsn)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microbiome)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gridExtra)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>soil <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/soil_processed/soil.RDS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exploration-of-the-phyloseq-object" class="level3">
<h3 class="anchored" data-anchor-id="exploration-of-the-phyloseq-object">Exploration of the phyloseq object</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>soil</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 116 taxa and 89 samples ]
sample_data() Sample Data:       [ 89 samples by 1 sample variables ]
tax_table()   Taxonomy Table:    [ 116 taxa by 7 taxonomic ranks ]</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read count table</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>otutab <span class="ot">&lt;-</span> <span class="fu">otu_table</span>(soil)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>otutab[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>OTU Table:          [6 taxa and 10 samples]
                     taxa are columns
        1124701 697997 203969 205391 843189 3431064
103.CA2      15      2      0      0      0       3
103.CO3      14      4      0      0      0       1
103.SR3       1      0      0      0      0       1
103.IE2       8      0      0      1      0       0
103.BP1      13     67      0      0      0       3
103.VC2       7      0      0      1      0       7
103.SA2       6      1      0      0      0       0
103.GB2       3      3      0      0      0       0
103.CO2       2      0      0      1      0       5
103.KP1       2      1      0      0      0       1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Taxonomy table</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>taxtab <span class="ot">&lt;-</span> <span class="fu">tax_table</span>(soil)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(taxtab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Taxonomy Table:     [6 taxa by 7 taxonomic ranks]:
        Kingdom       Phylum                Class                    
1124701 "k__Bacteria" "p__Bacteroidetes"    "c__[Saprospirae]"       
697997  "k__Bacteria" "p__Acidobacteria"    "c__[Chloracidobacteria]"
203969  "k__Bacteria" "p__Acidobacteria"    "c__DA052"               
205391  "k__Bacteria" "p__Acidobacteria"    "c__Solibacteres"        
843189  "k__Bacteria" "p__Acidobacteria"    "c__Solibacteres"        
3431064 "k__Bacteria" "p__Gemmatimonadetes" "c__Gemmatimonadetes"    
        Order                Family                Genus                     
1124701 "o__[Saprospirales]" "f__Chitinophagaceae" "g__"                     
697997  "o__RB41"            "f__"                 "g__"                     
203969  "o__Ellin6513"       "f__"                 "g__"                     
205391  "o__Solibacterales"  "f__"                 "g__"                     
843189  "o__Solibacterales"  "f__Solibacteraceae"  "g__Candidatus Solibacter"
3431064 "o__N1423WL"         "f__"                 "g__"                     
        Species
1124701 "s__"  
697997  "s__"  
203969  "s__"  
205391  "s__"  
843189  "s__"  
3431064 "s__"  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample data</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>sampdata <span class="ot">&lt;-</span> <span class="fu">sample_data</span>(soil)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sampdata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          ph
103.CA2 8.02
103.CO3 6.02
103.SR3 6.95
103.IE2 5.52
103.BP1 7.53
103.VC2 5.99</code></pre>
</div>
</div>
</section>
<section id="add-sample-variables" class="level3">
<h3 class="anchored" data-anchor-id="add-sample-variables">Add sample variables</h3>
<p>We add a few sample variables we will need later in this tutorial:</p>
<ul>
<li>sampleID</li>
<li>phType (acidic if pH&lt;6.7 and basic if pH&gt;6.7)</li>
<li>totalReads (total number of reads per sample)</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>sampdata <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sampleID =</span> <span class="fu">rownames</span>(sampdata),</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">totalReads =</span> <span class="fu">sample_sums</span>(soil),</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ph =</span> sampdata<span class="sc">$</span>ph)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>sampdata<span class="sc">$</span>phType <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(<span class="fu">ifelse</span>(sampdata<span class="sc">$</span>ph <span class="sc">&lt;</span> <span class="fl">6.7</span>, <span class="st">"acid"</span>, <span class="st">"basic"</span>)) </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sampdata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        sampleID totalReads   ph phType
103.CA2  103.CA2        205 8.02  basic
103.CO3  103.CO3        233 6.02   acid
103.SR3  103.SR3        169 6.95  basic
103.IE2  103.IE2        244 5.52   acid
103.BP1  103.BP1        297 7.53  basic
103.VC2  103.VC2        256 5.99   acid</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sampdata<span class="sc">$</span>phType)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 acid basic 
   59    30 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add data frame to phyloseq object</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(soil) <span class="ot">&lt;-</span> sampdata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sequencing-depth-library-size" class="level3">
<h3 class="anchored" data-anchor-id="sequencing-depth-library-size">Sequencing depth / library size</h3>
<p>Plot the sequencing depth (total number of reads) of each sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(sampdata, <span class="fu">aes</span>(<span class="at">x =</span> sampleID, <span class="at">y =</span> totalReads)) <span class="sc">+</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Sample"</span>)</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="4_eda_88soils_files/figure-html/sequencing_depth-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(sampdata<span class="sc">$</span>totalReads)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>torm <span class="ot">&lt;-</span> sampdata[<span class="fu">which.min</span>(sampdata<span class="sc">$</span>totalReads), <span class="st">"sampleID"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Remove the sample with only one read count.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>soil <span class="ot">&lt;-</span> <span class="fu">subset_samples</span>(soil, sampleID <span class="sc">!=</span> torm)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>soil</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 116 taxa and 88 samples ]
sample_data() Sample Data:       [ 88 samples by 4 sample variables ]
tax_table()   Taxonomy Table:    [ 116 taxa by 7 taxonomic ranks ]</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>otutab <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">otu_table</span>(soil), <span class="st">"matrix"</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>sampdata <span class="ot">&lt;-</span> <span class="fu">sample_data</span>(soil)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sparsity" class="level3">
<h3 class="anchored" data-anchor-id="sparsity">Sparsity</h3>
<p>Number of zeros and percentage of zeros in the OTU table</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>nvar <span class="ot">&lt;-</span><span class="fu">dim</span>(otutab)[<span class="dv">2</span>]</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>nsamp <span class="ot">&lt;-</span> <span class="fu">dim</span>(otutab)[<span class="dv">1</span>]</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(otutab <span class="sc">==</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6797</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(otutab <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">/</span> (nvar <span class="sc">*</span> nsamp) <span class="sc">*</span> <span class="dv">100</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 66.58503</code></pre>
</div>
</div>
<p>67 of the data are zeros.</p>
</section>
<section id="visualize-microbial-composition" class="level3">
<h3 class="anchored" data-anchor-id="visualize-microbial-composition">Visualize microbial composition</h3>
<p>Have a look at the microbial compositions on phylum level.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Agglomerate to phylum level</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>soil_phyla <span class="ot">&lt;-</span> <span class="fu">tax_glom</span>(soil, <span class="at">taxrank =</span> <span class="st">"Phylum"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Stacked bar plot of the observed “absolute” abundances:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_bar</span>(soil_phyla, <span class="at">fill =</span> <span class="st">"Phylum"</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Absolute abundance"</span>) <span class="sc">+</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Set2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="4_eda_88soils_files/figure-html/barplot_absolute-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Stacked bar plot of the relative abundances:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute relative abundances</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>soil_phyla_rel <span class="ot">&lt;-</span> <span class="fu">transform_sample_counts</span>(soil_phyla, <span class="cf">function</span>(x) x<span class="sc">/</span><span class="fu">sum</span>(x))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_bar</span>(soil_phyla_rel, <span class="at">fill =</span> <span class="st">"Phylum"</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Absolute abundance"</span>) <span class="sc">+</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Set2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="4_eda_88soils_files/figure-html/barplot_relative-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
<section id="zero-replacement" class="level3">
<h3 class="anchored" data-anchor-id="zero-replacement">Zero replacement</h3>
<p>Zero counts are replaced by a unit pseudo count. In doing so, ratios between non-zero counts are preserved, which is not the case if a pseudo count is added to the whole matrix (which is also common).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>soil_zrepl <span class="ot">&lt;-</span> soil</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">otu_table</span>(soil_zrepl)[<span class="fu">otu_table</span>(soil_zrepl) <span class="sc">==</span> <span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>otutab_zrepl <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">otu_table</span>(soil_zrepl), <span class="st">"matrix"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>zCompositions</code> R package provides more complex methods for zero replacement.</p>
</section>
<section id="mean---variance-relationship" class="level3">
<h3 class="anchored" data-anchor-id="mean---variance-relationship">Mean - variance relationship</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>vsn<span class="sc">::</span><span class="fu">meanSdPlot</span>(<span class="fu">t</span>(otutab), <span class="at">plot =</span> <span class="cn">FALSE</span>)<span class="sc">$</span>gg <span class="sc">+</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="4_eda_88soils_files/figure-html/mean-sd-plot_raw-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
<section id="normalization" class="level3">
<h3 class="anchored" data-anchor-id="normalization">Normalization</h3>
<p>We compare three normalization methods we have already seen in the lectures:</p>
<ul>
<li>CLR (centered log-ratio) transformation</li>
<li>Variance stabilizing transformation (vst)</li>
<li>Regularized log transformation (similar to vst but more robust when the size factors vary widely).</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clr</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>soil_clr <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">transform</span>(soil_zrepl, <span class="at">transform =</span> <span class="st">"clr"</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>clr_counts <span class="ot">&lt;-</span> <span class="fu">otu_table</span>(soil_clr)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="co"># vst (function expects samples in columns)</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>vst_counts <span class="ot">&lt;-</span> DESeq2<span class="sc">::</span><span class="fu">varianceStabilizingTransformation</span>(<span class="fu">t</span>(otutab_zrepl), </span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>                                                         <span class="at">fitType =</span> <span class="st">"local"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>converting counts to integer mode</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rlog</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>rlog_counts <span class="ot">&lt;-</span> DESeq2<span class="sc">::</span><span class="fu">rlog</span>(<span class="fu">t</span>(otutab_zrepl), <span class="at">fitType =</span> <span class="st">"local"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>rlog() may take a long time with 50 or more samples,
vst() is a much faster transformation
converting counts to integer mode</code></pre>
</div>
</div>
<p>Create mean-sd plots using the function from <code>vsn</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>pclr <span class="ot">&lt;-</span> <span class="fu">meanSdPlot</span>(<span class="fu">t</span>(clr_counts), <span class="at">plot =</span> <span class="cn">FALSE</span>)<span class="sc">$</span>gg <span class="sc">+</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>)) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"clr"</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>pvst <span class="ot">&lt;-</span> <span class="fu">meanSdPlot</span>(vst_counts, <span class="at">plot =</span> <span class="cn">FALSE</span>)<span class="sc">$</span>gg <span class="sc">+</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>)) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"vst"</span>)</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>prlog <span class="ot">&lt;-</span> <span class="fu">meanSdPlot</span>(rlog_counts, <span class="at">plot =</span> <span class="cn">FALSE</span>)<span class="sc">$</span>gg <span class="sc">+</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>)) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"rlog"</span>)</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(pclr, pvst, prlog, </span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">nrow =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="4_eda_88soils_files/figure-html/mean-sd-plots_norm-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
</section>
<section id="alpha-diversity" class="level3">
<h3 class="anchored" data-anchor-id="alpha-diversity">Alpha diversity</h3>
<p>Another common task is diversity analysis. Alpha diversity summarizes the distribution of species abundances in a given sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>pAlpha <span class="ot">&lt;-</span> <span class="fu">plot_richness</span>(soil,</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">color =</span> <span class="st">"phType"</span>,</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">"Observed"</span>, <span class="st">"Shannon"</span>, <span class="st">"InvSimpson"</span>, <span class="st">"Chao1"</span>),</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">title =</span> <span class="st">"Alpha diveristy for 88 soil data"</span>)</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>pAlpha <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="4_eda_88soils_files/figure-html/alpha_diversity_soil-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section>
<section id="low-dimensional-representation" class="level3">
<h3 class="anchored" data-anchor-id="low-dimensional-representation">Low-dimensional representation</h3>
<p>In this section, we want to plot the data in the two-dimensional space. We therefore compute distances between all samples and perform multi-dimensional scaling (MDS). Here, we use the non-normalized data (see <span class="citation" data-cites="mcKnight2019methods">(<a href="#ref-mcKnight2019methods" role="doc-biblioref">McKnight et al. 2019</a>)</span>)</p>
<p>Distance functions provided by the phyloseq package:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>dist_methods <span class="ot">&lt;-</span> <span class="fu">unlist</span>(distanceMethodList)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>dist_methods</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    UniFrac1     UniFrac2        DPCoA          JSD     vegdist1     vegdist2 
   "unifrac"   "wunifrac"      "dpcoa"        "jsd"  "manhattan"  "euclidean" 
    vegdist3     vegdist4     vegdist5     vegdist6     vegdist7     vegdist8 
  "canberra"       "bray" "kulczynski"    "jaccard"      "gower"   "altGower" 
    vegdist9    vegdist10    vegdist11    vegdist12    vegdist13    vegdist14 
  "morisita"       "horn"  "mountford"       "raup"   "binomial"       "chao" 
   vegdist15   betadiver1   betadiver2   betadiver3   betadiver4   betadiver5 
       "cao"          "w"         "-1"          "c"         "wb"          "r" 
  betadiver6   betadiver7   betadiver8   betadiver9  betadiver10  betadiver11 
         "I"          "e"          "t"         "me"          "j"        "sor" 
 betadiver12  betadiver13  betadiver14  betadiver15  betadiver16  betadiver17 
         "m"         "-2"         "co"         "cc"          "g"         "-3" 
 betadiver18  betadiver19  betadiver20  betadiver21  betadiver22  betadiver23 
         "l"         "19"         "hk"        "rlb"        "sim"         "gl" 
 betadiver24        dist1        dist2        dist3   designdist 
         "z"    "maximum"     "binary"  "minkowski"        "ANY" </code></pre>
</div>
</div>
<section id="mds-plot-with-bray-curtis-distance" class="level4">
<h4 class="anchored" data-anchor-id="mds-plot-with-bray-curtis-distance">MDS plot with Bray Curtis distance</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>soil_mds_bray <span class="ot">&lt;-</span> <span class="fu">ordinate</span>(soil, <span class="at">method =</span> <span class="st">"MDS"</span>, <span class="at">distance =</span> <span class="st">"bray"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>First a plot without coloring metadata.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>mds_soil_bray <span class="ot">&lt;-</span> <span class="fu">plot_ordination</span>(soil, soil_mds_bray, </span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">title =</span> <span class="st">"MDS of 88 soil data"</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>mds_soil_bray <span class="sc">+</span> </span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>)) <span class="sc">+</span> </span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="4_eda_88soils_files/figure-html/MDS_plot_bray-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>One could identify two main clusters, separated by the first axis.</p>
<p>Add the new data frame to the phyloseq object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(soil) <span class="ot">&lt;-</span> sampdata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>mds_soil_bray <span class="ot">&lt;-</span> <span class="fu">plot_ordination</span>(soil, soil_mds_bray, <span class="at">color =</span> <span class="st">"phType"</span>,</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">title =</span> <span class="st">""</span>)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>mds_soil_bray <span class="sc">+</span> </span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>)) <span class="sc">+</span> </span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"MDS of 88 soil data colored by pH type"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="4_eda_88soils_files/figure-html/MDS_plot_bray_labeled-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Indeed, pH type separates samples along Axis.1.</p>
</section>
<section id="mds-plot-with-aitchison-distance" class="level4">
<h4 class="anchored" data-anchor-id="mds-plot-with-aitchison-distance">MDS plot with Aitchison distance</h4>
<p>We generate another MDS plot, but this time using the Aitchison distance, which is simply the Euclidean distance of clr-transformed counts.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>soil_mds_ait <span class="ot">&lt;-</span> <span class="fu">ordinate</span>(soil_clr, <span class="at">method =</span> <span class="st">"MDS"</span>, <span class="at">distance =</span> <span class="st">"euclidean"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(soil_clr) <span class="ot">&lt;-</span> sampdata</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>mds_soil_ait <span class="ot">&lt;-</span> <span class="fu">plot_ordination</span>(soil_clr, soil_mds_ait, <span class="at">color =</span> <span class="st">"phType"</span>,</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">title =</span> <span class="st">""</span>)</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> mds_soil_bray <span class="sc">+</span> </span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>)) <span class="sc">+</span> </span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Bray Curtis dissimilarity"</span>)</span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> mds_soil_ait <span class="sc">+</span> </span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>)) <span class="sc">+</span> </span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Aitchison distance"</span>)</span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(p1, p2,  </span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a>             <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="4_eda_88soils_files/figure-html/MDS_plot_bray_aitchison-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
<section id="mds-plots-for-several-distance-measures" class="level4">
<h4 class="anchored" data-anchor-id="mds-plots-for-several-distance-measures">MDS plots for several distance measures</h4>
<p>Now, we apply all available distance functions to the data, compute the corresponding MDS embeddings, and plot them.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove some unwanted distances</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>sel_dist <span class="ot">&lt;-</span> dist_methods[<span class="sc">!</span>dist_methods <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"unifrac"</span>, <span class="st">"wunifrac"</span>,</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"dpcoa"</span>, <span class="st">"ANY"</span>, <span class="st">"z"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can apply these distance metrics to our data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>gather_ordination_data <span class="ot">&lt;-</span> <span class="cf">function</span>(d_method, phylo_obj){</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate the MDS matrix using the distance</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  ordinate_obj <span class="ot">&lt;-</span> <span class="fu">ordinate</span>(phylo_obj, <span class="at">method =</span> <span class="st">"MDS"</span>, <span class="at">distance =</span> d_method)</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get the data for plotting</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  plot_data <span class="ot">&lt;-</span> <span class="fu">plot_ordination</span>(phylo_obj, ordinate_obj)<span class="sc">$</span>data</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add a column for distance</span></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>  plot_data<span class="sc">$</span>distance <span class="ot">&lt;-</span> d_method</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(plot_data)</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>mds_plot_data <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(<span class="fu">lapply</span>(sel_dist, gather_ordination_data,</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">phylo_obj =</span> soil))</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mds_plot_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 Axis.1      Axis.2 sampleID totalReads   ph phType distance
103.CA2...1 -0.33649635  0.02831767  103.CA2        205 8.02  basic      jsd
103.CO3...2 -0.20666028  0.14139595  103.CO3        233 6.02   acid      jsd
103.SR3...3 -0.28623896 -0.06653403  103.SR3        169 6.95  basic      jsd
103.IE2...4  0.07986531  0.15001745  103.IE2        244 5.52   acid      jsd
103.BP1...5 -0.23627593  0.15166403  103.BP1        297 7.53  basic      jsd
103.VC2...6 -0.05674931  0.23499852  103.VC2        256 5.99   acid      jsd</code></pre>
</div>
</div>
<p>We color again according to pH type.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>mds_soil <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(mds_plot_data, <span class="fu">aes</span>(<span class="at">x=</span>Axis<span class="fl">.1</span>, <span class="at">y=</span>Axis<span class="fl">.2</span>, <span class="at">color=</span>phType)) <span class="sc">+</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> distance, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Axis.1"</span>,</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">y=</span><span class="st">"Axis.2"</span>,</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span><span class="st">"Separation of 88 soil samples according to different distances"</span>)</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>mds_soil <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="4_eda_88soils_files/figure-html/MDS_plots_multiple_dist-1.png" class="img-fluid" width="1344"></p>
</div>
</div>


<!-- -->


</section>
</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-lauber2009pyrosequencing" class="csl-entry" role="doc-biblioentry">
Lauber, Christian L, Micah Hamady, Rob Knight, and Noah Fierer. 2009. <span>“Pyrosequencing-Based Assessment of Soil pH as a Predictor of Soil Bacterial Community Structure at the Continental Scale.”</span> <em>Applied and Environmental Microbiology</em> 75 (15): 5111–20.
</div>
<div id="ref-mcKnight2019methods" class="csl-entry" role="doc-biblioentry">
McKnight, Donald T., Roger Huerlimann, Deborah S. Bower, Lin Schwarzkopf, Ross A. Alford, and Kyall R. Zenger. 2019. <span>“Methods for Normalizing Microbiome Data: An Ecological Perspective.”</span> <em>Methods in Ecology and Evolution</em> 10 (March): 389–400. <a href="https://doi.org/10.1111/2041-210X.13115">https://doi.org/10.1111/2041-210X.13115</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb50" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Exploratory analysis of compositional data (part II)"</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span><span class="co"> </span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a><span class="co">    code-fold: false</span></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a><span class="an">bibliography:</span><span class="co"> references.bib</span></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span><span class="co"> visual</span></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exploratory analysis of a microbial data set</span></span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a>Here we use the "88 soils" data set <span class="co">[</span><span class="ot">@lauber2009pyrosequencing</span><span class="co">]</span> containing bacterial communities in 88 soils from across North and South America.</span>
<span id="cb50-14"><a href="#cb50-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-15"><a href="#cb50-15" aria-hidden="true" tabindex="-1"></a><span class="fu">### Load packages and data data</span></span>
<span id="cb50-16"><a href="#cb50-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-19"><a href="#cb50-19" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-20"><a href="#cb50-20" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load_data</span></span>
<span id="cb50-21"><a href="#cb50-21" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-22"><a href="#cb50-22" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb50-23"><a href="#cb50-23" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-24"><a href="#cb50-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-25"><a href="#cb50-25" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span>
<span id="cb50-26"><a href="#cb50-26" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb50-27"><a href="#cb50-27" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vsn)</span>
<span id="cb50-28"><a href="#cb50-28" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microbiome)</span>
<span id="cb50-29"><a href="#cb50-29" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb50-30"><a href="#cb50-30" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb50-31"><a href="#cb50-31" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gridExtra)</span>
<span id="cb50-32"><a href="#cb50-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-33"><a href="#cb50-33" aria-hidden="true" tabindex="-1"></a>soil <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/soil_processed/soil.RDS"</span>)</span>
<span id="cb50-34"><a href="#cb50-34" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-35"><a href="#cb50-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-36"><a href="#cb50-36" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exploration of the phyloseq object</span></span>
<span id="cb50-37"><a href="#cb50-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-40"><a href="#cb50-40" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-41"><a href="#cb50-41" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: explore_phyloseq</span></span>
<span id="cb50-42"><a href="#cb50-42" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb50-43"><a href="#cb50-43" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb50-44"><a href="#cb50-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-45"><a href="#cb50-45" aria-hidden="true" tabindex="-1"></a>soil</span>
<span id="cb50-46"><a href="#cb50-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-47"><a href="#cb50-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Read count table</span></span>
<span id="cb50-48"><a href="#cb50-48" aria-hidden="true" tabindex="-1"></a>otutab <span class="ot">&lt;-</span> <span class="fu">otu_table</span>(soil)</span>
<span id="cb50-49"><a href="#cb50-49" aria-hidden="true" tabindex="-1"></a>otutab[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span>
<span id="cb50-50"><a href="#cb50-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-51"><a href="#cb50-51" aria-hidden="true" tabindex="-1"></a><span class="co"># Taxonomy table</span></span>
<span id="cb50-52"><a href="#cb50-52" aria-hidden="true" tabindex="-1"></a>taxtab <span class="ot">&lt;-</span> <span class="fu">tax_table</span>(soil)</span>
<span id="cb50-53"><a href="#cb50-53" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(taxtab)</span>
<span id="cb50-54"><a href="#cb50-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-55"><a href="#cb50-55" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample data</span></span>
<span id="cb50-56"><a href="#cb50-56" aria-hidden="true" tabindex="-1"></a>sampdata <span class="ot">&lt;-</span> <span class="fu">sample_data</span>(soil)</span>
<span id="cb50-57"><a href="#cb50-57" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sampdata)</span>
<span id="cb50-58"><a href="#cb50-58" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-59"><a href="#cb50-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-60"><a href="#cb50-60" aria-hidden="true" tabindex="-1"></a><span class="fu">### Add sample variables</span></span>
<span id="cb50-61"><a href="#cb50-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-62"><a href="#cb50-62" aria-hidden="true" tabindex="-1"></a>We add a few sample variables we will need later in this tutorial:</span>
<span id="cb50-63"><a href="#cb50-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-64"><a href="#cb50-64" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>sampleID</span>
<span id="cb50-65"><a href="#cb50-65" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>phType (acidic if pH&lt;6.7 and basic if pH&gt;6.7)</span>
<span id="cb50-66"><a href="#cb50-66" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>totalReads (total number of reads per sample)</span>
<span id="cb50-67"><a href="#cb50-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-70"><a href="#cb50-70" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-71"><a href="#cb50-71" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: add_sample_data</span></span>
<span id="cb50-72"><a href="#cb50-72" aria-hidden="true" tabindex="-1"></a>sampdata <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sampleID =</span> <span class="fu">rownames</span>(sampdata),</span>
<span id="cb50-73"><a href="#cb50-73" aria-hidden="true" tabindex="-1"></a>                       <span class="at">totalReads =</span> <span class="fu">sample_sums</span>(soil),</span>
<span id="cb50-74"><a href="#cb50-74" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ph =</span> sampdata<span class="sc">$</span>ph)</span>
<span id="cb50-75"><a href="#cb50-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-76"><a href="#cb50-76" aria-hidden="true" tabindex="-1"></a>sampdata<span class="sc">$</span>phType <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(<span class="fu">ifelse</span>(sampdata<span class="sc">$</span>ph <span class="sc">&lt;</span> <span class="fl">6.7</span>, <span class="st">"acid"</span>, <span class="st">"basic"</span>)) </span>
<span id="cb50-77"><a href="#cb50-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-78"><a href="#cb50-78" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sampdata)</span>
<span id="cb50-79"><a href="#cb50-79" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sampdata<span class="sc">$</span>phType)</span>
<span id="cb50-80"><a href="#cb50-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-81"><a href="#cb50-81" aria-hidden="true" tabindex="-1"></a><span class="co"># Add data frame to phyloseq object</span></span>
<span id="cb50-82"><a href="#cb50-82" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(soil) <span class="ot">&lt;-</span> sampdata</span>
<span id="cb50-83"><a href="#cb50-83" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-84"><a href="#cb50-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-85"><a href="#cb50-85" aria-hidden="true" tabindex="-1"></a><span class="fu">### Sequencing depth / library size</span></span>
<span id="cb50-86"><a href="#cb50-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-87"><a href="#cb50-87" aria-hidden="true" tabindex="-1"></a>Plot the sequencing depth (total number of reads) of each sample.</span>
<span id="cb50-88"><a href="#cb50-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-91"><a href="#cb50-91" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-92"><a href="#cb50-92" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: sequencing_depth</span></span>
<span id="cb50-93"><a href="#cb50-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-94"><a href="#cb50-94" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(sampdata, <span class="fu">aes</span>(<span class="at">x =</span> sampleID, <span class="at">y =</span> totalReads)) <span class="sc">+</span></span>
<span id="cb50-95"><a href="#cb50-95" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb50-96"><a href="#cb50-96" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb50-97"><a href="#cb50-97" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb50-98"><a href="#cb50-98" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Sample"</span>)</span>
<span id="cb50-99"><a href="#cb50-99" aria-hidden="true" tabindex="-1"></a>p</span>
<span id="cb50-100"><a href="#cb50-100" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-101"><a href="#cb50-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-102"><a href="#cb50-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-105"><a href="#cb50-105" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-106"><a href="#cb50-106" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(sampdata<span class="sc">$</span>totalReads)</span>
<span id="cb50-107"><a href="#cb50-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-108"><a href="#cb50-108" aria-hidden="true" tabindex="-1"></a>torm <span class="ot">&lt;-</span> sampdata[<span class="fu">which.min</span>(sampdata<span class="sc">$</span>totalReads), <span class="st">"sampleID"</span>]</span>
<span id="cb50-109"><a href="#cb50-109" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-110"><a href="#cb50-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-111"><a href="#cb50-111" aria-hidden="true" tabindex="-1"></a>Remove the sample with only one read count.</span>
<span id="cb50-112"><a href="#cb50-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-115"><a href="#cb50-115" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-116"><a href="#cb50-116" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: remove_sample</span></span>
<span id="cb50-117"><a href="#cb50-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-118"><a href="#cb50-118" aria-hidden="true" tabindex="-1"></a>soil <span class="ot">&lt;-</span> <span class="fu">subset_samples</span>(soil, sampleID <span class="sc">!=</span> torm)</span>
<span id="cb50-119"><a href="#cb50-119" aria-hidden="true" tabindex="-1"></a>soil</span>
<span id="cb50-120"><a href="#cb50-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-121"><a href="#cb50-121" aria-hidden="true" tabindex="-1"></a>otutab <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">otu_table</span>(soil), <span class="st">"matrix"</span>)</span>
<span id="cb50-122"><a href="#cb50-122" aria-hidden="true" tabindex="-1"></a>sampdata <span class="ot">&lt;-</span> <span class="fu">sample_data</span>(soil)</span>
<span id="cb50-123"><a href="#cb50-123" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-124"><a href="#cb50-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-125"><a href="#cb50-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-126"><a href="#cb50-126" aria-hidden="true" tabindex="-1"></a><span class="fu">### Sparsity</span></span>
<span id="cb50-127"><a href="#cb50-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-128"><a href="#cb50-128" aria-hidden="true" tabindex="-1"></a>Number of zeros and percentage of zeros in the OTU table</span>
<span id="cb50-129"><a href="#cb50-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-132"><a href="#cb50-132" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-133"><a href="#cb50-133" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: sparsity</span></span>
<span id="cb50-134"><a href="#cb50-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-135"><a href="#cb50-135" aria-hidden="true" tabindex="-1"></a>nvar <span class="ot">&lt;-</span><span class="fu">dim</span>(otutab)[<span class="dv">2</span>]</span>
<span id="cb50-136"><a href="#cb50-136" aria-hidden="true" tabindex="-1"></a>nsamp <span class="ot">&lt;-</span> <span class="fu">dim</span>(otutab)[<span class="dv">1</span>]</span>
<span id="cb50-137"><a href="#cb50-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-138"><a href="#cb50-138" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(otutab <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb50-139"><a href="#cb50-139" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(otutab <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">/</span> (nvar <span class="sc">*</span> nsamp) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb50-140"><a href="#cb50-140" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-141"><a href="#cb50-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-142"><a href="#cb50-142" aria-hidden="true" tabindex="-1"></a><span class="in">`r round(sum(otutab == 0) / (nvar * nsamp) * 100)`</span> of the data are zeros.</span>
<span id="cb50-143"><a href="#cb50-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-144"><a href="#cb50-144" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualize microbial composition</span></span>
<span id="cb50-145"><a href="#cb50-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-146"><a href="#cb50-146" aria-hidden="true" tabindex="-1"></a>Have a look at the microbial compositions on phylum level.</span>
<span id="cb50-147"><a href="#cb50-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-150"><a href="#cb50-150" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-151"><a href="#cb50-151" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: agglomerate_phylum</span></span>
<span id="cb50-152"><a href="#cb50-152" aria-hidden="true" tabindex="-1"></a><span class="co"># Agglomerate to phylum level</span></span>
<span id="cb50-153"><a href="#cb50-153" aria-hidden="true" tabindex="-1"></a>soil_phyla <span class="ot">&lt;-</span> <span class="fu">tax_glom</span>(soil, <span class="at">taxrank =</span> <span class="st">"Phylum"</span>)</span>
<span id="cb50-154"><a href="#cb50-154" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-155"><a href="#cb50-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-156"><a href="#cb50-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-157"><a href="#cb50-157" aria-hidden="true" tabindex="-1"></a>Stacked bar plot of the observed "absolute" abundances:</span>
<span id="cb50-158"><a href="#cb50-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-159"><a href="#cb50-159" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, fig.width=8, fig.height=3}</span></span>
<span id="cb50-160"><a href="#cb50-160" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: barplot_absolute</span></span>
<span id="cb50-161"><a href="#cb50-161" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb50-162"><a href="#cb50-162" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 3</span></span>
<span id="cb50-163"><a href="#cb50-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-164"><a href="#cb50-164" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_bar</span>(soil_phyla, <span class="at">fill =</span> <span class="st">"Phylum"</span>)</span>
<span id="cb50-165"><a href="#cb50-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-166"><a href="#cb50-166" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb50-167"><a href="#cb50-167" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Absolute abundance"</span>) <span class="sc">+</span></span>
<span id="cb50-168"><a href="#cb50-168" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb50-169"><a href="#cb50-169" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Set2"</span>)</span>
<span id="cb50-170"><a href="#cb50-170" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-171"><a href="#cb50-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-172"><a href="#cb50-172" aria-hidden="true" tabindex="-1"></a>Stacked bar plot of the relative abundances:</span>
<span id="cb50-173"><a href="#cb50-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-176"><a href="#cb50-176" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-177"><a href="#cb50-177" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute relative abundances</span></span>
<span id="cb50-178"><a href="#cb50-178" aria-hidden="true" tabindex="-1"></a>soil_phyla_rel <span class="ot">&lt;-</span> <span class="fu">transform_sample_counts</span>(soil_phyla, <span class="cf">function</span>(x) x<span class="sc">/</span><span class="fu">sum</span>(x))</span>
<span id="cb50-179"><a href="#cb50-179" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-180"><a href="#cb50-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-181"><a href="#cb50-181" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, fig.width=8, fig.height=3}</span></span>
<span id="cb50-182"><a href="#cb50-182" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: barplot_relative</span></span>
<span id="cb50-183"><a href="#cb50-183" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb50-184"><a href="#cb50-184" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 3</span></span>
<span id="cb50-185"><a href="#cb50-185" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_bar</span>(soil_phyla_rel, <span class="at">fill =</span> <span class="st">"Phylum"</span>)</span>
<span id="cb50-186"><a href="#cb50-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-187"><a href="#cb50-187" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb50-188"><a href="#cb50-188" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Absolute abundance"</span>) <span class="sc">+</span></span>
<span id="cb50-189"><a href="#cb50-189" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb50-190"><a href="#cb50-190" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Set2"</span>)</span>
<span id="cb50-191"><a href="#cb50-191" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-192"><a href="#cb50-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-193"><a href="#cb50-193" aria-hidden="true" tabindex="-1"></a><span class="fu">### Zero replacement </span></span>
<span id="cb50-194"><a href="#cb50-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-195"><a href="#cb50-195" aria-hidden="true" tabindex="-1"></a>Zero counts are replaced by a unit pseudo count. In doing so, ratios between non-zero counts are preserved, which is not the case if a pseudo count is added to the whole matrix (which is also common).</span>
<span id="cb50-196"><a href="#cb50-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-199"><a href="#cb50-199" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-200"><a href="#cb50-200" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: replace_zeros</span></span>
<span id="cb50-201"><a href="#cb50-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-202"><a href="#cb50-202" aria-hidden="true" tabindex="-1"></a>soil_zrepl <span class="ot">&lt;-</span> soil</span>
<span id="cb50-203"><a href="#cb50-203" aria-hidden="true" tabindex="-1"></a><span class="fu">otu_table</span>(soil_zrepl)[<span class="fu">otu_table</span>(soil_zrepl) <span class="sc">==</span> <span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb50-204"><a href="#cb50-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-205"><a href="#cb50-205" aria-hidden="true" tabindex="-1"></a>otutab_zrepl <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">otu_table</span>(soil_zrepl), <span class="st">"matrix"</span>)</span>
<span id="cb50-206"><a href="#cb50-206" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-207"><a href="#cb50-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-208"><a href="#cb50-208" aria-hidden="true" tabindex="-1"></a>The <span class="in">`zCompositions`</span> R package provides more complex methods for zero replacement.</span>
<span id="cb50-209"><a href="#cb50-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-210"><a href="#cb50-210" aria-hidden="true" tabindex="-1"></a><span class="fu">### Mean - variance relationship</span></span>
<span id="cb50-211"><a href="#cb50-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-212"><a href="#cb50-212" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, fig.width=8, fig.height=5}</span></span>
<span id="cb50-213"><a href="#cb50-213" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: mean-sd-plot_raw</span></span>
<span id="cb50-214"><a href="#cb50-214" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb50-215"><a href="#cb50-215" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb50-216"><a href="#cb50-216" aria-hidden="true" tabindex="-1"></a>vsn<span class="sc">::</span><span class="fu">meanSdPlot</span>(<span class="fu">t</span>(otutab), <span class="at">plot =</span> <span class="cn">FALSE</span>)<span class="sc">$</span>gg <span class="sc">+</span></span>
<span id="cb50-217"><a href="#cb50-217" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>))</span>
<span id="cb50-218"><a href="#cb50-218" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-219"><a href="#cb50-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-220"><a href="#cb50-220" aria-hidden="true" tabindex="-1"></a><span class="fu">### Normalization</span></span>
<span id="cb50-221"><a href="#cb50-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-222"><a href="#cb50-222" aria-hidden="true" tabindex="-1"></a>We compare three normalization methods we have already seen in the lectures:</span>
<span id="cb50-223"><a href="#cb50-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-224"><a href="#cb50-224" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>CLR (centered log-ratio) transformation</span>
<span id="cb50-225"><a href="#cb50-225" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Variance stabilizing transformation (vst)</span>
<span id="cb50-226"><a href="#cb50-226" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Regularized log transformation (similar to vst but more robust when the size factors vary widely).</span>
<span id="cb50-227"><a href="#cb50-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-230"><a href="#cb50-230" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-231"><a href="#cb50-231" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: normalization</span></span>
<span id="cb50-232"><a href="#cb50-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-233"><a href="#cb50-233" aria-hidden="true" tabindex="-1"></a><span class="co"># clr</span></span>
<span id="cb50-234"><a href="#cb50-234" aria-hidden="true" tabindex="-1"></a>soil_clr <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">transform</span>(soil_zrepl, <span class="at">transform =</span> <span class="st">"clr"</span>)</span>
<span id="cb50-235"><a href="#cb50-235" aria-hidden="true" tabindex="-1"></a>clr_counts <span class="ot">&lt;-</span> <span class="fu">otu_table</span>(soil_clr)</span>
<span id="cb50-236"><a href="#cb50-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-237"><a href="#cb50-237" aria-hidden="true" tabindex="-1"></a><span class="co"># vst (function expects samples in columns)</span></span>
<span id="cb50-238"><a href="#cb50-238" aria-hidden="true" tabindex="-1"></a>vst_counts <span class="ot">&lt;-</span> DESeq2<span class="sc">::</span><span class="fu">varianceStabilizingTransformation</span>(<span class="fu">t</span>(otutab_zrepl), </span>
<span id="cb50-239"><a href="#cb50-239" aria-hidden="true" tabindex="-1"></a>                                                         <span class="at">fitType =</span> <span class="st">"local"</span>)</span>
<span id="cb50-240"><a href="#cb50-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-241"><a href="#cb50-241" aria-hidden="true" tabindex="-1"></a><span class="co"># rlog</span></span>
<span id="cb50-242"><a href="#cb50-242" aria-hidden="true" tabindex="-1"></a>rlog_counts <span class="ot">&lt;-</span> DESeq2<span class="sc">::</span><span class="fu">rlog</span>(<span class="fu">t</span>(otutab_zrepl), <span class="at">fitType =</span> <span class="st">"local"</span>)</span>
<span id="cb50-243"><a href="#cb50-243" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-244"><a href="#cb50-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-245"><a href="#cb50-245" aria-hidden="true" tabindex="-1"></a>Create mean-sd plots using the function from <span class="in">`vsn`</span> package.</span>
<span id="cb50-246"><a href="#cb50-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-247"><a href="#cb50-247" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, fig.width=12, fig.height=4}</span></span>
<span id="cb50-248"><a href="#cb50-248" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: mean-sd-plots_norm</span></span>
<span id="cb50-249"><a href="#cb50-249" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 12</span></span>
<span id="cb50-250"><a href="#cb50-250" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb50-251"><a href="#cb50-251" aria-hidden="true" tabindex="-1"></a>pclr <span class="ot">&lt;-</span> <span class="fu">meanSdPlot</span>(<span class="fu">t</span>(clr_counts), <span class="at">plot =</span> <span class="cn">FALSE</span>)<span class="sc">$</span>gg <span class="sc">+</span></span>
<span id="cb50-252"><a href="#cb50-252" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>)) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"clr"</span>)</span>
<span id="cb50-253"><a href="#cb50-253" aria-hidden="true" tabindex="-1"></a>pvst <span class="ot">&lt;-</span> <span class="fu">meanSdPlot</span>(vst_counts, <span class="at">plot =</span> <span class="cn">FALSE</span>)<span class="sc">$</span>gg <span class="sc">+</span></span>
<span id="cb50-254"><a href="#cb50-254" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>)) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"vst"</span>)</span>
<span id="cb50-255"><a href="#cb50-255" aria-hidden="true" tabindex="-1"></a>prlog <span class="ot">&lt;-</span> <span class="fu">meanSdPlot</span>(rlog_counts, <span class="at">plot =</span> <span class="cn">FALSE</span>)<span class="sc">$</span>gg <span class="sc">+</span></span>
<span id="cb50-256"><a href="#cb50-256" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>)) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"rlog"</span>)</span>
<span id="cb50-257"><a href="#cb50-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-258"><a href="#cb50-258" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(pclr, pvst, prlog, </span>
<span id="cb50-259"><a href="#cb50-259" aria-hidden="true" tabindex="-1"></a>             <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">nrow =</span> <span class="dv">1</span>)</span>
<span id="cb50-260"><a href="#cb50-260" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-261"><a href="#cb50-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-262"><a href="#cb50-262" aria-hidden="true" tabindex="-1"></a><span class="fu">### Alpha diversity</span></span>
<span id="cb50-263"><a href="#cb50-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-264"><a href="#cb50-264" aria-hidden="true" tabindex="-1"></a>Another common task is diversity analysis. Alpha diversity summarizes the distribution of species abundances in a given sample.</span>
<span id="cb50-265"><a href="#cb50-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-266"><a href="#cb50-266" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, fig.width=10, fig.height=4}</span></span>
<span id="cb50-267"><a href="#cb50-267" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: alpha_diversity_soil</span></span>
<span id="cb50-268"><a href="#cb50-268" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb50-269"><a href="#cb50-269" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb50-270"><a href="#cb50-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-271"><a href="#cb50-271" aria-hidden="true" tabindex="-1"></a>pAlpha <span class="ot">&lt;-</span> <span class="fu">plot_richness</span>(soil,</span>
<span id="cb50-272"><a href="#cb50-272" aria-hidden="true" tabindex="-1"></a>                       <span class="at">color =</span> <span class="st">"phType"</span>,</span>
<span id="cb50-273"><a href="#cb50-273" aria-hidden="true" tabindex="-1"></a>                       <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">"Observed"</span>, <span class="st">"Shannon"</span>, <span class="st">"InvSimpson"</span>, <span class="st">"Chao1"</span>),</span>
<span id="cb50-274"><a href="#cb50-274" aria-hidden="true" tabindex="-1"></a>                       <span class="at">title =</span> <span class="st">"Alpha diveristy for 88 soil data"</span>)</span>
<span id="cb50-275"><a href="#cb50-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-276"><a href="#cb50-276" aria-hidden="true" tabindex="-1"></a>pAlpha <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>())</span>
<span id="cb50-277"><a href="#cb50-277" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-278"><a href="#cb50-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-279"><a href="#cb50-279" aria-hidden="true" tabindex="-1"></a><span class="fu">### Low-dimensional representation</span></span>
<span id="cb50-280"><a href="#cb50-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-281"><a href="#cb50-281" aria-hidden="true" tabindex="-1"></a>In this section, we want to plot the data in the two-dimensional space. We therefore compute distances between all samples and perform multi-dimensional scaling (MDS). Here, we use the non-normalized data (see <span class="co">[</span><span class="ot">@mcKnight2019methods</span><span class="co">]</span>)</span>
<span id="cb50-282"><a href="#cb50-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-283"><a href="#cb50-283" aria-hidden="true" tabindex="-1"></a>Distance functions provided by the phyloseq package:</span>
<span id="cb50-284"><a href="#cb50-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-287"><a href="#cb50-287" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-288"><a href="#cb50-288" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: distances</span></span>
<span id="cb50-289"><a href="#cb50-289" aria-hidden="true" tabindex="-1"></a>dist_methods <span class="ot">&lt;-</span> <span class="fu">unlist</span>(distanceMethodList)</span>
<span id="cb50-290"><a href="#cb50-290" aria-hidden="true" tabindex="-1"></a>dist_methods</span>
<span id="cb50-291"><a href="#cb50-291" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-292"><a href="#cb50-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-293"><a href="#cb50-293" aria-hidden="true" tabindex="-1"></a><span class="fu">#### MDS plot with Bray Curtis distance</span></span>
<span id="cb50-294"><a href="#cb50-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-297"><a href="#cb50-297" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-298"><a href="#cb50-298" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: mds_bray</span></span>
<span id="cb50-299"><a href="#cb50-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-300"><a href="#cb50-300" aria-hidden="true" tabindex="-1"></a>soil_mds_bray <span class="ot">&lt;-</span> <span class="fu">ordinate</span>(soil, <span class="at">method =</span> <span class="st">"MDS"</span>, <span class="at">distance =</span> <span class="st">"bray"</span>)</span>
<span id="cb50-301"><a href="#cb50-301" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-302"><a href="#cb50-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-303"><a href="#cb50-303" aria-hidden="true" tabindex="-1"></a>First a plot without coloring metadata.</span>
<span id="cb50-304"><a href="#cb50-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-305"><a href="#cb50-305" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, fig.width=8, fig.height=5}</span></span>
<span id="cb50-306"><a href="#cb50-306" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: MDS_plot_bray</span></span>
<span id="cb50-307"><a href="#cb50-307" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb50-308"><a href="#cb50-308" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb50-309"><a href="#cb50-309" aria-hidden="true" tabindex="-1"></a>mds_soil_bray <span class="ot">&lt;-</span> <span class="fu">plot_ordination</span>(soil, soil_mds_bray, </span>
<span id="cb50-310"><a href="#cb50-310" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">title =</span> <span class="st">"MDS of 88 soil data"</span>)</span>
<span id="cb50-311"><a href="#cb50-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-312"><a href="#cb50-312" aria-hidden="true" tabindex="-1"></a>mds_soil_bray <span class="sc">+</span> </span>
<span id="cb50-313"><a href="#cb50-313" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb50-314"><a href="#cb50-314" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>)) <span class="sc">+</span> </span>
<span id="cb50-315"><a href="#cb50-315" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>)</span>
<span id="cb50-316"><a href="#cb50-316" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-317"><a href="#cb50-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-318"><a href="#cb50-318" aria-hidden="true" tabindex="-1"></a>One could identify two main clusters, separated by the first axis. </span>
<span id="cb50-319"><a href="#cb50-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-320"><a href="#cb50-320" aria-hidden="true" tabindex="-1"></a>Add the new data frame to the phyloseq object.</span>
<span id="cb50-321"><a href="#cb50-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-324"><a href="#cb50-324" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-325"><a href="#cb50-325" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(soil) <span class="ot">&lt;-</span> sampdata</span>
<span id="cb50-326"><a href="#cb50-326" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-327"><a href="#cb50-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-328"><a href="#cb50-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-329"><a href="#cb50-329" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, fig.width=8, fig.height=5}</span></span>
<span id="cb50-330"><a href="#cb50-330" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: MDS_plot_bray_labeled</span></span>
<span id="cb50-331"><a href="#cb50-331" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb50-332"><a href="#cb50-332" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb50-333"><a href="#cb50-333" aria-hidden="true" tabindex="-1"></a>mds_soil_bray <span class="ot">&lt;-</span> <span class="fu">plot_ordination</span>(soil, soil_mds_bray, <span class="at">color =</span> <span class="st">"phType"</span>,</span>
<span id="cb50-334"><a href="#cb50-334" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">title =</span> <span class="st">""</span>)</span>
<span id="cb50-335"><a href="#cb50-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-336"><a href="#cb50-336" aria-hidden="true" tabindex="-1"></a>mds_soil_bray <span class="sc">+</span> </span>
<span id="cb50-337"><a href="#cb50-337" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb50-338"><a href="#cb50-338" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>)) <span class="sc">+</span> </span>
<span id="cb50-339"><a href="#cb50-339" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb50-340"><a href="#cb50-340" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"MDS of 88 soil data colored by pH type"</span>)</span>
<span id="cb50-341"><a href="#cb50-341" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-342"><a href="#cb50-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-343"><a href="#cb50-343" aria-hidden="true" tabindex="-1"></a>Indeed, pH type separates samples along Axis.1. </span>
<span id="cb50-344"><a href="#cb50-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-345"><a href="#cb50-345" aria-hidden="true" tabindex="-1"></a><span class="fu">#### MDS plot with Aitchison distance</span></span>
<span id="cb50-346"><a href="#cb50-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-347"><a href="#cb50-347" aria-hidden="true" tabindex="-1"></a>We generate another MDS plot, but this time using the Aitchison distance, which is simply the Euclidean distance of clr-transformed counts.</span>
<span id="cb50-348"><a href="#cb50-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-351"><a href="#cb50-351" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-352"><a href="#cb50-352" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: mds_aitchison</span></span>
<span id="cb50-353"><a href="#cb50-353" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-354"><a href="#cb50-354" aria-hidden="true" tabindex="-1"></a>soil_mds_ait <span class="ot">&lt;-</span> <span class="fu">ordinate</span>(soil_clr, <span class="at">method =</span> <span class="st">"MDS"</span>, <span class="at">distance =</span> <span class="st">"euclidean"</span>)</span>
<span id="cb50-355"><a href="#cb50-355" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-356"><a href="#cb50-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-357"><a href="#cb50-357" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, fig.width=8, fig.height=4}</span></span>
<span id="cb50-358"><a href="#cb50-358" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: MDS_plot_bray_aitchison</span></span>
<span id="cb50-359"><a href="#cb50-359" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb50-360"><a href="#cb50-360" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb50-361"><a href="#cb50-361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-362"><a href="#cb50-362" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(soil_clr) <span class="ot">&lt;-</span> sampdata</span>
<span id="cb50-363"><a href="#cb50-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-364"><a href="#cb50-364" aria-hidden="true" tabindex="-1"></a>mds_soil_ait <span class="ot">&lt;-</span> <span class="fu">plot_ordination</span>(soil_clr, soil_mds_ait, <span class="at">color =</span> <span class="st">"phType"</span>,</span>
<span id="cb50-365"><a href="#cb50-365" aria-hidden="true" tabindex="-1"></a>                                <span class="at">title =</span> <span class="st">""</span>)</span>
<span id="cb50-366"><a href="#cb50-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-367"><a href="#cb50-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-368"><a href="#cb50-368" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> mds_soil_bray <span class="sc">+</span> </span>
<span id="cb50-369"><a href="#cb50-369" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb50-370"><a href="#cb50-370" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>)) <span class="sc">+</span> </span>
<span id="cb50-371"><a href="#cb50-371" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb50-372"><a href="#cb50-372" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Bray Curtis dissimilarity"</span>)</span>
<span id="cb50-373"><a href="#cb50-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-374"><a href="#cb50-374" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> mds_soil_ait <span class="sc">+</span> </span>
<span id="cb50-375"><a href="#cb50-375" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb50-376"><a href="#cb50-376" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>)) <span class="sc">+</span> </span>
<span id="cb50-377"><a href="#cb50-377" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb50-378"><a href="#cb50-378" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Aitchison distance"</span>)</span>
<span id="cb50-379"><a href="#cb50-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-380"><a href="#cb50-380" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(p1, p2,  </span>
<span id="cb50-381"><a href="#cb50-381" aria-hidden="true" tabindex="-1"></a>             <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">1</span>)</span>
<span id="cb50-382"><a href="#cb50-382" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-383"><a href="#cb50-383" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-384"><a href="#cb50-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-385"><a href="#cb50-385" aria-hidden="true" tabindex="-1"></a><span class="fu">#### MDS plots for several distance measures</span></span>
<span id="cb50-386"><a href="#cb50-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-387"><a href="#cb50-387" aria-hidden="true" tabindex="-1"></a>Now, we apply all available distance functions to the data, compute the corresponding MDS embeddings, and plot them.</span>
<span id="cb50-388"><a href="#cb50-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-391"><a href="#cb50-391" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-392"><a href="#cb50-392" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove some unwanted distances</span></span>
<span id="cb50-393"><a href="#cb50-393" aria-hidden="true" tabindex="-1"></a>sel_dist <span class="ot">&lt;-</span> dist_methods[<span class="sc">!</span>dist_methods <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"unifrac"</span>, <span class="st">"wunifrac"</span>,</span>
<span id="cb50-394"><a href="#cb50-394" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"dpcoa"</span>, <span class="st">"ANY"</span>, <span class="st">"z"</span>)]</span>
<span id="cb50-395"><a href="#cb50-395" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-396"><a href="#cb50-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-397"><a href="#cb50-397" aria-hidden="true" tabindex="-1"></a>Now we can apply these distance metrics to our data. </span>
<span id="cb50-398"><a href="#cb50-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-401"><a href="#cb50-401" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb50-402"><a href="#cb50-402" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: MDS_multiple_dist</span></span>
<span id="cb50-403"><a href="#cb50-403" aria-hidden="true" tabindex="-1"></a>gather_ordination_data <span class="ot">&lt;-</span> <span class="cf">function</span>(d_method, phylo_obj){</span>
<span id="cb50-404"><a href="#cb50-404" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate the MDS matrix using the distance</span></span>
<span id="cb50-405"><a href="#cb50-405" aria-hidden="true" tabindex="-1"></a>  ordinate_obj <span class="ot">&lt;-</span> <span class="fu">ordinate</span>(phylo_obj, <span class="at">method =</span> <span class="st">"MDS"</span>, <span class="at">distance =</span> d_method)</span>
<span id="cb50-406"><a href="#cb50-406" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get the data for plotting</span></span>
<span id="cb50-407"><a href="#cb50-407" aria-hidden="true" tabindex="-1"></a>  plot_data <span class="ot">&lt;-</span> <span class="fu">plot_ordination</span>(phylo_obj, ordinate_obj)<span class="sc">$</span>data</span>
<span id="cb50-408"><a href="#cb50-408" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add a column for distance</span></span>
<span id="cb50-409"><a href="#cb50-409" aria-hidden="true" tabindex="-1"></a>  plot_data<span class="sc">$</span>distance <span class="ot">&lt;-</span> d_method</span>
<span id="cb50-410"><a href="#cb50-410" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(plot_data)</span>
<span id="cb50-411"><a href="#cb50-411" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb50-412"><a href="#cb50-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-413"><a href="#cb50-413" aria-hidden="true" tabindex="-1"></a>mds_plot_data <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(<span class="fu">lapply</span>(sel_dist, gather_ordination_data,</span>
<span id="cb50-414"><a href="#cb50-414" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">phylo_obj =</span> soil))</span>
<span id="cb50-415"><a href="#cb50-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-416"><a href="#cb50-416" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mds_plot_data)</span>
<span id="cb50-417"><a href="#cb50-417" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb50-418"><a href="#cb50-418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-419"><a href="#cb50-419" aria-hidden="true" tabindex="-1"></a>We color again according to pH type.</span>
<span id="cb50-420"><a href="#cb50-420" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-421"><a href="#cb50-421" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, fig.width=14, fig.height=9}</span></span>
<span id="cb50-422"><a href="#cb50-422" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: MDS_plots_multiple_dist</span></span>
<span id="cb50-423"><a href="#cb50-423" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 14</span></span>
<span id="cb50-424"><a href="#cb50-424" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 9</span></span>
<span id="cb50-425"><a href="#cb50-425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-426"><a href="#cb50-426" aria-hidden="true" tabindex="-1"></a>mds_soil <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(mds_plot_data, <span class="fu">aes</span>(<span class="at">x=</span>Axis<span class="fl">.1</span>, <span class="at">y=</span>Axis<span class="fl">.2</span>, <span class="at">color=</span>phType)) <span class="sc">+</span></span>
<span id="cb50-427"><a href="#cb50-427" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb50-428"><a href="#cb50-428" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> distance, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb50-429"><a href="#cb50-429" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Axis.1"</span>,</span>
<span id="cb50-430"><a href="#cb50-430" aria-hidden="true" tabindex="-1"></a>       <span class="at">y=</span><span class="st">"Axis.2"</span>,</span>
<span id="cb50-431"><a href="#cb50-431" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span><span class="st">"Separation of 88 soil samples according to different distances"</span>)</span>
<span id="cb50-432"><a href="#cb50-432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-433"><a href="#cb50-433" aria-hidden="true" tabindex="-1"></a>mds_soil <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>))</span>
<span id="cb50-434"><a href="#cb50-434" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>