<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Mini-course CoDA - Practical applications of learned concepts in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Mini-course CoDA</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html">
 <span class="menu-text">About</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">Practical applications of learned concepts in R</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation docked overflow-auto">
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./4_eda.html" class="sidebar-item-text sidebar-link">Exploratory data analysis of compositional data</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./9_practical.html" class="sidebar-item-text sidebar-link">Practical applications of learned concepts in R</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#network-learning-and-analysis" id="toc-network-learning-and-analysis" class="nav-link active" data-scroll-target="#network-learning-and-analysis">Network learning and analysis</a>
  <ul class="collapse">
  <li><a href="#install-and-load-packages" id="toc-install-and-load-packages" class="nav-link" data-scroll-target="#install-and-load-packages">Install and load packages</a></li>
  <li><a href="#load-the-data" id="toc-load-the-data" class="nav-link" data-scroll-target="#load-the-data">Load the data</a></li>
  <li><a href="#rename-genera" id="toc-rename-genera" class="nav-link" data-scroll-target="#rename-genera">Rename genera</a></li>
  <li><a href="#network-analysis" id="toc-network-analysis" class="nav-link" data-scroll-target="#network-analysis">Network analysis</a></li>
  <li><a href="#highlight-node-properties" id="toc-highlight-node-properties" class="nav-link" data-scroll-target="#highlight-node-properties">Highlight node properties</a></li>
  <li><a href="#highlight-data-features" id="toc-highlight-data-features" class="nav-link" data-scroll-target="#highlight-data-features">Highlight data features</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title d-none d-lg-block">Practical applications of learned concepts in R</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="network-learning-and-analysis" class="level2">
<h2 class="anchored" data-anchor-id="network-learning-and-analysis">Network learning and analysis</h2>
<p>Here, we learn how a microbial association network is constructed and analyzed using the R package NetCoMi.</p>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
The R package NetCoMi
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>The <a href="https://github.com/stefpeschel/NetCoMi">NetCoMi</a> <span class="citation" data-cites="peschel2021netcomi">(<a href="#ref-peschel2021netcomi" role="doc-biblioref">Peschel et al. 2021</a>)</span> package is specifically designed to <strong>construct, analyze, and compare networks</strong> for microbiome data. It implements the complete workflow of learning and analyzing microbial networks explained in lecture 8. Instead of using several functions for each of the steps, <code>NetCoMi</code> provides a single function for network construction (<code>netConstruct()</code>), so the package streamlines the workflow considerably. The user can choose from a variety of methods for <strong>data preprocessing, association estimation, sparsification, and transformation</strong>. The returned <code>microNet</code> object can then be passed to <code>netAnalyze()</code> (the network analysis function) so that all necessary information is available for the network analysis workflow.</p>
</div>
</div>
</div>
<section id="install-and-load-packages" class="level3">
<h3 class="anchored" data-anchor-id="install-and-load-packages">Install and load packages</h3>
<p>We start by installing NetCoMi from GitHub. Two of NetCoMi’s dependencies are also only available on GitHub. We will install them manually beforehand to avoid problems.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"zdk123/SpiecEasi"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"GraceYoon/SPRING"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"stefpeschel/NetCoMi"</span>, <span class="at">force =</span> <span class="cn">TRUE</span>, <span class="at">ref =</span> <span class="st">"TSE"</span>,</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>                         <span class="at">dependencies =</span> <span class="fu">c</span>(<span class="st">"Depends"</span>, <span class="st">"Imports"</span>, <span class="st">"LinkingTo"</span>),</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>                         <span class="at">repos =</span> <span class="fu">c</span>(<span class="st">"https://cloud.r-project.org/"</span>,</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>                                   BiocManager<span class="sc">::</span><span class="fu">repositories</span>()))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load packages needed in this exercise.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(NetCoMi)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SpiecEasi</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-the-data" class="level3">
<h3 class="anchored" data-anchor-id="load-the-data">Load the data</h3>
<p>The 88 soil dataset <span class="citation" data-cites="lauber2009pyrosequencing">(<a href="#ref-lauber2009pyrosequencing" role="doc-biblioref">Lauber et al. 2009</a>)</span> is again used in this example.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>soil <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/soil_processed/soil.RDS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Take a look at the data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>soil</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 116 taxa and 89 samples ]
sample_data() Sample Data:       [ 89 samples by 1 sample variables ]
tax_table()   Taxonomy Table:    [ 116 taxa by 7 taxonomic ranks ]</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read count table</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">otu_table</span>(soil)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>OTU Table:          [6 taxa and 10 samples]
                     taxa are columns
        1124701 697997 203969 205391 843189 3431064
103.CA2      15      2      0      0      0       3
103.CO3      14      4      0      0      0       1
103.SR3       1      0      0      0      0       1
103.IE2       8      0      0      1      0       0
103.BP1      13     67      0      0      0       3
103.VC2       7      0      0      1      0       7
103.SA2       6      1      0      0      0       0
103.GB2       3      3      0      0      0       0
103.CO2       2      0      0      1      0       5
103.KP1       2      1      0      0      0       1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Taxonomy table</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>taxtab <span class="ot">&lt;-</span> <span class="fu">tax_table</span>(soil)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(taxtab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Taxonomy Table:     [6 taxa by 7 taxonomic ranks]:
        Kingdom       Phylum                Class                    
1124701 "k__Bacteria" "p__Bacteroidetes"    "c__[Saprospirae]"       
697997  "k__Bacteria" "p__Acidobacteria"    "c__[Chloracidobacteria]"
203969  "k__Bacteria" "p__Acidobacteria"    "c__DA052"               
205391  "k__Bacteria" "p__Acidobacteria"    "c__Solibacteres"        
843189  "k__Bacteria" "p__Acidobacteria"    "c__Solibacteres"        
3431064 "k__Bacteria" "p__Gemmatimonadetes" "c__Gemmatimonadetes"    
        Order                Family                Genus                     
1124701 "o__[Saprospirales]" "f__Chitinophagaceae" "g__"                     
697997  "o__RB41"            "f__"                 "g__"                     
203969  "o__Ellin6513"       "f__"                 "g__"                     
205391  "o__Solibacterales"  "f__"                 "g__"                     
843189  "o__Solibacterales"  "f__Solibacteraceae"  "g__Candidatus Solibacter"
3431064 "o__N1423WL"         "f__"                 "g__"                     
        Species
1124701 "s__"  
697997  "s__"  
203969  "s__"  
205391  "s__"  
843189  "s__"  
3431064 "s__"  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample data</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>sampdata <span class="ot">&lt;-</span> <span class="fu">sample_data</span>(soil)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sampdata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          ph
103.CA2 8.02
103.CO3 6.02
103.SR3 6.95
103.IE2 5.52
103.BP1 7.53
103.VC2 5.99</code></pre>
</div>
</div>
</section>
<section id="rename-genera" class="level3">
<h3 class="anchored" data-anchor-id="rename-genera">Rename genera</h3>
<p>If we want to generate a network on genus level, we run into two problems:</p>
<ul>
<li>There are unidentified genera.</li>
<li>Not all genera are unique.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(taxtab[, <span class="st">"Genus"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
                     g__              g__Bacillus           g__Balneimonas 
                      83                        1                        4 
       g__Bradyrhizobium          g__Burkholderia g__Candidatus Solibacter 
                       4                        1                        2 
         g__Edaphobacter        g__Flavobacterium           g__Gallionella 
                       1                        1                        1 
     g__Geodermatophilus          g__Kaistobacter         g__Mycobacterium 
                       1                        1                        1 
           g__Phormidium        g__Planomicrobium           g__Rhodoplanes 
                       1                        1                        8 
          g__Rubrobacter          g__Sporosarcina 
                       4                        1 </code></pre>
</div>
</div>
<p>Solutions:</p>
<ul>
<li>Unknown genera are named by higher taxonomic ranks.</li>
<li>Duplicates are numbered to make them unique.</li>
</ul>
<p>NetCoMi provides a function <code>renameTaxa()</code>, which does this job for us. It can also be used to rename the taxa according to a desired pattern.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename taxonomic table and make Rank6 (genus) unique</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>soil_renamed <span class="ot">&lt;-</span> <span class="fu">renameTaxa</span>(soil, </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">pat =</span> <span class="st">"&lt;name&gt;"</span>, </span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">substPat =</span> <span class="st">"&lt;name&gt;_&lt;subst_name&gt;(&lt;subst_R&gt;)"</span>,</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>                           <span class="at">numDupli =</span> <span class="st">"Genus"</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>taxtab_renamed <span class="ot">&lt;-</span> <span class="fu">tax_table</span>(soil_renamed)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">cbind</span>(taxtab[, <span class="st">"Genus"</span>], taxtab_renamed[, <span class="st">"Genus"</span>]), <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Genus                      Genus                   
1124701 "g__"                      "1_Chitinophagaceae(F)" 
697997  "g__"                      "2_RB41(O)"             
203969  "g__"                      "3_Ellin6513(O)"        
205391  "g__"                      "4_Solibacterales(O)"   
843189  "g__Candidatus Solibacter" "Candidatus Solibacter1"
3431064 "g__"                      "5_N1423WL(O)"          
156611  "g__"                      "6_Acidobacteriaceae(F)"
2867534 "g__Candidatus Solibacter" "Candidatus Solibacter2"
4440638 "g__"                      "7_RB41(O)"             
216549  "g__"                      "8_RB41(O)"             
4440611 "g__"                      "9_RB41(O)"             
4416927 "g__Balneimonas"           "Balneimonas1"          
2578357 "g__"                      "10_Ellin6513(O)"       
1122202 "g__"                      "11_RB41(O)"            
3616127 "g__"                      "12_Solibacterales(O)"  </code></pre>
</div>
</div>
<p>Now, we can pass the phyloseq object with renamed genera to NetCoMi’s <code>netConstruct()</code> function.</p>
<p>SpiecEasi is used in the “MB” mode (Meinshausen &amp; Bühlman neighborhood selection) to measure associations between the genera.</p>
<p>A few comments on the arguments:</p>
<ul>
<li>The network is created on genus level.</li>
<li>The data are filtered as follows:
<ul>
<li>Only samples with a total of at least 100 reads are included (argument <code>filtSamp</code>).</li>
<li>Only taxa occuring in at least 5 samples are included (argument <code>filtTax</code>).</li>
</ul></li>
<li><code>measure</code> defines the association or dissimilarity measure, in our case SpiecEasi. Additional arguments are passed to <code>spiec.easi()</code> via <code>measurePar</code>. We reduce the number of StARS repetitions to 10 to reduce runtime.</li>
<li>Normalization and zero handling is done internally by <code>spiec.easi()</code>, so we set <code>normMethod</code> and <code>zeroMethod</code> to “none”.</li>
<li><code>sparsMethod</code> is also set to “none” (SpiecEasi includes StARS model selection to obtain a sparse network).</li>
<li>The “signed” method is used to transform associations into dissimilarities (argument <code>dissFunc</code>). Thus, strongly negatively associated taxa have a high dissimilarity and conversely a low similarity, corresponding to low edge weights on the network plot.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>net <span class="ot">&lt;-</span> <span class="fu">netConstruct</span>(soil_renamed,</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">taxRank =</span> <span class="st">"Genus"</span>,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">filtSamp =</span> <span class="st">"totalReads"</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">filtSampPar =</span> <span class="fu">list</span>(<span class="at">totalReads =</span> <span class="dv">100</span>),</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">filtTax =</span> <span class="st">"numbSamp"</span>,</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>                    <span class="at">filtTaxPar =</span> <span class="fu">list</span>(<span class="at">numbSamp =</span> <span class="dv">5</span>),</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">measure =</span> <span class="st">"spieceasi"</span>,</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">measurePar =</span> <span class="fu">list</span>(<span class="at">method =</span> <span class="st">"mb"</span>,</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">pulsar.params =</span> <span class="fu">list</span>(<span class="at">rep.num =</span> <span class="dv">10</span>)),</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>                    <span class="at">zeroMethod =</span> <span class="st">"none"</span>,</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>                    <span class="at">normMethod =</span> <span class="st">"none"</span>,</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>                    <span class="at">sparsMethod =</span> <span class="st">"none"</span>, </span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>                    <span class="at">dissFunc =</span> <span class="st">"signed"</span>,</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>                    <span class="at">verbose =</span> <span class="dv">3</span>,</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>                    <span class="at">seed =</span> <span class="dv">12345</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">

</div>
<div class="cell">

</div>
<p><code>netConstruct()</code> returns an object of the class <code>microNet</code>, which contains all matrices generated during network construction.</p>
<p>The object also contains an edge list, giving each edge’s estimated association, dissimilarity, and adjacency. Let’s take a quick look at the edges with the highest and lowest edge weights:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>edgelist <span class="ot">&lt;-</span> net<span class="sc">$</span>edgelist1[<span class="fu">order</span>(net<span class="sc">$</span>edgelist1<span class="sc">$</span>adja, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), ]</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(edgelist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                            v1                    v2      asso      diss
210 64_Syntrophobacteraceae(F)            73_PK29(O) 0.3597096 0.5658138
190                 50_RB41(O)            69_RB41(O) 0.3587220 0.5662499
196                 54_RB41(O)            82_RB41(O) 0.3560518 0.5674276
178            43_Ellin6513(O)       55_Ellin6513(O) 0.3485638 0.5707172
112      24_Sinobacteraceae(F) 68_Koribacteraceae(F) 0.3238421 0.5814456
2        1_Chitinophagaceae(F)             9_RB41(O) 0.3000597 0.5915827
         adja
210 0.4341862
190 0.4337501
196 0.4325724
178 0.4292828
112 0.4185544
2   0.4084173</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(edgelist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                         v1                      v2         asso      diss
129   30_Sinobacteraceae(F)   35_Sinobacteraceae(F) -0.009823811 0.7105715
193            Kaistobacter   75_Koribacteraceae(F) -0.010324659 0.7107477
35                8_RB41(O) 33_Rhodospirillaceae(F) -0.010716865 0.7108857
113 25_Rhodospirillaceae(F)   34_Koribacteraceae(F) -0.011317239 0.7110968
99  22_Bradyrhizobiaceae(F)            Rubrobacter3 -0.013935976 0.7120168
9            3_Ellin6513(O)         Bradyrhizobium4 -0.030935464 0.7179608
         adja
129 0.2894285
193 0.2892523
35  0.2891143
113 0.2889032
99  0.2879832
9   0.2820392</code></pre>
</div>
</div>
</section>
<section id="network-analysis" class="level3">
<h3 class="anchored" data-anchor-id="network-analysis">Network analysis</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>netprops <span class="ot">&lt;-</span> <span class="fu">netAnalyze</span>(net, </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">clustMethod =</span> <span class="st">"cluster_fast_greedy"</span>,</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">hubPar =</span> <span class="st">"eigenvector"</span>,</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">normDeg =</span> <span class="cn">FALSE</span>,</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">gcmHeat =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(netprops)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Component sizes
```````````````           
size: 111 1
   #:   1 3
______________________________
Global network properties
`````````````````````````
Largest connected component (LCC):
                                 
Relative LCC size         0.97368
Clustering coefficient    0.21135
Modularity                0.61887
Positive edge percentage 95.41284
Edge density              0.03571
Natural connectivity      0.01144
Vertex connectivity       1.00000
Edge connectivity         1.00000
Average dissimilarity*    0.98851
Average path length**     2.89288

Whole network:
                                 
Number of components      4.00000
Clustering coefficient    0.21135
Modularity                0.61887
Positive edge percentage 95.41284
Edge density              0.03385
Natural connectivity      0.01108
-----
*: Dissimilarity = 1 - edge weight
**: Path length = Units with average dissimilarity

______________________________
Clusters
- In the whole network
- Algorithm: cluster_fast_greedy
```````````````````````````````` 
                           
name: 0  1  2  3  4  5  6 7
   #: 3 19 31 18 10 16 14 3

______________________________
Hubs
- In alphabetical/numerical order
- Based on empirical quantiles of centralities
```````````````````````````````````````````````                      
 12_Solibacterales(O) 
 34_Koribacteraceae(F)
 42_Ellin6513(O)      
 61_Koribacteraceae(F)
 72_Koribacteraceae(F)
 75_Koribacteraceae(F)

______________________________
Centrality measures
- In decreasing order
- Centrality of disconnected components is zero
````````````````````````````````````````````````
Degree (unnormalized):
                          
75_Koribacteraceae(F)   11
68_Koribacteraceae(F)   10
17_RB41(O)               9
30_Sinobacteraceae(F)    8
33_Rhodospirillaceae(F)  8
34_Koribacteraceae(F)    8
42_Ellin6513(O)          8
61_Koribacteraceae(F)    8
12_Solibacterales(O)     7
16_Ellin6513(O)          7

Betweenness centrality (normalized):
                               
33_Rhodospirillaceae(F) 0.18098
30_Sinobacteraceae(F)   0.15913
50_RB41(O)              0.13044
75_Koribacteraceae(F)   0.12811
12_Solibacterales(O)    0.11526
42_Ellin6513(O)         0.10726
17_RB41(O)              0.10492
68_Koribacteraceae(F)   0.10392
Rubrobacter3            0.09958
35_Sinobacteraceae(F)   0.09358

Closeness centrality (normalized):
                               
75_Koribacteraceae(F)   0.56896
12_Solibacterales(O)    0.56828
33_Rhodospirillaceae(F) 0.56453
30_Sinobacteraceae(F)   0.56001
42_Ellin6513(O)         0.54787
17_RB41(O)              0.52447
50_RB41(O)              0.51804
61_Koribacteraceae(F)   0.51734
34_Koribacteraceae(F)   0.50888
56_Solibacterales(O)    0.50290

Eigenvector centrality (normalized):
                               
75_Koribacteraceae(F)   1.00000
12_Solibacterales(O)    0.72317
72_Koribacteraceae(F)   0.70549
34_Koribacteraceae(F)   0.70273
42_Ellin6513(O)         0.70056
61_Koribacteraceae(F)   0.64929
56_Solibacterales(O)    0.57609
16_Ellin6513(O)         0.55649
33_Rhodospirillaceae(F) 0.50822
68_Koribacteraceae(F)   0.49436</code></pre>
</div>
</div>
</section>
<section id="highlight-node-properties" class="level3">
<h3 class="anchored" data-anchor-id="highlight-node-properties">Highlight node properties</h3>
<p>In the first plot, node colors represent the detected clusters and node sizes are scaled by eigenvector centrality. Hub nodes are highlighted by default. Singletons are not included in the plot. To improve the readability, NetCoMi’s “intelligent” label shortening approach is used.</p>
<p>Note that nodes are sometimes placed too close together so that the labels overlap. You may need to play around with the repulsion argument until you find a value where the labels are legible, but also the clusters are still well recognizable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(netprops,</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">repulsion =</span> <span class="fl">0.95</span>,</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">rmSingles =</span> <span class="cn">TRUE</span>,</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">shortenLabels =</span> <span class="st">"none"</span>,</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">labelScale =</span> <span class="cn">TRUE</span>,</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeSize =</span> <span class="st">"eigenvector"</span>,</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeSizeSpread =</span> <span class="dv">3</span>,</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeColor =</span> <span class="st">"cluster"</span>, </span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">hubBorderCol =</span> <span class="st">"gray40"</span>,</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexNodes =</span> <span class="fl">1.8</span>,</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexLabels =</span> <span class="dv">3</span>,</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>     <span class="at">edgeTranspHigh =</span> <span class="dv">20</span>,</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>     <span class="at">title1 =</span> <span class="st">"Network properties highlighted"</span>, </span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>     <span class="at">showTitle =</span> <span class="cn">TRUE</span>,</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexTitle =</span> <span class="fl">2.3</span>,</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>     <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">8</span>))</span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="fl">0.7</span>, <span class="fl">1.1</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">title =</span> <span class="st">"estimated correlation:"</span>,</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"+"</span>,<span class="st">"-"</span>), <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#009900"</span>,<span class="st">"red"</span>), </span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">horiz =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="8_practical_networks_files/figure-html/network_plot_single_cluster-1.png" class="img-fluid" width="1248"></p>
</div>
</div>
</section>
<section id="highlight-data-features" class="level3">
<h3 class="anchored" data-anchor-id="highlight-data-features">Highlight data features</h3>
<p>We now color nodes according to their phylum. The node sizes are proportional to a taxon’s sum of mclr-transformed abundances. As already mentioned in Section @ref(spring-network), this is the normalization method used by <code>SPRING</code>. A color palette from <code>RColorBrewer</code> is used here.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate vector with phylum names for node coloring</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>phyla <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(taxtab_renamed[, <span class="st">"Phylum"</span>])</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(phyla) <span class="ot">&lt;-</span> taxtab_renamed[, <span class="st">"Genus"</span>]</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder levels</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>phyla <span class="ot">&lt;-</span> <span class="fu">reorder</span>(phyla, phyla, <span class="at">FUN =</span> length, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Create color vector</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>colvec <span class="ot">&lt;-</span> <span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">levels</span>(phyla)))</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(netprops,</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>     <span class="at">repulsion =</span> <span class="fl">0.95</span>,</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>     <span class="at">rmSingles =</span> <span class="cn">TRUE</span>,</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>     <span class="at">shortenLabels =</span> <span class="st">"none"</span>,</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>     <span class="at">labelScale =</span> <span class="cn">TRUE</span>,</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeSize =</span> <span class="st">"clr"</span>,</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeColor =</span> <span class="st">"feature"</span>, </span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>     <span class="at">featVecCol =</span> phyla, </span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>     <span class="at">colorVec =</span>  colvec,</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>     <span class="co">#nodeTransp = 20,</span></span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>     <span class="at">highlightHubs =</span> <span class="cn">FALSE</span>,</span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexNodes =</span> <span class="fl">1.2</span>,</span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>          <span class="at">cexLabels =</span> <span class="dv">3</span>,</span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>     <span class="at">edgeTranspHigh =</span> <span class="dv">20</span>,</span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a>     <span class="at">title1 =</span> <span class="st">"Data features highlighted"</span>, </span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a>     <span class="at">showTitle =</span> <span class="cn">TRUE</span>,</span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexTitle =</span> <span class="fl">2.3</span>,</span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a>     <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">4</span>, <span class="dv">6</span>))</span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Add legends</span></span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="fl">0.7</span>, <span class="fl">1.1</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">title =</span> <span class="st">"estimated correlation:"</span>,</span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"+"</span>,<span class="st">"-"</span>), <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#009900"</span>,<span class="st">"red"</span>), </span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">horiz =</span> <span class="cn">TRUE</span>)</span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-35"><a href="#cb28-35" aria-hidden="true" tabindex="-1"></a><span class="co"># Colors used in the legend should be equally transparent as in the plot</span></span>
<span id="cb28-36"><a href="#cb28-36" aria-hidden="true" tabindex="-1"></a>col_transp <span class="ot">&lt;-</span> <span class="fu">colToTransp</span>(colvec, <span class="dv">60</span>)</span>
<span id="cb28-37"><a href="#cb28-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-38"><a href="#cb28-38" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="sc">-</span><span class="fl">1.8</span>, <span class="fl">1.1</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">pt.cex =</span> <span class="fl">2.5</span>, <span class="at">title =</span> <span class="st">"Phylum:"</span>, </span>
<span id="cb28-39"><a href="#cb28-39" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend=</span><span class="fu">levels</span>(phyla), <span class="at">col =</span> col_transp, <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">pch =</span> <span class="dv">16</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="8_practical_networks_files/figure-html/network_plot_single_phylum-1.png" class="img-fluid" width="1440"></p>
</div>
</div>
<p>Select Proteobacteria</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> <span class="fu">names</span>(phyla[phyla <span class="sc">==</span> <span class="st">"Proteobacteria"</span>])</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> sel[sel <span class="sc">%in%</span> <span class="fu">colnames</span>(net<span class="sc">$</span>adjaMat1)]</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(netprops,</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">nodeFilter =</span> <span class="st">"names"</span>,</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeFilterPar =</span> sel,</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">repulsion =</span> <span class="fl">0.95</span>,</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">rmSingles =</span> <span class="cn">TRUE</span>,</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">shortenLabels =</span> <span class="st">"none"</span>,</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>     <span class="at">labelScale =</span> <span class="cn">TRUE</span>,</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeSize =</span> <span class="st">"eigenvector"</span>,</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeSizeSpread =</span> <span class="dv">3</span>,</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeColor =</span> <span class="st">"cluster"</span>, </span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>     <span class="at">hubBorderCol =</span> <span class="st">"gray40"</span>,</span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexNodes =</span> <span class="fl">1.8</span>,</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexLabels =</span> <span class="dv">3</span>,</span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>     <span class="at">edgeTranspHigh =</span> <span class="dv">20</span>,</span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>     <span class="at">title1 =</span> <span class="st">"Network properties highlighted"</span>, </span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>     <span class="at">showTitle =</span> <span class="cn">TRUE</span>,</span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexTitle =</span> <span class="fl">2.3</span>,</span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a>     <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">8</span>))</span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="fl">0.7</span>, <span class="fl">1.1</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">title =</span> <span class="st">"estimated correlation:"</span>,</span>
<span id="cb29-24"><a href="#cb29-24" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"+"</span>,<span class="st">"-"</span>), <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#009900"</span>,<span class="st">"red"</span>), </span>
<span id="cb29-25"><a href="#cb29-25" aria-hidden="true" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">horiz =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="8_practical_networks_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>


<!-- -->


</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-lauber2009pyrosequencing" class="csl-entry" role="doc-biblioentry">
Lauber, Christian L, Micah Hamady, Rob Knight, and Noah Fierer. 2009. <span>“Pyrosequencing-Based Assessment of Soil pH as a Predictor of Soil Bacterial Community Structure at the Continental Scale.”</span> <em>Applied and Environmental Microbiology</em> 75 (15): 5111–20.
</div>
<div id="ref-peschel2021netcomi" class="csl-entry" role="doc-biblioentry">
Peschel, Stefanie, Christian L Müller, Erika von Mutius, Anne-Laure Boulesteix, and Martin Depner. 2021. <span>“<span class="nocase">NetCoMi: network construction and comparison for microbiome data in R</span>.”</span> <em><span>Briefings in Bioinformatics</span></em> 22 (4): bbaa290. <a href="https://doi.org/10.1093/bib/bbaa290">https://doi.org/10.1093/bib/bbaa290</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb30" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Practical applications of learned concepts in R"</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span><span class="co"> </span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="co">    code-fold: false</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="an">bibliography:</span><span class="co"> references.bib</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span><span class="co"> visual</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a><span class="fu">## Network learning and analysis</span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>Here, we learn how a microbial association network is constructed and analyzed using the R package NetCoMi.</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip collapse="false"}</span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a><span class="fu">### The R package NetCoMi</span></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>The <span class="co">[</span><span class="ot">NetCoMi</span><span class="co">](https://github.com/stefpeschel/NetCoMi)</span> <span class="co">[</span><span class="ot">@peschel2021netcomi</span><span class="co">]</span> package is specifically designed to **construct, analyze, and compare networks** for microbiome data. It implements the complete workflow of learning and analyzing microbial networks explained in lecture 8. Instead of using several functions for each of the steps, `NetCoMi` provides a single function for network construction (`netConstruct()`), so the package streamlines the workflow considerably. The user can choose from a variety of methods for **data preprocessing, association estimation, sparsification, and transformation**. The returned <span class="in">`microNet`</span> object can then be passed to <span class="in">`netAnalyze()`</span> (the network analysis function) so that all necessary information is available for the network analysis workflow.</span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a><span class="fu">### Install and load packages</span></span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a>We start by installing NetCoMi from GitHub. Two of NetCoMi's dependencies are also only available on GitHub. We will install them manually beforehand to avoid problems.</span>
<span id="cb30-24"><a href="#cb30-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-27"><a href="#cb30-27" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-28"><a href="#cb30-28" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: install-netcomi</span></span>
<span id="cb30-29"><a href="#cb30-29" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb30-30"><a href="#cb30-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-31"><a href="#cb30-31" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"zdk123/SpiecEasi"</span>)</span>
<span id="cb30-32"><a href="#cb30-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-33"><a href="#cb30-33" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"GraceYoon/SPRING"</span>)</span>
<span id="cb30-34"><a href="#cb30-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-35"><a href="#cb30-35" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"stefpeschel/NetCoMi"</span>, <span class="at">force =</span> <span class="cn">TRUE</span>, <span class="at">ref =</span> <span class="st">"TSE"</span>,</span>
<span id="cb30-36"><a href="#cb30-36" aria-hidden="true" tabindex="-1"></a>                         <span class="at">dependencies =</span> <span class="fu">c</span>(<span class="st">"Depends"</span>, <span class="st">"Imports"</span>, <span class="st">"LinkingTo"</span>),</span>
<span id="cb30-37"><a href="#cb30-37" aria-hidden="true" tabindex="-1"></a>                         <span class="at">repos =</span> <span class="fu">c</span>(<span class="st">"https://cloud.r-project.org/"</span>,</span>
<span id="cb30-38"><a href="#cb30-38" aria-hidden="true" tabindex="-1"></a>                                   BiocManager<span class="sc">::</span><span class="fu">repositories</span>()))</span>
<span id="cb30-39"><a href="#cb30-39" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-40"><a href="#cb30-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-41"><a href="#cb30-41" aria-hidden="true" tabindex="-1"></a>Load packages needed in this exercise.</span>
<span id="cb30-42"><a href="#cb30-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-45"><a href="#cb30-45" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-46"><a href="#cb30-46" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(NetCoMi)</span>
<span id="cb30-47"><a href="#cb30-47" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span>
<span id="cb30-48"><a href="#cb30-48" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-49"><a href="#cb30-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-50"><a href="#cb30-50" aria-hidden="true" tabindex="-1"></a><span class="fu">### Load the data</span></span>
<span id="cb30-51"><a href="#cb30-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-52"><a href="#cb30-52" aria-hidden="true" tabindex="-1"></a>The 88 soil dataset <span class="co">[</span><span class="ot">@lauber2009pyrosequencing</span><span class="co">]</span> is again used in this example.</span>
<span id="cb30-53"><a href="#cb30-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-56"><a href="#cb30-56" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-57"><a href="#cb30-57" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-data</span></span>
<span id="cb30-58"><a href="#cb30-58" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb30-59"><a href="#cb30-59" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb30-60"><a href="#cb30-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-61"><a href="#cb30-61" aria-hidden="true" tabindex="-1"></a>soil <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/soil_processed/soil.RDS"</span>)</span>
<span id="cb30-62"><a href="#cb30-62" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-63"><a href="#cb30-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-64"><a href="#cb30-64" aria-hidden="true" tabindex="-1"></a>Take a look at the data</span>
<span id="cb30-65"><a href="#cb30-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-68"><a href="#cb30-68" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-69"><a href="#cb30-69" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: look-at-data</span></span>
<span id="cb30-70"><a href="#cb30-70" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb30-71"><a href="#cb30-71" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb30-72"><a href="#cb30-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-73"><a href="#cb30-73" aria-hidden="true" tabindex="-1"></a>soil</span>
<span id="cb30-74"><a href="#cb30-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-75"><a href="#cb30-75" aria-hidden="true" tabindex="-1"></a><span class="co"># Read count table</span></span>
<span id="cb30-76"><a href="#cb30-76" aria-hidden="true" tabindex="-1"></a><span class="fu">otu_table</span>(soil)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span>
<span id="cb30-77"><a href="#cb30-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-78"><a href="#cb30-78" aria-hidden="true" tabindex="-1"></a><span class="co"># Taxonomy table</span></span>
<span id="cb30-79"><a href="#cb30-79" aria-hidden="true" tabindex="-1"></a>taxtab <span class="ot">&lt;-</span> <span class="fu">tax_table</span>(soil)</span>
<span id="cb30-80"><a href="#cb30-80" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(taxtab)</span>
<span id="cb30-81"><a href="#cb30-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-82"><a href="#cb30-82" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample data</span></span>
<span id="cb30-83"><a href="#cb30-83" aria-hidden="true" tabindex="-1"></a>sampdata <span class="ot">&lt;-</span> <span class="fu">sample_data</span>(soil)</span>
<span id="cb30-84"><a href="#cb30-84" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sampdata)</span>
<span id="cb30-85"><a href="#cb30-85" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-86"><a href="#cb30-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-87"><a href="#cb30-87" aria-hidden="true" tabindex="-1"></a><span class="fu">### Rename genera</span></span>
<span id="cb30-88"><a href="#cb30-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-89"><a href="#cb30-89" aria-hidden="true" tabindex="-1"></a>If we want to generate a network on genus level, we run into two problems:</span>
<span id="cb30-90"><a href="#cb30-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-91"><a href="#cb30-91" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>There are unidentified genera.</span>
<span id="cb30-92"><a href="#cb30-92" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Not all genera are unique.</span>
<span id="cb30-93"><a href="#cb30-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-96"><a href="#cb30-96" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-97"><a href="#cb30-97" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(taxtab[, <span class="st">"Genus"</span>])</span>
<span id="cb30-98"><a href="#cb30-98" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-99"><a href="#cb30-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-100"><a href="#cb30-100" aria-hidden="true" tabindex="-1"></a>Solutions:</span>
<span id="cb30-101"><a href="#cb30-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-102"><a href="#cb30-102" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Unknown genera are named by higher taxonomic ranks.</span>
<span id="cb30-103"><a href="#cb30-103" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Duplicates are numbered to make them unique.</span>
<span id="cb30-104"><a href="#cb30-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-105"><a href="#cb30-105" aria-hidden="true" tabindex="-1"></a>NetCoMi provides a function <span class="in">`renameTaxa()`</span>, which does this job for us. It can also be used to rename the taxa according to a desired pattern.</span>
<span id="cb30-106"><a href="#cb30-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-109"><a href="#cb30-109" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-110"><a href="#cb30-110" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename taxonomic table and make Rank6 (genus) unique</span></span>
<span id="cb30-111"><a href="#cb30-111" aria-hidden="true" tabindex="-1"></a>soil_renamed <span class="ot">&lt;-</span> <span class="fu">renameTaxa</span>(soil, </span>
<span id="cb30-112"><a href="#cb30-112" aria-hidden="true" tabindex="-1"></a>                           <span class="at">pat =</span> <span class="st">"&lt;name&gt;"</span>, </span>
<span id="cb30-113"><a href="#cb30-113" aria-hidden="true" tabindex="-1"></a>                           <span class="at">substPat =</span> <span class="st">"&lt;name&gt;_&lt;subst_name&gt;(&lt;subst_R&gt;)"</span>,</span>
<span id="cb30-114"><a href="#cb30-114" aria-hidden="true" tabindex="-1"></a>                           <span class="at">numDupli =</span> <span class="st">"Genus"</span>)</span>
<span id="cb30-115"><a href="#cb30-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-116"><a href="#cb30-116" aria-hidden="true" tabindex="-1"></a>taxtab_renamed <span class="ot">&lt;-</span> <span class="fu">tax_table</span>(soil_renamed)</span>
<span id="cb30-117"><a href="#cb30-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-118"><a href="#cb30-118" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">cbind</span>(taxtab[, <span class="st">"Genus"</span>], taxtab_renamed[, <span class="st">"Genus"</span>]), <span class="dv">15</span>)</span>
<span id="cb30-119"><a href="#cb30-119" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-120"><a href="#cb30-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-121"><a href="#cb30-121" aria-hidden="true" tabindex="-1"></a>Now, we can pass the phyloseq object with renamed genera to NetCoMi's <span class="in">`netConstruct()`</span> function.</span>
<span id="cb30-122"><a href="#cb30-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-123"><a href="#cb30-123" aria-hidden="true" tabindex="-1"></a>SpiecEasi is used in the "MB" mode (Meinshausen &amp; Bühlman neighborhood selection) to measure associations between the genera.</span>
<span id="cb30-124"><a href="#cb30-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-125"><a href="#cb30-125" aria-hidden="true" tabindex="-1"></a>A few comments on the arguments:</span>
<span id="cb30-126"><a href="#cb30-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-127"><a href="#cb30-127" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The network is created on genus level.</span>
<span id="cb30-128"><a href="#cb30-128" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The data are filtered as follows:</span>
<span id="cb30-129"><a href="#cb30-129" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Only samples with a total of at least 100 reads are included (argument <span class="in">`filtSamp`</span>).</span>
<span id="cb30-130"><a href="#cb30-130" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Only taxa occuring in at least 5 samples are included (argument <span class="in">`filtTax`</span>).</span>
<span id="cb30-131"><a href="#cb30-131" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`measure`</span> defines the association or dissimilarity measure, in our case SpiecEasi. Additional arguments are passed to <span class="in">`spiec.easi()`</span> via <span class="in">`measurePar`</span>. We reduce the number of StARS repetitions to 10 to reduce runtime.</span>
<span id="cb30-132"><a href="#cb30-132" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Normalization and zero handling is done internally by <span class="in">`spiec.easi()`</span>, so we set <span class="in">`normMethod`</span> and <span class="in">`zeroMethod`</span> to "none".</span>
<span id="cb30-133"><a href="#cb30-133" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`sparsMethod`</span> is also set to "none" (SpiecEasi includes StARS model selection to obtain a sparse network).</span>
<span id="cb30-134"><a href="#cb30-134" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The "signed" method is used to transform associations into dissimilarities (argument <span class="in">`dissFunc`</span>). Thus, strongly negatively associated taxa have a high dissimilarity and conversely a low similarity, corresponding to low edge weights on the network plot.</span>
<span id="cb30-135"><a href="#cb30-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-138"><a href="#cb30-138" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-139"><a href="#cb30-139" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb30-140"><a href="#cb30-140" aria-hidden="true" tabindex="-1"></a>net <span class="ot">&lt;-</span> <span class="fu">netConstruct</span>(soil_renamed,</span>
<span id="cb30-141"><a href="#cb30-141" aria-hidden="true" tabindex="-1"></a>                    <span class="at">taxRank =</span> <span class="st">"Genus"</span>,</span>
<span id="cb30-142"><a href="#cb30-142" aria-hidden="true" tabindex="-1"></a>                    <span class="at">filtSamp =</span> <span class="st">"totalReads"</span>,</span>
<span id="cb30-143"><a href="#cb30-143" aria-hidden="true" tabindex="-1"></a>                    <span class="at">filtSampPar =</span> <span class="fu">list</span>(<span class="at">totalReads =</span> <span class="dv">100</span>),</span>
<span id="cb30-144"><a href="#cb30-144" aria-hidden="true" tabindex="-1"></a>                    <span class="at">filtTax =</span> <span class="st">"numbSamp"</span>,</span>
<span id="cb30-145"><a href="#cb30-145" aria-hidden="true" tabindex="-1"></a>                    <span class="at">filtTaxPar =</span> <span class="fu">list</span>(<span class="at">numbSamp =</span> <span class="dv">5</span>),</span>
<span id="cb30-146"><a href="#cb30-146" aria-hidden="true" tabindex="-1"></a>                    <span class="at">measure =</span> <span class="st">"spieceasi"</span>,</span>
<span id="cb30-147"><a href="#cb30-147" aria-hidden="true" tabindex="-1"></a>                    <span class="at">measurePar =</span> <span class="fu">list</span>(<span class="at">method =</span> <span class="st">"mb"</span>,</span>
<span id="cb30-148"><a href="#cb30-148" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">pulsar.params =</span> <span class="fu">list</span>(<span class="at">rep.num =</span> <span class="dv">10</span>)),</span>
<span id="cb30-149"><a href="#cb30-149" aria-hidden="true" tabindex="-1"></a>                    <span class="at">zeroMethod =</span> <span class="st">"none"</span>,</span>
<span id="cb30-150"><a href="#cb30-150" aria-hidden="true" tabindex="-1"></a>                    <span class="at">normMethod =</span> <span class="st">"none"</span>,</span>
<span id="cb30-151"><a href="#cb30-151" aria-hidden="true" tabindex="-1"></a>                    <span class="at">sparsMethod =</span> <span class="st">"none"</span>, </span>
<span id="cb30-152"><a href="#cb30-152" aria-hidden="true" tabindex="-1"></a>                    <span class="at">dissFunc =</span> <span class="st">"signed"</span>,</span>
<span id="cb30-153"><a href="#cb30-153" aria-hidden="true" tabindex="-1"></a>                    <span class="at">verbose =</span> <span class="dv">3</span>,</span>
<span id="cb30-154"><a href="#cb30-154" aria-hidden="true" tabindex="-1"></a>                    <span class="at">seed =</span> <span class="dv">12345</span>)</span>
<span id="cb30-155"><a href="#cb30-155" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-156"><a href="#cb30-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-159"><a href="#cb30-159" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-160"><a href="#cb30-160" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb30-161"><a href="#cb30-161" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb30-162"><a href="#cb30-162" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(net, <span class="at">file =</span> <span class="st">"data/networks/spieceasi_net.RDS"</span>)</span>
<span id="cb30-163"><a href="#cb30-163" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-164"><a href="#cb30-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-167"><a href="#cb30-167" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-168"><a href="#cb30-168" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb30-169"><a href="#cb30-169" aria-hidden="true" tabindex="-1"></a>net <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/networks/spieceasi_net.RDS"</span>)</span>
<span id="cb30-170"><a href="#cb30-170" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-171"><a href="#cb30-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-172"><a href="#cb30-172" aria-hidden="true" tabindex="-1"></a><span class="in">`netConstruct()`</span> returns an object of the class <span class="in">`microNet`</span>, which contains all matrices generated during network construction.</span>
<span id="cb30-173"><a href="#cb30-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-174"><a href="#cb30-174" aria-hidden="true" tabindex="-1"></a>The object also contains an edge list, giving each edge's estimated association, dissimilarity, and adjacency. Let's take a quick look at the edges with the highest and lowest edge weights:</span>
<span id="cb30-175"><a href="#cb30-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-176"><a href="#cb30-176" aria-hidden="true" tabindex="-1"></a><span class="in">```{r edge_list}</span></span>
<span id="cb30-177"><a href="#cb30-177" aria-hidden="true" tabindex="-1"></a>edgelist <span class="ot">&lt;-</span> net<span class="sc">$</span>edgelist1[<span class="fu">order</span>(net<span class="sc">$</span>edgelist1<span class="sc">$</span>adja, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), ]</span>
<span id="cb30-178"><a href="#cb30-178" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(edgelist)</span>
<span id="cb30-179"><a href="#cb30-179" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(edgelist)</span>
<span id="cb30-180"><a href="#cb30-180" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-181"><a href="#cb30-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-182"><a href="#cb30-182" aria-hidden="true" tabindex="-1"></a><span class="fu">### Network analysis</span></span>
<span id="cb30-183"><a href="#cb30-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-186"><a href="#cb30-186" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-187"><a href="#cb30-187" aria-hidden="true" tabindex="-1"></a>netprops <span class="ot">&lt;-</span> <span class="fu">netAnalyze</span>(net, </span>
<span id="cb30-188"><a href="#cb30-188" aria-hidden="true" tabindex="-1"></a>                       <span class="at">clustMethod =</span> <span class="st">"cluster_fast_greedy"</span>,</span>
<span id="cb30-189"><a href="#cb30-189" aria-hidden="true" tabindex="-1"></a>                       <span class="at">hubPar =</span> <span class="st">"eigenvector"</span>,</span>
<span id="cb30-190"><a href="#cb30-190" aria-hidden="true" tabindex="-1"></a>                       <span class="at">normDeg =</span> <span class="cn">FALSE</span>,</span>
<span id="cb30-191"><a href="#cb30-191" aria-hidden="true" tabindex="-1"></a>                       <span class="at">gcmHeat =</span> <span class="cn">FALSE</span>)</span>
<span id="cb30-192"><a href="#cb30-192" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-193"><a href="#cb30-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-196"><a href="#cb30-196" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-197"><a href="#cb30-197" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(netprops)</span>
<span id="cb30-198"><a href="#cb30-198" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-199"><a href="#cb30-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-200"><a href="#cb30-200" aria-hidden="true" tabindex="-1"></a><span class="fu">### Highlight node properties</span></span>
<span id="cb30-201"><a href="#cb30-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-202"><a href="#cb30-202" aria-hidden="true" tabindex="-1"></a>In the first plot, node colors represent the detected clusters and node sizes are scaled by eigenvector centrality. Hub nodes are highlighted by default. Singletons are not included in the plot. To improve the readability, NetCoMi's "intelligent" label shortening approach is used.</span>
<span id="cb30-203"><a href="#cb30-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-204"><a href="#cb30-204" aria-hidden="true" tabindex="-1"></a>Note that nodes are sometimes placed too close together so that the labels overlap. You may need to play around with the repulsion argument until you find a value where the labels are legible, but also the clusters are still well recognizable.</span>
<span id="cb30-205"><a href="#cb30-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-206"><a href="#cb30-206" aria-hidden="true" tabindex="-1"></a><span class="in">```{r network_plot_single_cluster, fig.width=13, fig.height=10}</span></span>
<span id="cb30-207"><a href="#cb30-207" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(netprops,</span>
<span id="cb30-208"><a href="#cb30-208" aria-hidden="true" tabindex="-1"></a>     <span class="at">repulsion =</span> <span class="fl">0.95</span>,</span>
<span id="cb30-209"><a href="#cb30-209" aria-hidden="true" tabindex="-1"></a>     <span class="at">rmSingles =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-210"><a href="#cb30-210" aria-hidden="true" tabindex="-1"></a>     <span class="at">shortenLabels =</span> <span class="st">"none"</span>,</span>
<span id="cb30-211"><a href="#cb30-211" aria-hidden="true" tabindex="-1"></a>     <span class="at">labelScale =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-212"><a href="#cb30-212" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeSize =</span> <span class="st">"eigenvector"</span>,</span>
<span id="cb30-213"><a href="#cb30-213" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeSizeSpread =</span> <span class="dv">3</span>,</span>
<span id="cb30-214"><a href="#cb30-214" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeColor =</span> <span class="st">"cluster"</span>, </span>
<span id="cb30-215"><a href="#cb30-215" aria-hidden="true" tabindex="-1"></a>     <span class="at">hubBorderCol =</span> <span class="st">"gray40"</span>,</span>
<span id="cb30-216"><a href="#cb30-216" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexNodes =</span> <span class="fl">1.8</span>,</span>
<span id="cb30-217"><a href="#cb30-217" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexLabels =</span> <span class="dv">3</span>,</span>
<span id="cb30-218"><a href="#cb30-218" aria-hidden="true" tabindex="-1"></a>     <span class="at">edgeTranspHigh =</span> <span class="dv">20</span>,</span>
<span id="cb30-219"><a href="#cb30-219" aria-hidden="true" tabindex="-1"></a>     <span class="at">title1 =</span> <span class="st">"Network properties highlighted"</span>, </span>
<span id="cb30-220"><a href="#cb30-220" aria-hidden="true" tabindex="-1"></a>     <span class="at">showTitle =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-221"><a href="#cb30-221" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexTitle =</span> <span class="fl">2.3</span>,</span>
<span id="cb30-222"><a href="#cb30-222" aria-hidden="true" tabindex="-1"></a>     <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">8</span>))</span>
<span id="cb30-223"><a href="#cb30-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-224"><a href="#cb30-224" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="fl">0.7</span>, <span class="fl">1.1</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">title =</span> <span class="st">"estimated correlation:"</span>,</span>
<span id="cb30-225"><a href="#cb30-225" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"+"</span>,<span class="st">"-"</span>), <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#009900"</span>,<span class="st">"red"</span>), </span>
<span id="cb30-226"><a href="#cb30-226" aria-hidden="true" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">horiz =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-227"><a href="#cb30-227" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-228"><a href="#cb30-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-229"><a href="#cb30-229" aria-hidden="true" tabindex="-1"></a><span class="fu">### Highlight data features</span></span>
<span id="cb30-230"><a href="#cb30-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-231"><a href="#cb30-231" aria-hidden="true" tabindex="-1"></a>We now color nodes according to their phylum. The node sizes are proportional to a taxon's sum of mclr-transformed abundances. As already mentioned in Section \@ref(spring-network), this is the normalization method used by <span class="in">`SPRING`</span>. A color palette from <span class="in">`RColorBrewer`</span> is used here.</span>
<span id="cb30-232"><a href="#cb30-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-233"><a href="#cb30-233" aria-hidden="true" tabindex="-1"></a><span class="in">```{r network_plot_single_phylum, fig.width=15, fig.height=10}</span></span>
<span id="cb30-234"><a href="#cb30-234" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate vector with phylum names for node coloring</span></span>
<span id="cb30-235"><a href="#cb30-235" aria-hidden="true" tabindex="-1"></a>phyla <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(taxtab_renamed[, <span class="st">"Phylum"</span>])</span>
<span id="cb30-236"><a href="#cb30-236" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(phyla) <span class="ot">&lt;-</span> taxtab_renamed[, <span class="st">"Genus"</span>]</span>
<span id="cb30-237"><a href="#cb30-237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-238"><a href="#cb30-238" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder levels</span></span>
<span id="cb30-239"><a href="#cb30-239" aria-hidden="true" tabindex="-1"></a>phyla <span class="ot">&lt;-</span> <span class="fu">reorder</span>(phyla, phyla, <span class="at">FUN =</span> length, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-240"><a href="#cb30-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-241"><a href="#cb30-241" aria-hidden="true" tabindex="-1"></a><span class="co"># Create color vector</span></span>
<span id="cb30-242"><a href="#cb30-242" aria-hidden="true" tabindex="-1"></a>colvec <span class="ot">&lt;-</span> <span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">levels</span>(phyla)))</span>
<span id="cb30-243"><a href="#cb30-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-244"><a href="#cb30-244" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(netprops,</span>
<span id="cb30-245"><a href="#cb30-245" aria-hidden="true" tabindex="-1"></a>     <span class="at">repulsion =</span> <span class="fl">0.95</span>,</span>
<span id="cb30-246"><a href="#cb30-246" aria-hidden="true" tabindex="-1"></a>     <span class="at">rmSingles =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-247"><a href="#cb30-247" aria-hidden="true" tabindex="-1"></a>     <span class="at">shortenLabels =</span> <span class="st">"none"</span>,</span>
<span id="cb30-248"><a href="#cb30-248" aria-hidden="true" tabindex="-1"></a>     <span class="at">labelScale =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-249"><a href="#cb30-249" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeSize =</span> <span class="st">"clr"</span>,</span>
<span id="cb30-250"><a href="#cb30-250" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeColor =</span> <span class="st">"feature"</span>, </span>
<span id="cb30-251"><a href="#cb30-251" aria-hidden="true" tabindex="-1"></a>     <span class="at">featVecCol =</span> phyla, </span>
<span id="cb30-252"><a href="#cb30-252" aria-hidden="true" tabindex="-1"></a>     <span class="at">colorVec =</span>  colvec,</span>
<span id="cb30-253"><a href="#cb30-253" aria-hidden="true" tabindex="-1"></a>     <span class="co">#nodeTransp = 20,</span></span>
<span id="cb30-254"><a href="#cb30-254" aria-hidden="true" tabindex="-1"></a>     <span class="at">highlightHubs =</span> <span class="cn">FALSE</span>,</span>
<span id="cb30-255"><a href="#cb30-255" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexNodes =</span> <span class="fl">1.2</span>,</span>
<span id="cb30-256"><a href="#cb30-256" aria-hidden="true" tabindex="-1"></a>          <span class="at">cexLabels =</span> <span class="dv">3</span>,</span>
<span id="cb30-257"><a href="#cb30-257" aria-hidden="true" tabindex="-1"></a>     <span class="at">edgeTranspHigh =</span> <span class="dv">20</span>,</span>
<span id="cb30-258"><a href="#cb30-258" aria-hidden="true" tabindex="-1"></a>     <span class="at">title1 =</span> <span class="st">"Data features highlighted"</span>, </span>
<span id="cb30-259"><a href="#cb30-259" aria-hidden="true" tabindex="-1"></a>     <span class="at">showTitle =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-260"><a href="#cb30-260" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexTitle =</span> <span class="fl">2.3</span>,</span>
<span id="cb30-261"><a href="#cb30-261" aria-hidden="true" tabindex="-1"></a>     <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">4</span>, <span class="dv">6</span>))</span>
<span id="cb30-262"><a href="#cb30-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-263"><a href="#cb30-263" aria-hidden="true" tabindex="-1"></a><span class="co"># Add legends</span></span>
<span id="cb30-264"><a href="#cb30-264" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="fl">0.7</span>, <span class="fl">1.1</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">title =</span> <span class="st">"estimated correlation:"</span>,</span>
<span id="cb30-265"><a href="#cb30-265" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"+"</span>,<span class="st">"-"</span>), <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#009900"</span>,<span class="st">"red"</span>), </span>
<span id="cb30-266"><a href="#cb30-266" aria-hidden="true" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">horiz =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-267"><a href="#cb30-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-268"><a href="#cb30-268" aria-hidden="true" tabindex="-1"></a><span class="co"># Colors used in the legend should be equally transparent as in the plot</span></span>
<span id="cb30-269"><a href="#cb30-269" aria-hidden="true" tabindex="-1"></a>col_transp <span class="ot">&lt;-</span> <span class="fu">colToTransp</span>(colvec, <span class="dv">60</span>)</span>
<span id="cb30-270"><a href="#cb30-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-271"><a href="#cb30-271" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="sc">-</span><span class="fl">1.8</span>, <span class="fl">1.1</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">pt.cex =</span> <span class="fl">2.5</span>, <span class="at">title =</span> <span class="st">"Phylum:"</span>, </span>
<span id="cb30-272"><a href="#cb30-272" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend=</span><span class="fu">levels</span>(phyla), <span class="at">col =</span> col_transp, <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">pch =</span> <span class="dv">16</span>) </span>
<span id="cb30-273"><a href="#cb30-273" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb30-274"><a href="#cb30-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-275"><a href="#cb30-275" aria-hidden="true" tabindex="-1"></a>Select Proteobacteria</span>
<span id="cb30-276"><a href="#cb30-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-279"><a href="#cb30-279" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb30-280"><a href="#cb30-280" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> <span class="fu">names</span>(phyla[phyla <span class="sc">==</span> <span class="st">"Proteobacteria"</span>])</span>
<span id="cb30-281"><a href="#cb30-281" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> sel[sel <span class="sc">%in%</span> <span class="fu">colnames</span>(net<span class="sc">$</span>adjaMat1)]</span>
<span id="cb30-282"><a href="#cb30-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-283"><a href="#cb30-283" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(netprops,</span>
<span id="cb30-284"><a href="#cb30-284" aria-hidden="true" tabindex="-1"></a>          <span class="at">nodeFilter =</span> <span class="st">"names"</span>,</span>
<span id="cb30-285"><a href="#cb30-285" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeFilterPar =</span> sel,</span>
<span id="cb30-286"><a href="#cb30-286" aria-hidden="true" tabindex="-1"></a>     <span class="at">repulsion =</span> <span class="fl">0.95</span>,</span>
<span id="cb30-287"><a href="#cb30-287" aria-hidden="true" tabindex="-1"></a>     <span class="at">rmSingles =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-288"><a href="#cb30-288" aria-hidden="true" tabindex="-1"></a>     <span class="at">shortenLabels =</span> <span class="st">"none"</span>,</span>
<span id="cb30-289"><a href="#cb30-289" aria-hidden="true" tabindex="-1"></a>     <span class="at">labelScale =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-290"><a href="#cb30-290" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeSize =</span> <span class="st">"eigenvector"</span>,</span>
<span id="cb30-291"><a href="#cb30-291" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeSizeSpread =</span> <span class="dv">3</span>,</span>
<span id="cb30-292"><a href="#cb30-292" aria-hidden="true" tabindex="-1"></a>     <span class="at">nodeColor =</span> <span class="st">"cluster"</span>, </span>
<span id="cb30-293"><a href="#cb30-293" aria-hidden="true" tabindex="-1"></a>     <span class="at">hubBorderCol =</span> <span class="st">"gray40"</span>,</span>
<span id="cb30-294"><a href="#cb30-294" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexNodes =</span> <span class="fl">1.8</span>,</span>
<span id="cb30-295"><a href="#cb30-295" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexLabels =</span> <span class="dv">3</span>,</span>
<span id="cb30-296"><a href="#cb30-296" aria-hidden="true" tabindex="-1"></a>     <span class="at">edgeTranspHigh =</span> <span class="dv">20</span>,</span>
<span id="cb30-297"><a href="#cb30-297" aria-hidden="true" tabindex="-1"></a>     <span class="at">title1 =</span> <span class="st">"Network properties highlighted"</span>, </span>
<span id="cb30-298"><a href="#cb30-298" aria-hidden="true" tabindex="-1"></a>     <span class="at">showTitle =</span> <span class="cn">TRUE</span>,</span>
<span id="cb30-299"><a href="#cb30-299" aria-hidden="true" tabindex="-1"></a>     <span class="at">cexTitle =</span> <span class="fl">2.3</span>,</span>
<span id="cb30-300"><a href="#cb30-300" aria-hidden="true" tabindex="-1"></a>     <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">8</span>))</span>
<span id="cb30-301"><a href="#cb30-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-302"><a href="#cb30-302" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="fl">0.7</span>, <span class="fl">1.1</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">title =</span> <span class="st">"estimated correlation:"</span>,</span>
<span id="cb30-303"><a href="#cb30-303" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"+"</span>,<span class="st">"-"</span>), <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#009900"</span>,<span class="st">"red"</span>), </span>
<span id="cb30-304"><a href="#cb30-304" aria-hidden="true" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">horiz =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-305"><a href="#cb30-305" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>