<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Mini-course CoDA - Sparse log-contrast and trac on 88 soil</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Mini-course CoDA</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Sparse log-contrast and trac on 88 soil</li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
 <span class="menu-text">4_eda.qmd</span>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./9_practical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Practical applications of learned concepts in R</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#run-sparse-log-contrast-and-trac-on-88-soil" id="toc-run-sparse-log-contrast-and-trac-on-88-soil" class="nav-link active" data-scroll-target="#run-sparse-log-contrast-and-trac-on-88-soil">Run sparse log-contrast and trac on 88 soil</a>
  <ul class="collapse">
  <li><a href="#load-the-data" id="toc-load-the-data" class="nav-link" data-scroll-target="#load-the-data">Load the data</a></li>
  <li><a href="#create-a-matrix" id="toc-create-a-matrix" class="nav-link" data-scroll-target="#create-a-matrix">Create A matrix</a></li>
  </ul></li>
  <li><a href="#pseudo-count-log-transformation" id="toc-pseudo-count-log-transformation" class="nav-link" data-scroll-target="#pseudo-count-log-transformation">Pseudo-count + log transformation</a></li>
  <li><a href="#model-fitting" id="toc-model-fitting" class="nav-link" data-scroll-target="#model-fitting">Model fitting</a>
  <ul class="collapse">
  <li><a href="#train-test-split" id="toc-train-test-split" class="nav-link" data-scroll-target="#train-test-split">Train-test split</a></li>
  <li><a href="#sparse-log-contrast" id="toc-sparse-log-contrast" class="nav-link" data-scroll-target="#sparse-log-contrast">Sparse log-contrast</a></li>
  <li><a href="#trac" id="toc-trac" class="nav-link" data-scroll-target="#trac">trac</a></li>
  <li><a href="#weighted-trac" id="toc-weighted-trac" class="nav-link" data-scroll-target="#weighted-trac">Weighted trac</a></li>
  </ul></li>
  <li><a href="#check-prediction" id="toc-check-prediction" class="nav-link" data-scroll-target="#check-prediction">Check prediction</a></li>
  <li><a href="#plot-coefficients-and-trac-results-on-a-tree" id="toc-plot-coefficients-and-trac-results-on-a-tree" class="nav-link" data-scroll-target="#plot-coefficients-and-trac-results-on-a-tree">Plot coefficients and trac results on a tree</a>
  <ul class="collapse">
  <li><a href="#slc" id="toc-slc" class="nav-link" data-scroll-target="#slc">SLC</a></li>
  <li><a href="#trac-1" id="toc-trac-1" class="nav-link" data-scroll-target="#trac-1">trac</a></li>
  <li><a href="#weighted-trac-1" id="toc-weighted-trac-1" class="nav-link" data-scroll-target="#weighted-trac-1">Weighted trac</a></li>
  </ul></li>
  <li><a href="#computational-environment" id="toc-computational-environment" class="nav-link" data-scroll-target="#computational-environment">Computational environment</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Sparse log-contrast and trac on 88 soil</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="run-sparse-log-contrast-and-trac-on-88-soil" class="level2">
<h2 class="anchored" data-anchor-id="run-sparse-log-contrast-and-trac-on-88-soil">Run sparse log-contrast and trac on 88 soil</h2>
<p>We will need several packages for the analysis:</p>
<ol type="1">
<li><p><code>phyloseq</code> for the data structure,</p></li>
<li><p><code>tidyverse</code> for data wrangling and plotting,</p></li>
<li><p><code>trac</code> for fitting models with sparse log-contrast (SLC) and trac,</p></li>
<li><p><code>kableExtra</code> for pretty tables,</p></li>
<li><p><code>reticulate</code> to specify the python environment and</p></li>
<li><p><code>ggtree</code> to plot the tree.</p></li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse) <span class="co"># data wrangling and visualization</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.1     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.1     ✔ tibble    3.2.1
✔ lubridate 1.9.2     ✔ tidyr     1.3.0
✔ purrr     1.0.1     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq) <span class="co"># microbiome data wrangling</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(trac) <span class="co"># trac</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(kableExtra) <span class="co"># creating tables in rmarkdown</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'kableExtra'

The following object is masked from 'package:dplyr':

    group_rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate) <span class="co"># specify the python environment</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'reticulate'

The following object is masked from 'package:phyloseq':

    import</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>ggtree v3.4.1 For help: https://yulab-smu.top/treedata-book/

If you use the ggtree package suite in published research, please cite
the appropriate paper(s):

Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam.
ggtree: an R package for visualization and annotation of phylogenetic
trees with their covariates and other associated data. Methods in
Ecology and Evolution. 2017, 8(1):28-36. doi:10.1111/2041-210X.12628

Guangchuang Yu. Using ggtree to visualize data on tree-like structures.
Current Protocols in Bioinformatics. 2020, 69:e96. doi:10.1002/cpbi.96

Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods
for mapping and visualizing associated data on phylogeny using ggtree.
Molecular Biology and Evolution. 2018, 35(12):3041-3043.
doi:10.1093/molbev/msy194

Attaching package: 'ggtree'

The following object is masked from 'package:tidyr':

    expand</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>reticulate<span class="sc">::</span><span class="fu">use_condaenv</span>(<span class="st">"trac"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="load-the-data" class="level3">
<h3 class="anchored" data-anchor-id="load-the-data">Load the data</h3>
<p>Read the 88 soil dataset which consist in this case of 89 samples with 116 OTUs and the covariate pH.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>soil <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/soil_processed/soil.RDS"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(soil<span class="sc">@</span>otu_table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  89 116</code></pre>
</div>
</div>
<p>For SLC and trac we need the ingredient:</p>
<ol type="1">
<li>Response y</li>
<li>A compositional data matrix x, (i.e.&nbsp;OTUs).</li>
<li>(trac only:) A binary matrix descendent matrix A.</li>
</ol>
</section>
<section id="create-a-matrix" class="level3">
<h3 class="anchored" data-anchor-id="create-a-matrix">Create A matrix</h3>
<p>We will first derive the descendant matrix A.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. extract the taxonomic table from the phyloseq object</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>tax <span class="ot">&lt;-</span> soil<span class="sc">@</span>tax_table<span class="sc">@</span>.Data</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. add an OTU column based on the rownames and name the column OTU</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>tax <span class="ot">&lt;-</span> <span class="fu">cbind</span>(tax, <span class="fu">rownames</span>(tax))</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tax)[<span class="fu">ncol</span>(tax)] <span class="ot">&lt;-</span> <span class="st">"OTU"</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># In this example is rooted in bacteria since we only consider bacteria. If one also wants to</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="co"># include archea one needs to add another root node e.g. cbind(root = "life", tax). Need to</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co"># adjust the for loops in the next steps and add the root to the formula in step 5</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="co"># 3. Replace unknown taxonomic assignement on the corresponding level</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="co"># iterate over all levels</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_len</span>(<span class="dv">7</span>)) {</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add a number when the type is unknown... e.g. "g__"</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>  ii <span class="ot">&lt;-</span> <span class="fu">nchar</span>(tax[, i]) <span class="sc">==</span> <span class="dv">3</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">sum</span>(ii) <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>    tax[ii, i] <span class="ot">&lt;-</span> <span class="fu">paste0</span>(tax[ii, i], <span class="dv">1</span><span class="sc">:</span><span class="fu">sum</span>(ii))</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a><span class="co"># 4. add the ancestor information with ::</span></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a><span class="co"># iterate over all levels</span></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">8</span>) {</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>  tax[, i] <span class="ot">&lt;-</span> <span class="fu">paste</span>(tax[, i<span class="dv">-1</span>], tax[, i], <span class="at">sep =</span> <span class="st">"::"</span>)</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>tax <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(tax, <span class="at">stringsAsFactors =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a><span class="co"># 5. form phylo object:</span></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>tree1 <span class="ot">&lt;-</span> <span class="fu">tax_table_to_phylo</span>(<span class="sc">~</span>Kingdom<span class="sc">/</span>Phylum<span class="sc">/</span>Class<span class="sc">/</span>Order<span class="sc">/</span></span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>                              Family<span class="sc">/</span>Genus<span class="sc">/</span>Species<span class="sc">/</span>OTU,</span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>                            <span class="at">data =</span> tax, <span class="at">collapse =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a><span class="co"># 6. convert this to an A matrix to be used for aggregation:</span></span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">phylo_to_A</span>(tree1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s see how the taxonomic tree looks like.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the different taxonomic levels</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>tax_levels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Kingdom"</span>,<span class="st">"Phylum"</span>,<span class="st">"Class"</span>,<span class="st">"Order"</span>,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Family"</span>,<span class="st">"Genus"</span>,<span class="st">"Species"</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Summarize different OTUs by their phylum level</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>phylum_preprocess <span class="ot">&lt;-</span> <span class="fu">enframe</span>(tree1<span class="sc">$</span>tip.label) <span class="sc">%&gt;%</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(value,</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>           <span class="at">into =</span> <span class="fu">as.character</span>(tax_levels),</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>           <span class="at">sep =</span> <span class="st">"::"</span>,</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>           <span class="at">remove =</span> <span class="cn">FALSE</span>,</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="st">"right"</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Phylum =</span> <span class="fu">str_remove_all</span>(Phylum, <span class="st">"[a-z]__"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>  <span class="co">#  mutate(Genus = str_remove_all(Genus, "[a-z]__")) %&gt;%</span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Phylum, value) <span class="sc">%&gt;%</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Phylum =</span> <span class="fu">case_when</span>(</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>    Phylum <span class="sc">==</span> <span class="st">"Acidobacteria"</span> <span class="sc">~</span> <span class="st">"Acidobacteria"</span>,</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>    Phylum <span class="sc">==</span> <span class="st">"Proteobacteria"</span> <span class="sc">~</span> <span class="st">"Proteobacteria"</span>,</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>    Phylum <span class="sc">==</span> <span class="st">"Actinobacteria"</span> <span class="sc">~</span> <span class="st">"Actinobacteria"</span>,</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>    Phylum <span class="sc">==</span> <span class="st">"Firmicutes"</span> <span class="sc">~</span> <span class="st">"Firmicutes"</span>,</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"Others"</span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>  ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Expected 7 pieces. Additional pieces discarded in 116 rows [1, 2, 3, 4, 5, 6,
7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, ...].</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>phylum_group <span class="ot">&lt;-</span> <span class="fu">split</span>(phylum_preprocess<span class="sc">$</span>value, phylum_preprocess<span class="sc">$</span>Phylum)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="co"># match the name of alpha to the nodes of the tree</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>nodes_tree <span class="ot">&lt;-</span> <span class="fu">c</span>(tree1<span class="sc">$</span>tip.label, </span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>                tree1<span class="sc">$</span>node.label)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>tree_1 <span class="ot">&lt;-</span> <span class="fu">groupOTU</span>(tree1, phylum_group)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree_1, <span class="fu">aes</span>(<span class="at">color =</span> group), <span class="at">layout =</span> <span class="st">"circular"</span>) <span class="sc">+</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>, <span class="at">name =</span> <span class="st">"Phylum"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'tidytree'</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>tree</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/first-glimps-tree-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="pseudo-count-log-transformation" class="level2">
<h2 class="anchored" data-anchor-id="pseudo-count-log-transformation">Pseudo-count + log transformation</h2>
<p>trac and SLC expect log transformed values as inputs. Since <span class="math inline">\(log(0)\)</span> is undefined we need to add a pseudo-count (e.g.&nbsp;<span class="math inline">\(1\)</span>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>log_pseudo <span class="ot">&lt;-</span> <span class="cf">function</span>(x, <span class="at">pseudo_count =</span> <span class="dv">1</span>) <span class="fu">log</span>(x <span class="sc">+</span> pseudo_count)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>z <span class="ot">&lt;-</span> <span class="fu">log_pseudo</span>(soil<span class="sc">@</span>otu_table<span class="sc">@</span>.Data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="model-fitting" class="level2">
<h2 class="anchored" data-anchor-id="model-fitting">Model fitting</h2>
<section id="train-test-split" class="level3">
<h3 class="anchored" data-anchor-id="train-test-split">Train-test split</h3>
<p>First we split the data into a train-test split to evaluate the predictive performance at the end. We’ll take 2/3 of the observations for training.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extact pH</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> soil<span class="sc">@</span>sam_data<span class="sc">$</span>ph</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># define seed for reproducibility</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co"># train test split</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>ntot <span class="ot">&lt;-</span> <span class="fu">length</span>(y)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fl">0.75</span> <span class="sc">*</span> ntot)</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>tr <span class="ot">&lt;-</span> <span class="fu">sample</span>(ntot, n)</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a><span class="co"># define training and test data</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>ytr <span class="ot">&lt;-</span> y[tr]</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>yte <span class="ot">&lt;-</span> y[<span class="sc">-</span>tr]</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>ztr <span class="ot">&lt;-</span> z[tr, ]</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>zte <span class="ot">&lt;-</span> z[<span class="sc">-</span>tr, ]</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a><span class="co"># set the folds for cross-validation</span></span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>folds <span class="ot">&lt;-</span> ggb<span class="sc">:::</span><span class="fu">make_folds</span>(<span class="fu">length</span>(ytr), <span class="at">nfolds =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(ztr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  67 116</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(zte)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  22 116</code></pre>
</div>
</div>
</section>
<section id="sparse-log-contrast" class="level3">
<h3 class="anchored" data-anchor-id="sparse-log-contrast">Sparse log-contrast</h3>
<p>We will fit the model on the train data. The task is to predict pH based on the OTUs available. The method will solve the lambda path. Selecting the optimal tuning parameter <span class="math inline">\(\lambda\)</span> is a crucial.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>fit_log_contrast <span class="ot">&lt;-</span> <span class="fu">sparse_log_contrast</span>(<span class="at">Z =</span> ztr, <span class="at">y =</span> ytr)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>fit_log_contrast<span class="sc">$</span>beta <span class="sc">%&gt;%</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">frac =</span> fit_log_contrast<span class="sc">$</span>fraclist) <span class="sc">%&gt;%</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>frac) <span class="sc">%&gt;%</span> </span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> frac, <span class="at">y =</span> value, </span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>                               <span class="at">group =</span> name, <span class="at">color =</span> name)) <span class="sc">+</span> </span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span> </span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">y =</span> <span class="fu">expression</span>(beta), <span class="at">x =</span> <span class="st">"Fraction of lambda_max"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/SLC-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Cross validate SLC to choose tuning parameter <span class="math inline">\(\lambda\)</span>. The function <code>cv_sparse_log_contrast</code> can be used to determine the optimal parameter. Dotted line represents the 1SE rule.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>cvfit_log_contrast <span class="ot">&lt;-</span> </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cv_sparse_log_contrast</span>(fit_log_contrast, <span class="at">Z =</span> ztr, <span class="at">y =</span> ytr,</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">folds =</span> folds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>fold 1
fold 2
fold 3
fold 4
fold 5
fold 6
fold 7
fold 8
fold 9
fold 10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>plot_cv_log_contrast <span class="ot">&lt;-</span> <span class="cf">function</span>(cvfit) {</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract number of non-zero coefficients, mean error and sd</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">non_zero =</span> cvfit<span class="sc">$</span>cv<span class="sc">$</span>nonzeros,</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">error =</span> cvfit<span class="sc">$</span>cv<span class="sc">$</span>m,</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">sd =</span> cvfit<span class="sc">$</span>cv<span class="sc">$</span>se) <span class="sc">%&gt;%</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> non_zero, <span class="at">y =</span> error)) <span class="sc">+</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 1se line</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> cvfit<span class="sc">$</span>cv<span class="sc">$</span>nonzeros[cvfit<span class="sc">$</span>cv<span class="sc">$</span>i1se],</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>               <span class="at">linetype =</span> <span class="st">"longdash"</span>) <span class="sc">+</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ibest line</span></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> cvfit<span class="sc">$</span>cv<span class="sc">$</span>nonzeros[cvfit<span class="sc">$</span>cv<span class="sc">$</span>ibest]) <span class="sc">+</span></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># line + point and error bars</span></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> error <span class="sc">-</span> sd, <span class="at">ymax =</span> error <span class="sc">+</span> sd), <span class="at">width =</span> .<span class="dv">2</span>,</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>                  <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="fl">0.05</span>))</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_cv_log_contrast</span>(cvfit_log_contrast)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/SLC-cv-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Plot the path again, this time with the cross-validation results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>fit_log_contrast<span class="sc">$</span>beta <span class="sc">%&gt;%</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">frac =</span> fit_log_contrast<span class="sc">$</span>fraclist) <span class="sc">%&gt;%</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>frac) <span class="sc">%&gt;%</span> </span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> frac, <span class="at">y =</span> value, </span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>                               <span class="at">group =</span> name, <span class="at">color =</span> name)) <span class="sc">+</span> </span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span> </span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">y =</span> <span class="fu">expression</span>(beta), <span class="at">x =</span> <span class="st">"Fraction of lambda_max"</span>) <span class="sc">+</span></span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> fit_log_contrast<span class="sc">$</span>fraclist[cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se],</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"longdash"</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> fit_log_contrast<span class="sc">$</span>fraclist[cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>ibest],</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/plot-slc-path-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="trac" class="level3">
<h3 class="anchored" data-anchor-id="trac">trac</h3>
<p>Let’s continue with the trac model using <code>trac</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>fit_trac <span class="ot">&lt;-</span> <span class="fu">trac</span>(<span class="at">Z =</span> ztr, <span class="at">y =</span> ytr, <span class="at">A =</span> A)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Cross validate trac to choose tuning parameter <span class="math inline">\(\lambda\)</span>. The function <code>cv_trac</code> can be used to determine the optimal parameter. Dotted line represents the 1SE rule.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>cvfit_trac <span class="ot">&lt;-</span> </span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cv_trac</span>(fit_trac, <span class="at">Z =</span> ztr, <span class="at">y =</span> ytr, <span class="at">A =</span> A,</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">folds =</span> folds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>fold 1
fold 2
fold 3
fold 4
fold 5
fold 6
fold 7
fold 8
fold 9
fold 10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>plot_cv_trac <span class="ot">&lt;-</span> <span class="cf">function</span>(cvfit) {</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract number of non-zero coefficients, mean error and sd</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">non_zero =</span> cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>nonzeros,</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">error =</span> cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>m,</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">sd =</span> cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>se) <span class="sc">%&gt;%</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> non_zero, <span class="at">y =</span> error)) <span class="sc">+</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 1se line</span></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>nonzeros[cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se],</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">"gray"</span>, <span class="at">linetype =</span> <span class="st">"longdash"</span>) <span class="sc">+</span></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ibest line</span></span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>nonzeros[cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest],</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">"gray50"</span>, <span class="at">linetype =</span> <span class="st">"longdash"</span>) <span class="sc">+</span></span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># line + point and error bars</span></span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> error <span class="sc">-</span> sd, <span class="at">ymax =</span> error <span class="sc">+</span> sd), <span class="at">width =</span> .<span class="dv">2</span>,</span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>                  <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="fl">0.05</span>))</span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-21"><a href="#cb33-21" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_cv_trac</span>(cvfit_trac)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/trac-cv-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Plot the coefficient path.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">frac =</span> fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>fraclist) <span class="sc">%&gt;%</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>frac) <span class="sc">%&gt;%</span> </span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> frac, <span class="at">y =</span> value, </span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>                               <span class="at">group =</span> name, <span class="at">color =</span> name)) <span class="sc">+</span> </span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span> </span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">y =</span> <span class="fu">expression</span>(alpha), <span class="at">x =</span> <span class="st">"Fraction of lambda_max"</span>) <span class="sc">+</span></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>fraclist[cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se],</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"longdash"</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>fraclist[cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest],</span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/trac-path-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="weighted-trac" class="level3">
<h3 class="anchored" data-anchor-id="weighted-trac">Weighted trac</h3>
<p>The path of trac with normal weights seems interesting. It might be better to use the weighted trac, nudging the model to select on a lower taxonomic level.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>fit_weighted_trac <span class="ot">&lt;-</span> <span class="fu">trac</span>(<span class="at">Z =</span> ztr, <span class="at">y =</span> ytr, <span class="at">A =</span> A, <span class="at">w =</span> <span class="fu">colSums</span>(<span class="fu">as.matrix</span>(A))<span class="sc">^</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>cvfit_weighted_trac <span class="ot">&lt;-</span> </span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cv_trac</span>(fit_weighted_trac, <span class="at">Z =</span> ztr, <span class="at">y =</span> ytr, <span class="at">A =</span> A,</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">folds =</span> folds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>fold 1
fold 2
fold 3
fold 4
fold 5
fold 6
fold 7
fold 8
fold 9
fold 10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_cv_trac</span>(cvfit_weighted_trac)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/weighted-trac-cv-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha <span class="sc">%&gt;%</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">frac =</span> fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>fraclist) <span class="sc">%&gt;%</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>frac) <span class="sc">%&gt;%</span> </span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> frac, <span class="at">y =</span> value, </span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>                               <span class="at">group =</span> name, <span class="at">color =</span> name)) <span class="sc">+</span> </span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span> </span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">y =</span> <span class="fu">expression</span>(alpha), <span class="at">x =</span> <span class="st">"Fraction of lambda_max"</span>) <span class="sc">+</span></span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>fraclist[cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se],</span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"longdash"</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>fraclist[cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest],</span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/weighted-trac-path-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="check-prediction" class="level2">
<h2 class="anchored" data-anchor-id="check-prediction">Check prediction</h2>
<p>Let’s evaluate the different models on the test set and see how they perform based on the RMSE.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>yhat_slc <span class="ot">&lt;-</span> <span class="fu">predict_sparse_log_contrast</span>(fit_log_contrast, <span class="at">new_Z =</span> zte)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>yhat_trac <span class="ot">&lt;-</span> <span class="fu">predict_trac</span>(fit_trac, <span class="at">new_Z =</span> zte)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>yhat_weighted_trac <span class="ot">&lt;-</span> <span class="fu">predict_trac</span>(fit_weighted_trac, <span class="at">new_Z =</span> zte)</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>yhat_slc_1se <span class="ot">&lt;-</span> yhat_slc[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se]</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>yhat_slc_best <span class="ot">&lt;-</span> yhat_slc[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>ibest]</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>yhat_trac_1se <span class="ot">&lt;-</span> yhat_trac[[<span class="dv">1</span>]][, cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se]</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>yhat_trac_best <span class="ot">&lt;-</span> yhat_trac[[<span class="dv">1</span>]][, cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest]</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>yhat_weighted_trac_1se <span class="ot">&lt;-</span> yhat_weighted_trac[[<span class="dv">1</span>]][, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se]</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>yhat_weighted_trac_best <span class="ot">&lt;-</span> yhat_weighted_trac[[<span class="dv">1</span>]][, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest]</span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a>method_col <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"SLC (1SE)"</span>,  <span class="st">"SLC (best)"</span>, <span class="st">"trac (1SE)"</span>, <span class="st">"trac (best)"</span>,</span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Weighted trac (1SE)"</span>, <span class="st">"Weighted trac (best)"</span>)</span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a>RMSE <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">sqrt</span>(<span class="fu">mean</span>((yte <span class="sc">-</span> yhat_slc_1se)<span class="sc">^</span><span class="dv">2</span>)),</span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a>          <span class="fu">sqrt</span>(<span class="fu">mean</span>((yte <span class="sc">-</span> yhat_slc_best)<span class="sc">^</span><span class="dv">2</span>)),</span>
<span id="cb40-18"><a href="#cb40-18" aria-hidden="true" tabindex="-1"></a>          <span class="fu">sqrt</span>(<span class="fu">mean</span>((yte <span class="sc">-</span> yhat_trac_1se)<span class="sc">^</span><span class="dv">2</span>)),</span>
<span id="cb40-19"><a href="#cb40-19" aria-hidden="true" tabindex="-1"></a>          <span class="fu">sqrt</span>(<span class="fu">mean</span>((yte <span class="sc">-</span> yhat_trac_best)<span class="sc">^</span><span class="dv">2</span>)),</span>
<span id="cb40-20"><a href="#cb40-20" aria-hidden="true" tabindex="-1"></a>          <span class="fu">sqrt</span>(<span class="fu">mean</span>((yte <span class="sc">-</span> yhat_weighted_trac_1se)<span class="sc">^</span><span class="dv">2</span>)),</span>
<span id="cb40-21"><a href="#cb40-21" aria-hidden="true" tabindex="-1"></a>          <span class="fu">sqrt</span>(<span class="fu">mean</span>((yte <span class="sc">-</span> yhat_weighted_trac_best)<span class="sc">^</span><span class="dv">2</span>)))</span>
<span id="cb40-22"><a href="#cb40-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-23"><a href="#cb40-23" aria-hidden="true" tabindex="-1"></a>NZ <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">sum</span>(fit_log_contrast<span class="sc">$</span>beta[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se] <span class="sc">!=</span> <span class="dv">0</span>),</span>
<span id="cb40-24"><a href="#cb40-24" aria-hidden="true" tabindex="-1"></a>        <span class="fu">sum</span>(fit_log_contrast<span class="sc">$</span>beta[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>ibest] <span class="sc">!=</span> <span class="dv">0</span>),</span>
<span id="cb40-25"><a href="#cb40-25" aria-hidden="true" tabindex="-1"></a>        <span class="fu">sum</span>(fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">!=</span> <span class="dv">0</span>),</span>
<span id="cb40-26"><a href="#cb40-26" aria-hidden="true" tabindex="-1"></a>        <span class="fu">sum</span>(fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest] <span class="sc">!=</span> <span class="dv">0</span>),</span>
<span id="cb40-27"><a href="#cb40-27" aria-hidden="true" tabindex="-1"></a>        <span class="fu">sum</span>(fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">!=</span> <span class="dv">0</span>),</span>
<span id="cb40-28"><a href="#cb40-28" aria-hidden="true" tabindex="-1"></a>        <span class="fu">sum</span>(fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest] <span class="sc">!=</span> <span class="dv">0</span>))</span>
<span id="cb40-29"><a href="#cb40-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-30"><a href="#cb40-30" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(method_col, NZ, RMSE) <span class="sc">%&gt;%</span></span>
<span id="cb40-31"><a href="#cb40-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">Method =</span> method_col) <span class="sc">%&gt;%</span></span>
<span id="cb40-32"><a href="#cb40-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">RMSE =</span> <span class="fu">round</span>(RMSE, <span class="at">digits =</span> <span class="dv">3</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb40-33"><a href="#cb40-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kbl</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">Method</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">NZ</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">RMSE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">SLC (1SE)</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">0.518</td>
</tr>
<tr class="even">
<td style="text-align: left;">SLC (best)</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">0.467</td>
</tr>
<tr class="odd">
<td style="text-align: left;">trac (1SE)</td>
<td style="text-align: right;">31</td>
<td style="text-align: right;">0.517</td>
</tr>
<tr class="even">
<td style="text-align: left;">trac (best)</td>
<td style="text-align: right;">38</td>
<td style="text-align: right;">0.530</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Weighted trac (1SE)</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">0.500</td>
</tr>
<tr class="even">
<td style="text-align: left;">Weighted trac (best)</td>
<td style="text-align: right;">43</td>
<td style="text-align: right;">0.601</td>
</tr>
</tbody>
</table>


</div>
</div>
<p>Let’s plot the out of sample prediction vs the ground truth.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(<span class="at">yte =</span> yte,</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>           <span class="st">"SLC (1se)"</span> <span class="ot">=</span> yhat_slc_1se,</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>           <span class="st">"SLC (best)"</span> <span class="ot">=</span> yhat_slc_best,</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>           <span class="st">"trac (1se)"</span> <span class="ot">=</span> yhat_trac_1se,</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>           <span class="st">"trac (best)"</span> <span class="ot">=</span> yhat_trac_best,</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>           <span class="st">"Weighted trac (1se)"</span> <span class="ot">=</span> yhat_weighted_trac_1se,</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>           <span class="st">"Weighted trac (best)"</span> <span class="ot">=</span> yhat_weighted_trac_best) <span class="sc">%&gt;%</span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>yte) <span class="sc">%&gt;%</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> yte, <span class="at">y =</span> value)) <span class="sc">+</span></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>() <span class="sc">+</span></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>name, <span class="at">ncol =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Predicted pH"</span>) <span class="sc">+</span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"True pH"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/out-of-sample-plot-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Finally, plot out of sample error vs the different number of non-zero coefficients. for SLC, trac, and weighted trac.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>plot_nzz_mse <span class="ot">&lt;-</span> <span class="cf">function</span>(yhat, y, coefs, i1se, ibest) {</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  mse <span class="ot">&lt;-</span> <span class="fu">colMeans</span>((yhat <span class="sc">-</span> y)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>  nnz <span class="ot">&lt;-</span> <span class="fu">colSums</span>(coefs <span class="sc">!=</span> <span class="dv">0</span>)</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(mse, nnz) <span class="sc">%&gt;%</span></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> nnz, <span class="at">y =</span> mse)) <span class="sc">+</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> nnz[i1se], <span class="at">linetype =</span> <span class="st">"longdash"</span>) <span class="sc">+</span> </span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> nnz[ibest])</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_nzz_mse</span>(<span class="at">yhat =</span> yhat_slc, <span class="at">y =</span> yte, <span class="at">coefs =</span> fit_log_contrast<span class="sc">$</span>beta,</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">i1se =</span> cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se,</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>             <span class="at">ibest =</span> cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>ibest)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/plot-slc-sparsity-vs-mse-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_nzz_mse</span>(<span class="at">yhat =</span> yhat_trac[[<span class="dv">1</span>]], <span class="at">y =</span> yte, <span class="at">coefs =</span> fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha,</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">i1se =</span> cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se,</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">ibest =</span> cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/plot-trac-sparsity-vs-mse-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_nzz_mse</span>(<span class="at">yhat =</span> yhat_weighted_trac[[<span class="dv">1</span>]], </span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> yte, <span class="at">coefs =</span> fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha,</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">i1se =</span> cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se,</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">ibest =</span> cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/plot-weighted-trac-sparsity-vs-mse-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="plot-coefficients-and-trac-results-on-a-tree" class="level2">
<h2 class="anchored" data-anchor-id="plot-coefficients-and-trac-results-on-a-tree">Plot coefficients and trac results on a tree</h2>
<section id="slc" class="level3">
<h3 class="anchored" data-anchor-id="slc">SLC</h3>
<p>Start with the <span class="math inline">\(\beta\)</span> coefficients of the 1SE solution. The first plot shows the last known taxonomic annotation for the corresponding OTU.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get which coefficients are not zero</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>nzz <span class="ot">&lt;-</span> <span class="cf">function</span>(coefs) {</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  coefs[coefs <span class="sc">!=</span> <span class="dv">0</span>]</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a><span class="co"># for slc 1se</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>slc_names <span class="ot">&lt;-</span> fit_log_contrast<span class="sc">$</span>beta[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>  nzz </span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a><span class="co"># get the last known name for slc</span></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>slc_names_last_known <span class="ot">&lt;-</span> soil<span class="sc">@</span>tax_table<span class="sc">@</span>.Data <span class="sc">%&gt;%</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>  as.data.frame <span class="sc">%&gt;%</span></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>() <span class="sc">%&gt;%</span></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(rowname <span class="sc">%in%</span> <span class="fu">names</span>(slc_names))</span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a>last_known_tax <span class="ot">&lt;-</span> <span class="fu">apply</span>(slc_names_last_known, <span class="at">MARGIN =</span> <span class="dv">2</span>, </span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">str_length</span>(x) <span class="sc">==</span> <span class="dv">3</span>) </span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a>last_known_slc <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(slc_names_last_known)) {</span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a>  last_known_slc <span class="ot">&lt;-</span> <span class="fu">append</span>(last_known_slc,</span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(slc_names_last_known[i, (<span class="dv">8</span> <span class="sc">-</span> <span class="fu">rowSums</span>(last_known_tax)[i])], <span class="st">"::"</span>,</span>
<span id="cb45-22"><a href="#cb45-22" aria-hidden="true" tabindex="-1"></a>         slc_names_last_known[i, <span class="dv">1</span>]))</span>
<span id="cb45-23"><a href="#cb45-23" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb45-24"><a href="#cb45-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-25"><a href="#cb45-25" aria-hidden="true" tabindex="-1"></a>last_known_slc <span class="ot">&lt;-</span> </span>
<span id="cb45-26"><a href="#cb45-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(<span class="at">name =</span> slc_names_last_known<span class="sc">$</span>rowname, <span class="at">last_name =</span> last_known_slc)</span>
<span id="cb45-27"><a href="#cb45-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-28"><a href="#cb45-28" aria-hidden="true" tabindex="-1"></a><span class="co"># start plotting</span></span>
<span id="cb45-29"><a href="#cb45-29" aria-hidden="true" tabindex="-1"></a>fit_log_contrast<span class="sc">$</span>beta[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb45-30"><a href="#cb45-30" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb45-31"><a href="#cb45-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">enframe</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb45-32"><a href="#cb45-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(last_known_slc) <span class="sc">%&gt;%</span></span>
<span id="cb45-33"><a href="#cb45-33" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">fct_reorder</span>(last_name, value), <span class="at">y =</span> value)) <span class="sc">+</span> </span>
<span id="cb45-34"><a href="#cb45-34" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb45-35"><a href="#cb45-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">fct_reorder</span>(last_name, value), </span>
<span id="cb45-36"><a href="#cb45-36" aria-hidden="true" tabindex="-1"></a>                   <span class="at">xend=</span> <span class="fu">fct_reorder</span>(last_name, value),</span>
<span id="cb45-37"><a href="#cb45-37" aria-hidden="true" tabindex="-1"></a>                   <span class="at">y=</span><span class="dv">0</span>, <span class="at">yend=</span>value)) <span class="sc">+</span></span>
<span id="cb45-38"><a href="#cb45-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()<span class="sc">+</span> </span>
<span id="cb45-39"><a href="#cb45-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb45-40"><a href="#cb45-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">label=</span><span class="cf">function</span>(x) <span class="fu">str_split_i</span>(x, <span class="at">pattern =</span> <span class="st">"::"</span>, <span class="at">i =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb45-41"><a href="#cb45-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"OTUs (as lowest known taxonomic rank)"</span>) <span class="sc">+</span></span>
<span id="cb45-42"><a href="#cb45-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(beta))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(name)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/plot-slc-coef-last-known-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Once again only with the the OTU ids.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>fit_log_contrast<span class="sc">$</span>beta[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">enframe</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">fct_reorder</span>(name, value), <span class="at">y =</span> value)) <span class="sc">+</span> </span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">fct_reorder</span>(name, value), </span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">xend=</span> <span class="fu">fct_reorder</span>(name, value),</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">y=</span><span class="dv">0</span>, <span class="at">yend=</span>value)) <span class="sc">+</span></span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()<span class="sc">+</span> </span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">label=</span><span class="cf">function</span>(x) <span class="fu">str_split_i</span>(x, <span class="at">pattern =</span> <span class="st">"::"</span>, <span class="at">i =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"OTUs"</span>) <span class="sc">+</span></span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(beta))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/plot-slc-coef-otu-name-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Sparse log-contrast tree</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>slc_selected <span class="ot">&lt;-</span> tax <span class="sc">%&gt;%</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>() <span class="sc">%&gt;%</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(rowname <span class="sc">%in%</span> <span class="fu">names</span>(slc_names)) <span class="sc">%&gt;%</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(OTU) <span class="sc">%&gt;%</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">OTU =</span> <span class="fu">as.character</span>(OTU),</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">SLC =</span> <span class="cn">TRUE</span>)</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>tree_slc <span class="ot">&lt;-</span> tree <span class="sc">%&lt;+%</span> slc_selected</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>tree_slc <span class="sc">+</span></span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point2</span>(<span class="fu">aes</span>(<span class="at">subset =</span> SLC <span class="sc">==</span> <span class="cn">TRUE</span>), <span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/plot-slc-coef-tree-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="trac-1" class="level3">
<h3 class="anchored" data-anchor-id="trac-1">trac</h3>
<p>A possible reason for the behavior of trac could be because of the filter (OTUs present at least <span class="math inline">\(100\)</span> times). The included OTUs could be biased towards a specific Phylum.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">enframe</span>() <span class="sc">%&gt;%</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(value)) <span class="sc">%&gt;%</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kbl</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">name</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">'k__Bacteria::p__Proteobacteria::c__Alphaproteobacteria'</td>
<td style="text-align: right;">0.9117530</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Actinobacteria::c__Actinobacteria::o__Micrococcales::f__21::g__45::s__63::238800</td>
<td style="text-align: right;">0.2618305</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Gammaproteobacteria::o__Xanthomonadales::f__Sinobacteraceae::g__24::s__31::1121144</td>
<td style="text-align: right;">0.2306242</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__[Chloracidobacteria]::o__RB41::f__6::g__8::s__10::216549</td>
<td style="text-align: right;">0.1966367</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__Acidobacteriia::o__Acidobacteriales::f__Acidobacteriaceae::g__37::s__51::4440396</td>
<td style="text-align: right;">0.1813117</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__[Chloracidobacteria]::o__RB41::f__35::g__69::s__93::1142972</td>
<td style="text-align: right;">0.1800244</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__Solibacteres::o__Solibacterales::f__10::g__12::s__15::3616127</td>
<td style="text-align: right;">0.1735738</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__Acidobacteriia::o__Acidobacteriales::f__Koribacteraceae::g__79::s__106::4341561</td>
<td style="text-align: right;">0.1234441</td>
</tr>
<tr class="odd">
<td style="text-align: left;">'k__Bacteria::p__Gemmatimonadetes::c__Gemmatimonadetes::o__N1423WL'</td>
<td style="text-align: right;">0.1200462</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Alphaproteobacteria::o__Rhizobiales::f__Hyphomicrobiaceae::g__Rhodoplanes::s__57::319907</td>
<td style="text-align: right;">0.0993545</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__DA052::o__Ellin6513::f__12::g__16::s__19::113212</td>
<td style="text-align: right;">0.0932988</td>
</tr>
<tr class="even">
<td style="text-align: left;">'k__Bacteria::p__Acidobacteria::c__DA052::o__Ellin6513'</td>
<td style="text-align: right;">0.0867691</td>
</tr>
<tr class="odd">
<td style="text-align: left;">'k__Bacteria::p__Proteobacteria::c__Alphaproteobacteria::o__Rhizobiales::f__Bradyrhizobiaceae::g__Balneimonas'</td>
<td style="text-align: right;">0.0687826</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Gemmatimonadetes::c__Gemmatimonadetes::o__N1423WL::f__31::g__62::s__84::4423681</td>
<td style="text-align: right;">0.0415569</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Alphaproteobacteria::o__Rhodospirillales::f__Rhodospirillaceae::g__25::s__32::357011</td>
<td style="text-align: right;">0.0058405</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__[Chloracidobacteria]::o__RB41::f__24::g__50::s__68::4468200</td>
<td style="text-align: right;">-0.0184326</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__DA052::o__Ellin6513::f__17::g__36::s__50::1024089</td>
<td style="text-align: right;">-0.0210397</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Betaproteobacteria::o__Burkholderiales::f__Comamonadaceae::g__26::s__33::214611</td>
<td style="text-align: right;">-0.0240086</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Alphaproteobacteria::o__1::f__18::g__40::s__56::732929</td>
<td style="text-align: right;">-0.0268979</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Gammaproteobacteria::o__Xanthomonadales::f__Sinobacteraceae::g__51::s__69::1122517</td>
<td style="text-align: right;">-0.0425186</td>
</tr>
<tr class="odd">
<td style="text-align: left;">'k__Bacteria::p__Actinobacteria::c__Actinobacteria::o__Actinomycetales::f__Propionibacteriaceae'</td>
<td style="text-align: right;">-0.0441450</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__DA052::o__Ellin6513::f__8::g__10::s__13::2578357</td>
<td style="text-align: right;">-0.0694426</td>
</tr>
<tr class="odd">
<td style="text-align: left;">'k__Bacteria::p__Proteobacteria::c__Betaproteobacteria'</td>
<td style="text-align: right;">-0.0791409</td>
</tr>
<tr class="even">
<td style="text-align: left;">'k__Bacteria::p__Acidobacteria::c__Acidobacteriia::o__Acidobacteriales::f__Acidobacteriaceae'</td>
<td style="text-align: right;">-0.0935579</td>
</tr>
<tr class="odd">
<td style="text-align: left;">'k__Bacteria::p__Actinobacteria::c__Actinobacteria::o__Actinomycetales'</td>
<td style="text-align: right;">-0.1007290</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Alphaproteobacteria::o__Rhizobiales::f__Bradyrhizobiaceae::g__Bradyrhizobium::s__95::146397</td>
<td style="text-align: right;">-0.1351057</td>
</tr>
<tr class="odd">
<td style="text-align: left;">'k__Bacteria::p__Acidobacteria'</td>
<td style="text-align: right;">-0.1652062</td>
</tr>
<tr class="even">
<td style="text-align: left;">'k__Bacteria::p__Firmicutes::c__Bacilli::o__Bacillales'</td>
<td style="text-align: right;">-0.2156016</td>
</tr>
<tr class="odd">
<td style="text-align: left;">'k__Bacteria::p__Bacteroidetes'</td>
<td style="text-align: right;">-0.3390730</td>
</tr>
<tr class="even">
<td style="text-align: left;">'k__Bacteria::p__Actinobacteria::c__Rubrobacteria::o__Rubrobacterales::f__Rubrobacteraceae::g__Rubrobacter'</td>
<td style="text-align: right;">-0.3443785</td>
</tr>
<tr class="odd">
<td style="text-align: left;">'k__Bacteria::p__Acidobacteria::c__[Chloracidobacteria]::o__RB41'</td>
<td style="text-align: right;">-1.0555691</td>
</tr>
</tbody>
</table>


</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>tree_df <span class="ot">&lt;-</span> <span class="fu">enframe</span>(nodes_tree) <span class="sc">%&gt;%</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>name)</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>trac_selected <span class="ot">&lt;-</span> fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>  names <span class="sc">%&gt;%</span></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>  enframe <span class="sc">%&gt;%</span></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">selected_trac =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>name) <span class="sc">%&gt;%</span></span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">right_join</span>(tree_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(value)`</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>tree_trac <span class="ot">&lt;-</span> tree <span class="sc">%&lt;+%</span> trac_selected</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>tree_trac <span class="sc">+</span></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point2</span>(<span class="fu">aes</span>(<span class="at">subset =</span> selected_trac <span class="sc">==</span> <span class="cn">TRUE</span>), <span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/plot-trac-tree-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="weighted-trac-1" class="level3">
<h3 class="anchored" data-anchor-id="weighted-trac-1">Weighted trac</h3>
<p>Therefore, using weighted trac could be better here.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">enframe</span>() <span class="sc">%&gt;%</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(value)) <span class="sc">%&gt;%</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kbl</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">name</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Gemmatimonadetes::c__Gemmatimonadetes::o__N1423WL::f__31::g__62::s__84::4423681</td>
<td style="text-align: right;">0.2469191</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Actinobacteria::c__Actinobacteria::o__Micrococcales::f__21::g__45::s__63::238800</td>
<td style="text-align: right;">0.2042559</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__Acidobacteriia::o__Acidobacteriales::f__Acidobacteriaceae::g__37::s__51::4440396</td>
<td style="text-align: right;">0.1874899</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__[Chloracidobacteria]::o__RB41::f__35::g__69::s__93::1142972</td>
<td style="text-align: right;">0.0958719</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Gammaproteobacteria::o__Xanthomonadales::f__Sinobacteraceae::g__24::s__31::1121144</td>
<td style="text-align: right;">0.0327851</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__DA052::o__Ellin6513::f__12::g__16::s__19::113212</td>
<td style="text-align: right;">0.0292049</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__[Chloracidobacteria]::o__RB41::f__6::g__8::s__10::216549</td>
<td style="text-align: right;">0.0146034</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__Acidobacteriia::o__Acidobacteriales::f__Koribacteraceae::g__79::s__106::4341561</td>
<td style="text-align: right;">0.0135046</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Actinobacteria::c__Actinobacteria::o__Actinomycetales::f__Geodermatophilaceae::g__Geodermatophilus::s__35::154567</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__Solibacteres::o__Solibacterales::f__15::g__29::s__38::156065</td>
<td style="text-align: right;">-0.0244083</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Alphaproteobacteria::o__1::f__18::g__40::s__56::732929</td>
<td style="text-align: right;">-0.0344659</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__[Chloracidobacteria]::o__RB41::f__29::g__59::s__81::4346060</td>
<td style="text-align: right;">-0.0854696</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Gammaproteobacteria::o__Xanthomonadales::f__Sinobacteraceae::g__51::s__69::1122517</td>
<td style="text-align: right;">-0.1771935</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Betaproteobacteria::o__Burkholderiales::f__Comamonadaceae::g__26::s__33::214611</td>
<td style="text-align: right;">-0.2393001</td>
</tr>
<tr class="odd">
<td style="text-align: left;">'k__Bacteria::p__Actinobacteria::c__Actinobacteria::o__Actinomycetales'</td>
<td style="text-align: right;">-0.2637977</td>
</tr>
</tbody>
</table>


</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">enframe</span>() <span class="sc">%&gt;%</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(value)) <span class="sc">%&gt;%</span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kbl</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">name</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Gemmatimonadetes::c__Gemmatimonadetes::o__N1423WL::f__31::g__62::s__84::4423681</td>
<td style="text-align: right;">0.2469191</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Actinobacteria::c__Actinobacteria::o__Micrococcales::f__21::g__45::s__63::238800</td>
<td style="text-align: right;">0.2042559</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__Acidobacteriia::o__Acidobacteriales::f__Acidobacteriaceae::g__37::s__51::4440396</td>
<td style="text-align: right;">0.1874899</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__[Chloracidobacteria]::o__RB41::f__35::g__69::s__93::1142972</td>
<td style="text-align: right;">0.0958719</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Gammaproteobacteria::o__Xanthomonadales::f__Sinobacteraceae::g__24::s__31::1121144</td>
<td style="text-align: right;">0.0327851</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__DA052::o__Ellin6513::f__12::g__16::s__19::113212</td>
<td style="text-align: right;">0.0292049</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__[Chloracidobacteria]::o__RB41::f__6::g__8::s__10::216549</td>
<td style="text-align: right;">0.0146034</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__Acidobacteriia::o__Acidobacteriales::f__Koribacteraceae::g__79::s__106::4341561</td>
<td style="text-align: right;">0.0135046</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Actinobacteria::c__Actinobacteria::o__Actinomycetales::f__Geodermatophilaceae::g__Geodermatophilus::s__35::154567</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__Solibacteres::o__Solibacterales::f__15::g__29::s__38::156065</td>
<td style="text-align: right;">-0.0244083</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Alphaproteobacteria::o__1::f__18::g__40::s__56::732929</td>
<td style="text-align: right;">-0.0344659</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Acidobacteria::c__[Chloracidobacteria]::o__RB41::f__29::g__59::s__81::4346060</td>
<td style="text-align: right;">-0.0854696</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Gammaproteobacteria::o__Xanthomonadales::f__Sinobacteraceae::g__51::s__69::1122517</td>
<td style="text-align: right;">-0.1771935</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria::p__Proteobacteria::c__Betaproteobacteria::o__Burkholderiales::f__Comamonadaceae::g__26::s__33::214611</td>
<td style="text-align: right;">-0.2393001</td>
</tr>
<tr class="odd">
<td style="text-align: left;">'k__Bacteria::p__Actinobacteria::c__Actinobacteria::o__Actinomycetales'</td>
<td style="text-align: right;">-0.2637977</td>
</tr>
</tbody>
</table>


</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>weighted_trac_selected <span class="ot">&lt;-</span> </span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>  fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>  names <span class="sc">%&gt;%</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>  enframe <span class="sc">%&gt;%</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">selected_weighted_trac =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>name) <span class="sc">%&gt;%</span></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">right_join</span>(tree_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(value)`</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>tree_weighted_trac <span class="ot">&lt;-</span> tree <span class="sc">%&lt;+%</span> weighted_trac_selected</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>tree_weighted_trac <span class="sc">+</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point2</span>(<span class="fu">aes</span>(<span class="at">subset =</span> selected_weighted_trac <span class="sc">==</span> <span class="cn">TRUE</span>), </span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="91_soil_slc_files/figure-html/plot-weighted-trac-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="computational-environment" class="level2">
<h2 class="anchored" data-anchor-id="computational-environment">Computational environment</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.2.3 (2023-03-15 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default

locale:
[1] LC_COLLATE=English_Germany.utf8  LC_CTYPE=English_Germany.utf8   
[3] LC_MONETARY=English_Germany.utf8 LC_NUMERIC=C                    
[5] LC_TIME=English_Germany.utf8    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] ggtree_3.4.1     reticulate_1.28  kableExtra_1.3.4 trac_0.0.2      
 [5] phyloseq_1.40.0  lubridate_1.9.2  forcats_1.0.0    stringr_1.5.0   
 [9] dplyr_1.1.1      purrr_1.0.1      readr_2.1.4      tidyr_1.3.0     
[13] tibble_3.2.1     ggplot2_3.4.1    tidyverse_2.0.0 

loaded via a namespace (and not attached):
  [1] nlme_3.1-162           bitops_1.0-7           RColorBrewer_1.1-3    
  [4] webshot_0.5.4          httr_1.4.5             rprojroot_2.0.3       
  [7] GenomeInfoDb_1.32.2    tools_4.2.3            utf8_1.2.3            
 [10] R6_2.5.1               vegan_2.6-4            lazyeval_0.2.2        
 [13] DBI_1.1.3              BiocGenerics_0.42.0    mgcv_1.8-42           
 [16] colorspace_2.1-0       permute_0.9-7          rhdf5filters_1.8.0    
 [19] ade4_1.7-22            withr_2.5.0            tidyselect_1.2.0      
 [22] compiler_4.2.3         cli_3.6.1              rvest_1.0.3           
 [25] Biobase_2.56.0         xml2_1.3.3             labeling_0.4.2        
 [28] scales_1.2.1           rappdirs_0.3.3         yulab.utils_0.0.6     
 [31] systemfonts_1.0.4      digest_0.6.31          rmarkdown_2.21        
 [34] svglite_2.1.1          XVector_0.36.0         pkgconfig_2.0.3       
 [37] htmltools_0.5.5        highr_0.10             fastmap_1.1.1         
 [40] htmlwidgets_1.6.2      rlang_1.1.0            rstudioapi_0.14       
 [43] farver_2.1.1           gridGraphics_0.5-1     generics_0.1.3        
 [46] jsonlite_1.8.4         RCurl_1.98-1.12        magrittr_2.0.3        
 [49] ggplotify_0.1.0        GenomeInfoDbData_1.2.8 patchwork_1.1.2       
 [52] biomformat_1.24.0      Matrix_1.5-3           Rcpp_1.0.10           
 [55] munsell_0.5.0          S4Vectors_0.34.0       Rhdf5lib_1.18.2       
 [58] fansi_1.0.4            ape_5.7-1              lifecycle_1.0.3       
 [61] stringi_1.7.12         yaml_2.3.7             MASS_7.3-58.2         
 [64] zlibbioc_1.42.0        hsm_0.2.0              rhdf5_2.40.0          
 [67] plyr_1.8.8             grid_4.2.3             parallel_4.2.3        
 [70] crayon_1.5.2           lattice_0.20-45        Biostrings_2.64.0     
 [73] splines_4.2.3          multtest_2.52.0        hms_1.1.3             
 [76] knitr_1.42             pillar_1.9.0           igraph_1.4.1          
 [79] reshape2_1.4.4         codetools_0.2-19       stats4_4.2.3          
 [82] glue_1.6.2             evaluate_0.20          ggfun_0.0.9           
 [85] data.table_1.14.8      treeio_1.20.1          vctrs_0.6.1           
 [88] png_0.1-8              tzdb_0.3.0             foreach_1.5.2         
 [91] gtable_0.3.3           ggb_0.10.0             xfun_0.38             
 [94] tidytree_0.4.2         survival_3.5-3         viridisLite_0.4.1     
 [97] aplot_0.1.10           iterators_1.0.14       IRanges_2.30.0        
[100] cluster_2.1.4          timechange_0.2.0       here_1.0.1            </code></pre>
</div>
</div>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb60" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Sparse log-contrast and trac on 88 soil"</span></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span><span class="co"> </span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a><span class="co">    code-fold: false</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a><span class="an">bibliography:</span><span class="co"> references.bib</span></span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span><span class="co"> visual</span></span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a><span class="fu">## Run sparse log-contrast and trac on 88 soil</span></span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a>We will need several packages for the analysis:</span>
<span id="cb60-14"><a href="#cb60-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-15"><a href="#cb60-15" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span><span class="in">`phyloseq`</span> for the data structure,</span>
<span id="cb60-16"><a href="#cb60-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-17"><a href="#cb60-17" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span><span class="in">`tidyverse`</span> for data wrangling and plotting,</span>
<span id="cb60-18"><a href="#cb60-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-19"><a href="#cb60-19" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span><span class="in">`trac`</span> for fitting models with sparse log-contrast (SLC) and trac,</span>
<span id="cb60-20"><a href="#cb60-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-21"><a href="#cb60-21" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span><span class="in">`kableExtra`</span> for pretty tables,</span>
<span id="cb60-22"><a href="#cb60-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-23"><a href="#cb60-23" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span><span class="in">`reticulate`</span> to specify the python environment and</span>
<span id="cb60-24"><a href="#cb60-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-25"><a href="#cb60-25" aria-hidden="true" tabindex="-1"></a><span class="ss">6.  </span><span class="in">`ggtree`</span> to plot the tree.</span>
<span id="cb60-26"><a href="#cb60-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-29"><a href="#cb60-29" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-30"><a href="#cb60-30" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-packages</span></span>
<span id="cb60-31"><a href="#cb60-31" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse) <span class="co"># data wrangling and visualization</span></span>
<span id="cb60-32"><a href="#cb60-32" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq) <span class="co"># microbiome data wrangling</span></span>
<span id="cb60-33"><a href="#cb60-33" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(trac) <span class="co"># trac</span></span>
<span id="cb60-34"><a href="#cb60-34" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(kableExtra) <span class="co"># creating tables in rmarkdown</span></span>
<span id="cb60-35"><a href="#cb60-35" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate) <span class="co"># specify the python environment</span></span>
<span id="cb60-36"><a href="#cb60-36" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb60-37"><a href="#cb60-37" aria-hidden="true" tabindex="-1"></a>reticulate<span class="sc">::</span><span class="fu">use_condaenv</span>(<span class="st">"trac"</span>)</span>
<span id="cb60-38"><a href="#cb60-38" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-39"><a href="#cb60-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-40"><a href="#cb60-40" aria-hidden="true" tabindex="-1"></a><span class="fu">### Load the data</span></span>
<span id="cb60-41"><a href="#cb60-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-42"><a href="#cb60-42" aria-hidden="true" tabindex="-1"></a>Read the 88 soil dataset which consist in this case of 89 samples with 116 OTUs and the covariate pH.</span>
<span id="cb60-43"><a href="#cb60-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-46"><a href="#cb60-46" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-47"><a href="#cb60-47" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-data</span></span>
<span id="cb60-48"><a href="#cb60-48" aria-hidden="true" tabindex="-1"></a>soil <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/soil_processed/soil.RDS"</span>)</span>
<span id="cb60-49"><a href="#cb60-49" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(soil<span class="sc">@</span>otu_table)</span>
<span id="cb60-50"><a href="#cb60-50" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-51"><a href="#cb60-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-52"><a href="#cb60-52" aria-hidden="true" tabindex="-1"></a>For SLC and trac we need the ingredient:</span>
<span id="cb60-53"><a href="#cb60-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-54"><a href="#cb60-54" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Response y</span>
<span id="cb60-55"><a href="#cb60-55" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>A compositional data matrix x, (i.e. OTUs).</span>
<span id="cb60-56"><a href="#cb60-56" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>(trac only:) A binary matrix descendent matrix A.</span>
<span id="cb60-57"><a href="#cb60-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-58"><a href="#cb60-58" aria-hidden="true" tabindex="-1"></a><span class="fu">### Create A matrix</span></span>
<span id="cb60-59"><a href="#cb60-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-60"><a href="#cb60-60" aria-hidden="true" tabindex="-1"></a>We will first derive the descendant matrix A.</span>
<span id="cb60-61"><a href="#cb60-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-64"><a href="#cb60-64" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-65"><a href="#cb60-65" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: create-A-matrix</span></span>
<span id="cb60-66"><a href="#cb60-66" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb60-67"><a href="#cb60-67" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb60-68"><a href="#cb60-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-69"><a href="#cb60-69" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. extract the taxonomic table from the phyloseq object</span></span>
<span id="cb60-70"><a href="#cb60-70" aria-hidden="true" tabindex="-1"></a>tax <span class="ot">&lt;-</span> soil<span class="sc">@</span>tax_table<span class="sc">@</span>.Data</span>
<span id="cb60-71"><a href="#cb60-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-72"><a href="#cb60-72" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. add an OTU column based on the rownames and name the column OTU</span></span>
<span id="cb60-73"><a href="#cb60-73" aria-hidden="true" tabindex="-1"></a>tax <span class="ot">&lt;-</span> <span class="fu">cbind</span>(tax, <span class="fu">rownames</span>(tax))</span>
<span id="cb60-74"><a href="#cb60-74" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tax)[<span class="fu">ncol</span>(tax)] <span class="ot">&lt;-</span> <span class="st">"OTU"</span></span>
<span id="cb60-75"><a href="#cb60-75" aria-hidden="true" tabindex="-1"></a><span class="co"># In this example is rooted in bacteria since we only consider bacteria. If one also wants to</span></span>
<span id="cb60-76"><a href="#cb60-76" aria-hidden="true" tabindex="-1"></a><span class="co"># include archea one needs to add another root node e.g. cbind(root = "life", tax). Need to</span></span>
<span id="cb60-77"><a href="#cb60-77" aria-hidden="true" tabindex="-1"></a><span class="co"># adjust the for loops in the next steps and add the root to the formula in step 5</span></span>
<span id="cb60-78"><a href="#cb60-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-79"><a href="#cb60-79" aria-hidden="true" tabindex="-1"></a><span class="co"># 3. Replace unknown taxonomic assignement on the corresponding level</span></span>
<span id="cb60-80"><a href="#cb60-80" aria-hidden="true" tabindex="-1"></a><span class="co"># iterate over all levels</span></span>
<span id="cb60-81"><a href="#cb60-81" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_len</span>(<span class="dv">7</span>)) {</span>
<span id="cb60-82"><a href="#cb60-82" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add a number when the type is unknown... e.g. "g__"</span></span>
<span id="cb60-83"><a href="#cb60-83" aria-hidden="true" tabindex="-1"></a>  ii <span class="ot">&lt;-</span> <span class="fu">nchar</span>(tax[, i]) <span class="sc">==</span> <span class="dv">3</span></span>
<span id="cb60-84"><a href="#cb60-84" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">sum</span>(ii) <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb60-85"><a href="#cb60-85" aria-hidden="true" tabindex="-1"></a>    tax[ii, i] <span class="ot">&lt;-</span> <span class="fu">paste0</span>(tax[ii, i], <span class="dv">1</span><span class="sc">:</span><span class="fu">sum</span>(ii))</span>
<span id="cb60-86"><a href="#cb60-86" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-87"><a href="#cb60-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-88"><a href="#cb60-88" aria-hidden="true" tabindex="-1"></a><span class="co"># 4. add the ancestor information with ::</span></span>
<span id="cb60-89"><a href="#cb60-89" aria-hidden="true" tabindex="-1"></a><span class="co"># iterate over all levels</span></span>
<span id="cb60-90"><a href="#cb60-90" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">8</span>) {</span>
<span id="cb60-91"><a href="#cb60-91" aria-hidden="true" tabindex="-1"></a>  tax[, i] <span class="ot">&lt;-</span> <span class="fu">paste</span>(tax[, i<span class="dv">-1</span>], tax[, i], <span class="at">sep =</span> <span class="st">"::"</span>)</span>
<span id="cb60-92"><a href="#cb60-92" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-93"><a href="#cb60-93" aria-hidden="true" tabindex="-1"></a>tax <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(tax, <span class="at">stringsAsFactors =</span> <span class="cn">TRUE</span>)</span>
<span id="cb60-94"><a href="#cb60-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-95"><a href="#cb60-95" aria-hidden="true" tabindex="-1"></a><span class="co"># 5. form phylo object:</span></span>
<span id="cb60-96"><a href="#cb60-96" aria-hidden="true" tabindex="-1"></a>tree1 <span class="ot">&lt;-</span> <span class="fu">tax_table_to_phylo</span>(<span class="sc">~</span>Kingdom<span class="sc">/</span>Phylum<span class="sc">/</span>Class<span class="sc">/</span>Order<span class="sc">/</span></span>
<span id="cb60-97"><a href="#cb60-97" aria-hidden="true" tabindex="-1"></a>                              Family<span class="sc">/</span>Genus<span class="sc">/</span>Species<span class="sc">/</span>OTU,</span>
<span id="cb60-98"><a href="#cb60-98" aria-hidden="true" tabindex="-1"></a>                            <span class="at">data =</span> tax, <span class="at">collapse =</span> <span class="cn">TRUE</span>)</span>
<span id="cb60-99"><a href="#cb60-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-100"><a href="#cb60-100" aria-hidden="true" tabindex="-1"></a><span class="co"># 6. convert this to an A matrix to be used for aggregation:</span></span>
<span id="cb60-101"><a href="#cb60-101" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">phylo_to_A</span>(tree1)</span>
<span id="cb60-102"><a href="#cb60-102" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-103"><a href="#cb60-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-104"><a href="#cb60-104" aria-hidden="true" tabindex="-1"></a>Let's see how the taxonomic tree looks like.</span>
<span id="cb60-105"><a href="#cb60-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-108"><a href="#cb60-108" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-109"><a href="#cb60-109" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: first-glimps-tree</span></span>
<span id="cb60-110"><a href="#cb60-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-111"><a href="#cb60-111" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the different taxonomic levels</span></span>
<span id="cb60-112"><a href="#cb60-112" aria-hidden="true" tabindex="-1"></a>tax_levels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Kingdom"</span>,<span class="st">"Phylum"</span>,<span class="st">"Class"</span>,<span class="st">"Order"</span>,</span>
<span id="cb60-113"><a href="#cb60-113" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Family"</span>,<span class="st">"Genus"</span>,<span class="st">"Species"</span>)</span>
<span id="cb60-114"><a href="#cb60-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-115"><a href="#cb60-115" aria-hidden="true" tabindex="-1"></a><span class="co"># Summarize different OTUs by their phylum level</span></span>
<span id="cb60-116"><a href="#cb60-116" aria-hidden="true" tabindex="-1"></a>phylum_preprocess <span class="ot">&lt;-</span> <span class="fu">enframe</span>(tree1<span class="sc">$</span>tip.label) <span class="sc">%&gt;%</span></span>
<span id="cb60-117"><a href="#cb60-117" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(value,</span>
<span id="cb60-118"><a href="#cb60-118" aria-hidden="true" tabindex="-1"></a>           <span class="at">into =</span> <span class="fu">as.character</span>(tax_levels),</span>
<span id="cb60-119"><a href="#cb60-119" aria-hidden="true" tabindex="-1"></a>           <span class="at">sep =</span> <span class="st">"::"</span>,</span>
<span id="cb60-120"><a href="#cb60-120" aria-hidden="true" tabindex="-1"></a>           <span class="at">remove =</span> <span class="cn">FALSE</span>,</span>
<span id="cb60-121"><a href="#cb60-121" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="st">"right"</span></span>
<span id="cb60-122"><a href="#cb60-122" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb60-123"><a href="#cb60-123" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Phylum =</span> <span class="fu">str_remove_all</span>(Phylum, <span class="st">"[a-z]__"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb60-124"><a href="#cb60-124" aria-hidden="true" tabindex="-1"></a>  <span class="co">#  mutate(Genus = str_remove_all(Genus, "[a-z]__")) %&gt;%</span></span>
<span id="cb60-125"><a href="#cb60-125" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Phylum, value) <span class="sc">%&gt;%</span></span>
<span id="cb60-126"><a href="#cb60-126" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Phylum =</span> <span class="fu">case_when</span>(</span>
<span id="cb60-127"><a href="#cb60-127" aria-hidden="true" tabindex="-1"></a>    Phylum <span class="sc">==</span> <span class="st">"Acidobacteria"</span> <span class="sc">~</span> <span class="st">"Acidobacteria"</span>,</span>
<span id="cb60-128"><a href="#cb60-128" aria-hidden="true" tabindex="-1"></a>    Phylum <span class="sc">==</span> <span class="st">"Proteobacteria"</span> <span class="sc">~</span> <span class="st">"Proteobacteria"</span>,</span>
<span id="cb60-129"><a href="#cb60-129" aria-hidden="true" tabindex="-1"></a>    Phylum <span class="sc">==</span> <span class="st">"Actinobacteria"</span> <span class="sc">~</span> <span class="st">"Actinobacteria"</span>,</span>
<span id="cb60-130"><a href="#cb60-130" aria-hidden="true" tabindex="-1"></a>    Phylum <span class="sc">==</span> <span class="st">"Firmicutes"</span> <span class="sc">~</span> <span class="st">"Firmicutes"</span>,</span>
<span id="cb60-131"><a href="#cb60-131" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"Others"</span></span>
<span id="cb60-132"><a href="#cb60-132" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb60-133"><a href="#cb60-133" aria-hidden="true" tabindex="-1"></a>phylum_group <span class="ot">&lt;-</span> <span class="fu">split</span>(phylum_preprocess<span class="sc">$</span>value, phylum_preprocess<span class="sc">$</span>Phylum)</span>
<span id="cb60-134"><a href="#cb60-134" aria-hidden="true" tabindex="-1"></a><span class="co"># match the name of alpha to the nodes of the tree</span></span>
<span id="cb60-135"><a href="#cb60-135" aria-hidden="true" tabindex="-1"></a>nodes_tree <span class="ot">&lt;-</span> <span class="fu">c</span>(tree1<span class="sc">$</span>tip.label, </span>
<span id="cb60-136"><a href="#cb60-136" aria-hidden="true" tabindex="-1"></a>                tree1<span class="sc">$</span>node.label)</span>
<span id="cb60-137"><a href="#cb60-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-138"><a href="#cb60-138" aria-hidden="true" tabindex="-1"></a>tree_1 <span class="ot">&lt;-</span> <span class="fu">groupOTU</span>(tree1, phylum_group)</span>
<span id="cb60-139"><a href="#cb60-139" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree_1, <span class="fu">aes</span>(<span class="at">color =</span> group), <span class="at">layout =</span> <span class="st">"circular"</span>) <span class="sc">+</span></span>
<span id="cb60-140"><a href="#cb60-140" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>, <span class="at">name =</span> <span class="st">"Phylum"</span>)</span>
<span id="cb60-141"><a href="#cb60-141" aria-hidden="true" tabindex="-1"></a>tree</span>
<span id="cb60-142"><a href="#cb60-142" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-143"><a href="#cb60-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-144"><a href="#cb60-144" aria-hidden="true" tabindex="-1"></a><span class="fu">## Pseudo-count + log transformation</span></span>
<span id="cb60-145"><a href="#cb60-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-146"><a href="#cb60-146" aria-hidden="true" tabindex="-1"></a>trac and SLC expect log transformed values as inputs. Since $log(0)$ is undefined we need to add a pseudo-count (e.g. $1$).</span>
<span id="cb60-147"><a href="#cb60-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-150"><a href="#cb60-150" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-151"><a href="#cb60-151" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: log-pseudo</span></span>
<span id="cb60-152"><a href="#cb60-152" aria-hidden="true" tabindex="-1"></a>log_pseudo <span class="ot">&lt;-</span> <span class="cf">function</span>(x, <span class="at">pseudo_count =</span> <span class="dv">1</span>) <span class="fu">log</span>(x <span class="sc">+</span> pseudo_count)</span>
<span id="cb60-153"><a href="#cb60-153" aria-hidden="true" tabindex="-1"></a>z <span class="ot">&lt;-</span> <span class="fu">log_pseudo</span>(soil<span class="sc">@</span>otu_table<span class="sc">@</span>.Data)</span>
<span id="cb60-154"><a href="#cb60-154" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-155"><a href="#cb60-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-156"><a href="#cb60-156" aria-hidden="true" tabindex="-1"></a><span class="fu">## Model fitting</span></span>
<span id="cb60-157"><a href="#cb60-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-158"><a href="#cb60-158" aria-hidden="true" tabindex="-1"></a><span class="fu">### Train-test split</span></span>
<span id="cb60-159"><a href="#cb60-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-160"><a href="#cb60-160" aria-hidden="true" tabindex="-1"></a>First we split the data into a train-test split to evaluate the predictive performance at the end. We'll take 2/3 of the observations for training.</span>
<span id="cb60-161"><a href="#cb60-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-164"><a href="#cb60-164" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-165"><a href="#cb60-165" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: train-test-split</span></span>
<span id="cb60-166"><a href="#cb60-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-167"><a href="#cb60-167" aria-hidden="true" tabindex="-1"></a><span class="co"># extact pH</span></span>
<span id="cb60-168"><a href="#cb60-168" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> soil<span class="sc">@</span>sam_data<span class="sc">$</span>ph</span>
<span id="cb60-169"><a href="#cb60-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-170"><a href="#cb60-170" aria-hidden="true" tabindex="-1"></a><span class="co"># define seed for reproducibility</span></span>
<span id="cb60-171"><a href="#cb60-171" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb60-172"><a href="#cb60-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-173"><a href="#cb60-173" aria-hidden="true" tabindex="-1"></a><span class="co"># train test split</span></span>
<span id="cb60-174"><a href="#cb60-174" aria-hidden="true" tabindex="-1"></a>ntot <span class="ot">&lt;-</span> <span class="fu">length</span>(y)</span>
<span id="cb60-175"><a href="#cb60-175" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fl">0.75</span> <span class="sc">*</span> ntot)</span>
<span id="cb60-176"><a href="#cb60-176" aria-hidden="true" tabindex="-1"></a>tr <span class="ot">&lt;-</span> <span class="fu">sample</span>(ntot, n)</span>
<span id="cb60-177"><a href="#cb60-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-178"><a href="#cb60-178" aria-hidden="true" tabindex="-1"></a><span class="co"># define training and test data</span></span>
<span id="cb60-179"><a href="#cb60-179" aria-hidden="true" tabindex="-1"></a>ytr <span class="ot">&lt;-</span> y[tr]</span>
<span id="cb60-180"><a href="#cb60-180" aria-hidden="true" tabindex="-1"></a>yte <span class="ot">&lt;-</span> y[<span class="sc">-</span>tr]</span>
<span id="cb60-181"><a href="#cb60-181" aria-hidden="true" tabindex="-1"></a>ztr <span class="ot">&lt;-</span> z[tr, ]</span>
<span id="cb60-182"><a href="#cb60-182" aria-hidden="true" tabindex="-1"></a>zte <span class="ot">&lt;-</span> z[<span class="sc">-</span>tr, ]</span>
<span id="cb60-183"><a href="#cb60-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-184"><a href="#cb60-184" aria-hidden="true" tabindex="-1"></a><span class="co"># set the folds for cross-validation</span></span>
<span id="cb60-185"><a href="#cb60-185" aria-hidden="true" tabindex="-1"></a>folds <span class="ot">&lt;-</span> ggb<span class="sc">:::</span><span class="fu">make_folds</span>(<span class="fu">length</span>(ytr), <span class="at">nfolds =</span> <span class="dv">10</span>)</span>
<span id="cb60-186"><a href="#cb60-186" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-187"><a href="#cb60-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-190"><a href="#cb60-190" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-191"><a href="#cb60-191" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: dim-test</span></span>
<span id="cb60-192"><a href="#cb60-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-193"><a href="#cb60-193" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(ztr)</span>
<span id="cb60-194"><a href="#cb60-194" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-195"><a href="#cb60-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-198"><a href="#cb60-198" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-199"><a href="#cb60-199" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: dim-train</span></span>
<span id="cb60-200"><a href="#cb60-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-201"><a href="#cb60-201" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(zte)</span>
<span id="cb60-202"><a href="#cb60-202" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-203"><a href="#cb60-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-204"><a href="#cb60-204" aria-hidden="true" tabindex="-1"></a><span class="fu">### Sparse log-contrast</span></span>
<span id="cb60-205"><a href="#cb60-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-206"><a href="#cb60-206" aria-hidden="true" tabindex="-1"></a>We will fit the model on the train data. The task is to predict pH based on the OTUs available. The method will solve the lambda path. Selecting the optimal tuning parameter $\lambda$ is a crucial.</span>
<span id="cb60-207"><a href="#cb60-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-210"><a href="#cb60-210" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-211"><a href="#cb60-211" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: SLC</span></span>
<span id="cb60-212"><a href="#cb60-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-213"><a href="#cb60-213" aria-hidden="true" tabindex="-1"></a>fit_log_contrast <span class="ot">&lt;-</span> <span class="fu">sparse_log_contrast</span>(<span class="at">Z =</span> ztr, <span class="at">y =</span> ytr)</span>
<span id="cb60-214"><a href="#cb60-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-215"><a href="#cb60-215" aria-hidden="true" tabindex="-1"></a>fit_log_contrast<span class="sc">$</span>beta <span class="sc">%&gt;%</span></span>
<span id="cb60-216"><a href="#cb60-216" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-217"><a href="#cb60-217" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-218"><a href="#cb60-218" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-219"><a href="#cb60-219" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">frac =</span> fit_log_contrast<span class="sc">$</span>fraclist) <span class="sc">%&gt;%</span></span>
<span id="cb60-220"><a href="#cb60-220" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>frac) <span class="sc">%&gt;%</span> </span>
<span id="cb60-221"><a href="#cb60-221" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> frac, <span class="at">y =</span> value, </span>
<span id="cb60-222"><a href="#cb60-222" aria-hidden="true" tabindex="-1"></a>                               <span class="at">group =</span> name, <span class="at">color =</span> name)) <span class="sc">+</span> </span>
<span id="cb60-223"><a href="#cb60-223" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb60-224"><a href="#cb60-224" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb60-225"><a href="#cb60-225" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span> </span>
<span id="cb60-226"><a href="#cb60-226" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">y =</span> <span class="fu">expression</span>(beta), <span class="at">x =</span> <span class="st">"Fraction of lambda_max"</span>)</span>
<span id="cb60-227"><a href="#cb60-227" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-228"><a href="#cb60-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-229"><a href="#cb60-229" aria-hidden="true" tabindex="-1"></a>Cross validate SLC to choose tuning parameter $\lambda$. The function <span class="in">`cv_sparse_log_contrast`</span> can be used to determine the optimal parameter. Dotted line represents the 1SE rule.</span>
<span id="cb60-230"><a href="#cb60-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-233"><a href="#cb60-233" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-234"><a href="#cb60-234" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: SLC-cv</span></span>
<span id="cb60-235"><a href="#cb60-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-236"><a href="#cb60-236" aria-hidden="true" tabindex="-1"></a>cvfit_log_contrast <span class="ot">&lt;-</span> </span>
<span id="cb60-237"><a href="#cb60-237" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cv_sparse_log_contrast</span>(fit_log_contrast, <span class="at">Z =</span> ztr, <span class="at">y =</span> ytr,</span>
<span id="cb60-238"><a href="#cb60-238" aria-hidden="true" tabindex="-1"></a>                         <span class="at">folds =</span> folds)</span>
<span id="cb60-239"><a href="#cb60-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-240"><a href="#cb60-240" aria-hidden="true" tabindex="-1"></a>plot_cv_log_contrast <span class="ot">&lt;-</span> <span class="cf">function</span>(cvfit) {</span>
<span id="cb60-241"><a href="#cb60-241" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract number of non-zero coefficients, mean error and sd</span></span>
<span id="cb60-242"><a href="#cb60-242" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">non_zero =</span> cvfit<span class="sc">$</span>cv<span class="sc">$</span>nonzeros,</span>
<span id="cb60-243"><a href="#cb60-243" aria-hidden="true" tabindex="-1"></a>         <span class="at">error =</span> cvfit<span class="sc">$</span>cv<span class="sc">$</span>m,</span>
<span id="cb60-244"><a href="#cb60-244" aria-hidden="true" tabindex="-1"></a>         <span class="at">sd =</span> cvfit<span class="sc">$</span>cv<span class="sc">$</span>se) <span class="sc">%&gt;%</span></span>
<span id="cb60-245"><a href="#cb60-245" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> non_zero, <span class="at">y =</span> error)) <span class="sc">+</span></span>
<span id="cb60-246"><a href="#cb60-246" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb60-247"><a href="#cb60-247" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 1se line</span></span>
<span id="cb60-248"><a href="#cb60-248" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> cvfit<span class="sc">$</span>cv<span class="sc">$</span>nonzeros[cvfit<span class="sc">$</span>cv<span class="sc">$</span>i1se],</span>
<span id="cb60-249"><a href="#cb60-249" aria-hidden="true" tabindex="-1"></a>               <span class="at">linetype =</span> <span class="st">"longdash"</span>) <span class="sc">+</span></span>
<span id="cb60-250"><a href="#cb60-250" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ibest line</span></span>
<span id="cb60-251"><a href="#cb60-251" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> cvfit<span class="sc">$</span>cv<span class="sc">$</span>nonzeros[cvfit<span class="sc">$</span>cv<span class="sc">$</span>ibest]) <span class="sc">+</span></span>
<span id="cb60-252"><a href="#cb60-252" aria-hidden="true" tabindex="-1"></a>    <span class="co"># line + point and error bars</span></span>
<span id="cb60-253"><a href="#cb60-253" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb60-254"><a href="#cb60-254" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb60-255"><a href="#cb60-255" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> error <span class="sc">-</span> sd, <span class="at">ymax =</span> error <span class="sc">+</span> sd), <span class="at">width =</span> .<span class="dv">2</span>,</span>
<span id="cb60-256"><a href="#cb60-256" aria-hidden="true" tabindex="-1"></a>                  <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="fl">0.05</span>))</span>
<span id="cb60-257"><a href="#cb60-257" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-258"><a href="#cb60-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-259"><a href="#cb60-259" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_cv_log_contrast</span>(cvfit_log_contrast)</span>
<span id="cb60-260"><a href="#cb60-260" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-261"><a href="#cb60-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-262"><a href="#cb60-262" aria-hidden="true" tabindex="-1"></a>Plot the path again, this time with the cross-validation results.</span>
<span id="cb60-263"><a href="#cb60-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-266"><a href="#cb60-266" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-267"><a href="#cb60-267" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: plot-slc-path</span></span>
<span id="cb60-268"><a href="#cb60-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-269"><a href="#cb60-269" aria-hidden="true" tabindex="-1"></a>fit_log_contrast<span class="sc">$</span>beta <span class="sc">%&gt;%</span></span>
<span id="cb60-270"><a href="#cb60-270" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-271"><a href="#cb60-271" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-272"><a href="#cb60-272" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-273"><a href="#cb60-273" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">frac =</span> fit_log_contrast<span class="sc">$</span>fraclist) <span class="sc">%&gt;%</span></span>
<span id="cb60-274"><a href="#cb60-274" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>frac) <span class="sc">%&gt;%</span> </span>
<span id="cb60-275"><a href="#cb60-275" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> frac, <span class="at">y =</span> value, </span>
<span id="cb60-276"><a href="#cb60-276" aria-hidden="true" tabindex="-1"></a>                               <span class="at">group =</span> name, <span class="at">color =</span> name)) <span class="sc">+</span> </span>
<span id="cb60-277"><a href="#cb60-277" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb60-278"><a href="#cb60-278" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb60-279"><a href="#cb60-279" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span> </span>
<span id="cb60-280"><a href="#cb60-280" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">y =</span> <span class="fu">expression</span>(beta), <span class="at">x =</span> <span class="st">"Fraction of lambda_max"</span>) <span class="sc">+</span></span>
<span id="cb60-281"><a href="#cb60-281" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> fit_log_contrast<span class="sc">$</span>fraclist[cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se],</span>
<span id="cb60-282"><a href="#cb60-282" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"longdash"</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb60-283"><a href="#cb60-283" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> fit_log_contrast<span class="sc">$</span>fraclist[cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>ibest],</span>
<span id="cb60-284"><a href="#cb60-284" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha =</span> <span class="fl">0.5</span>)</span>
<span id="cb60-285"><a href="#cb60-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-286"><a href="#cb60-286" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-287"><a href="#cb60-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-288"><a href="#cb60-288" aria-hidden="true" tabindex="-1"></a><span class="fu">### trac</span></span>
<span id="cb60-289"><a href="#cb60-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-290"><a href="#cb60-290" aria-hidden="true" tabindex="-1"></a>Let's continue with the trac model using <span class="in">`trac`</span>.</span>
<span id="cb60-291"><a href="#cb60-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-294"><a href="#cb60-294" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-295"><a href="#cb60-295" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: trac</span></span>
<span id="cb60-296"><a href="#cb60-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-297"><a href="#cb60-297" aria-hidden="true" tabindex="-1"></a>fit_trac <span class="ot">&lt;-</span> <span class="fu">trac</span>(<span class="at">Z =</span> ztr, <span class="at">y =</span> ytr, <span class="at">A =</span> A)</span>
<span id="cb60-298"><a href="#cb60-298" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-299"><a href="#cb60-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-300"><a href="#cb60-300" aria-hidden="true" tabindex="-1"></a>Cross validate trac to choose tuning parameter $\lambda$. The function <span class="in">`cv_trac`</span> can be used to determine the optimal parameter. Dotted line represents the 1SE rule.</span>
<span id="cb60-301"><a href="#cb60-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-304"><a href="#cb60-304" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-305"><a href="#cb60-305" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: trac-cv</span></span>
<span id="cb60-306"><a href="#cb60-306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-307"><a href="#cb60-307" aria-hidden="true" tabindex="-1"></a>cvfit_trac <span class="ot">&lt;-</span> </span>
<span id="cb60-308"><a href="#cb60-308" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cv_trac</span>(fit_trac, <span class="at">Z =</span> ztr, <span class="at">y =</span> ytr, <span class="at">A =</span> A,</span>
<span id="cb60-309"><a href="#cb60-309" aria-hidden="true" tabindex="-1"></a>                         <span class="at">folds =</span> folds)</span>
<span id="cb60-310"><a href="#cb60-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-311"><a href="#cb60-311" aria-hidden="true" tabindex="-1"></a>plot_cv_trac <span class="ot">&lt;-</span> <span class="cf">function</span>(cvfit) {</span>
<span id="cb60-312"><a href="#cb60-312" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract number of non-zero coefficients, mean error and sd</span></span>
<span id="cb60-313"><a href="#cb60-313" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">non_zero =</span> cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>nonzeros,</span>
<span id="cb60-314"><a href="#cb60-314" aria-hidden="true" tabindex="-1"></a>         <span class="at">error =</span> cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>m,</span>
<span id="cb60-315"><a href="#cb60-315" aria-hidden="true" tabindex="-1"></a>         <span class="at">sd =</span> cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>se) <span class="sc">%&gt;%</span></span>
<span id="cb60-316"><a href="#cb60-316" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> non_zero, <span class="at">y =</span> error)) <span class="sc">+</span></span>
<span id="cb60-317"><a href="#cb60-317" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb60-318"><a href="#cb60-318" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 1se line</span></span>
<span id="cb60-319"><a href="#cb60-319" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>nonzeros[cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se],</span>
<span id="cb60-320"><a href="#cb60-320" aria-hidden="true" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">"gray"</span>, <span class="at">linetype =</span> <span class="st">"longdash"</span>) <span class="sc">+</span></span>
<span id="cb60-321"><a href="#cb60-321" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ibest line</span></span>
<span id="cb60-322"><a href="#cb60-322" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>nonzeros[cvfit<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest],</span>
<span id="cb60-323"><a href="#cb60-323" aria-hidden="true" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">"gray50"</span>, <span class="at">linetype =</span> <span class="st">"longdash"</span>) <span class="sc">+</span></span>
<span id="cb60-324"><a href="#cb60-324" aria-hidden="true" tabindex="-1"></a>    <span class="co"># line + point and error bars</span></span>
<span id="cb60-325"><a href="#cb60-325" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb60-326"><a href="#cb60-326" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb60-327"><a href="#cb60-327" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> error <span class="sc">-</span> sd, <span class="at">ymax =</span> error <span class="sc">+</span> sd), <span class="at">width =</span> .<span class="dv">2</span>,</span>
<span id="cb60-328"><a href="#cb60-328" aria-hidden="true" tabindex="-1"></a>                  <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="fl">0.05</span>))</span>
<span id="cb60-329"><a href="#cb60-329" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-330"><a href="#cb60-330" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-331"><a href="#cb60-331" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_cv_trac</span>(cvfit_trac)</span>
<span id="cb60-332"><a href="#cb60-332" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-333"><a href="#cb60-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-334"><a href="#cb60-334" aria-hidden="true" tabindex="-1"></a>Plot the coefficient path.</span>
<span id="cb60-335"><a href="#cb60-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-338"><a href="#cb60-338" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-339"><a href="#cb60-339" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: trac-path</span></span>
<span id="cb60-340"><a href="#cb60-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-341"><a href="#cb60-341" aria-hidden="true" tabindex="-1"></a>fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha <span class="sc">%&gt;%</span></span>
<span id="cb60-342"><a href="#cb60-342" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-343"><a href="#cb60-343" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-344"><a href="#cb60-344" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-345"><a href="#cb60-345" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">frac =</span> fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>fraclist) <span class="sc">%&gt;%</span></span>
<span id="cb60-346"><a href="#cb60-346" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>frac) <span class="sc">%&gt;%</span> </span>
<span id="cb60-347"><a href="#cb60-347" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> frac, <span class="at">y =</span> value, </span>
<span id="cb60-348"><a href="#cb60-348" aria-hidden="true" tabindex="-1"></a>                               <span class="at">group =</span> name, <span class="at">color =</span> name)) <span class="sc">+</span> </span>
<span id="cb60-349"><a href="#cb60-349" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb60-350"><a href="#cb60-350" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb60-351"><a href="#cb60-351" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span> </span>
<span id="cb60-352"><a href="#cb60-352" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">y =</span> <span class="fu">expression</span>(alpha), <span class="at">x =</span> <span class="st">"Fraction of lambda_max"</span>) <span class="sc">+</span></span>
<span id="cb60-353"><a href="#cb60-353" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>fraclist[cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se],</span>
<span id="cb60-354"><a href="#cb60-354" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"longdash"</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb60-355"><a href="#cb60-355" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>fraclist[cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest],</span>
<span id="cb60-356"><a href="#cb60-356" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha =</span> <span class="fl">0.5</span>)</span>
<span id="cb60-357"><a href="#cb60-357" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-358"><a href="#cb60-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-359"><a href="#cb60-359" aria-hidden="true" tabindex="-1"></a><span class="fu">### Weighted trac</span></span>
<span id="cb60-360"><a href="#cb60-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-361"><a href="#cb60-361" aria-hidden="true" tabindex="-1"></a>The path of trac with normal weights seems interesting. It might be better to use the weighted trac, nudging the model to select on a lower taxonomic level.</span>
<span id="cb60-362"><a href="#cb60-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-365"><a href="#cb60-365" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-366"><a href="#cb60-366" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: weighted-trac</span></span>
<span id="cb60-367"><a href="#cb60-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-368"><a href="#cb60-368" aria-hidden="true" tabindex="-1"></a>fit_weighted_trac <span class="ot">&lt;-</span> <span class="fu">trac</span>(<span class="at">Z =</span> ztr, <span class="at">y =</span> ytr, <span class="at">A =</span> A, <span class="at">w =</span> <span class="fu">colSums</span>(<span class="fu">as.matrix</span>(A))<span class="sc">^</span><span class="fl">0.5</span>)</span>
<span id="cb60-369"><a href="#cb60-369" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-370"><a href="#cb60-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-373"><a href="#cb60-373" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-374"><a href="#cb60-374" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: weighted-trac-cv</span></span>
<span id="cb60-375"><a href="#cb60-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-376"><a href="#cb60-376" aria-hidden="true" tabindex="-1"></a>cvfit_weighted_trac <span class="ot">&lt;-</span> </span>
<span id="cb60-377"><a href="#cb60-377" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cv_trac</span>(fit_weighted_trac, <span class="at">Z =</span> ztr, <span class="at">y =</span> ytr, <span class="at">A =</span> A,</span>
<span id="cb60-378"><a href="#cb60-378" aria-hidden="true" tabindex="-1"></a>          <span class="at">folds =</span> folds)</span>
<span id="cb60-379"><a href="#cb60-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-380"><a href="#cb60-380" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_cv_trac</span>(cvfit_weighted_trac)</span>
<span id="cb60-381"><a href="#cb60-381" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-382"><a href="#cb60-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-385"><a href="#cb60-385" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-386"><a href="#cb60-386" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: weighted-trac-path</span></span>
<span id="cb60-387"><a href="#cb60-387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-388"><a href="#cb60-388" aria-hidden="true" tabindex="-1"></a>fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha <span class="sc">%&gt;%</span></span>
<span id="cb60-389"><a href="#cb60-389" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-390"><a href="#cb60-390" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-391"><a href="#cb60-391" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-392"><a href="#cb60-392" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">frac =</span> fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>fraclist) <span class="sc">%&gt;%</span></span>
<span id="cb60-393"><a href="#cb60-393" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>frac) <span class="sc">%&gt;%</span> </span>
<span id="cb60-394"><a href="#cb60-394" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> frac, <span class="at">y =</span> value, </span>
<span id="cb60-395"><a href="#cb60-395" aria-hidden="true" tabindex="-1"></a>                               <span class="at">group =</span> name, <span class="at">color =</span> name)) <span class="sc">+</span> </span>
<span id="cb60-396"><a href="#cb60-396" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb60-397"><a href="#cb60-397" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb60-398"><a href="#cb60-398" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span> </span>
<span id="cb60-399"><a href="#cb60-399" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">y =</span> <span class="fu">expression</span>(alpha), <span class="at">x =</span> <span class="st">"Fraction of lambda_max"</span>) <span class="sc">+</span></span>
<span id="cb60-400"><a href="#cb60-400" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>fraclist[cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se],</span>
<span id="cb60-401"><a href="#cb60-401" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">"longdash"</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb60-402"><a href="#cb60-402" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>fraclist[cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest],</span>
<span id="cb60-403"><a href="#cb60-403" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha =</span> <span class="fl">0.5</span>)</span>
<span id="cb60-404"><a href="#cb60-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-405"><a href="#cb60-405" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-406"><a href="#cb60-406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-407"><a href="#cb60-407" aria-hidden="true" tabindex="-1"></a><span class="fu">## Check prediction</span></span>
<span id="cb60-408"><a href="#cb60-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-409"><a href="#cb60-409" aria-hidden="true" tabindex="-1"></a>Let's evaluate the different models on the test set and see how they perform based on the RMSE.</span>
<span id="cb60-410"><a href="#cb60-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-413"><a href="#cb60-413" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-414"><a href="#cb60-414" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: predictions</span></span>
<span id="cb60-415"><a href="#cb60-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-416"><a href="#cb60-416" aria-hidden="true" tabindex="-1"></a>yhat_slc <span class="ot">&lt;-</span> <span class="fu">predict_sparse_log_contrast</span>(fit_log_contrast, <span class="at">new_Z =</span> zte)</span>
<span id="cb60-417"><a href="#cb60-417" aria-hidden="true" tabindex="-1"></a>yhat_trac <span class="ot">&lt;-</span> <span class="fu">predict_trac</span>(fit_trac, <span class="at">new_Z =</span> zte)</span>
<span id="cb60-418"><a href="#cb60-418" aria-hidden="true" tabindex="-1"></a>yhat_weighted_trac <span class="ot">&lt;-</span> <span class="fu">predict_trac</span>(fit_weighted_trac, <span class="at">new_Z =</span> zte)</span>
<span id="cb60-419"><a href="#cb60-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-420"><a href="#cb60-420" aria-hidden="true" tabindex="-1"></a>yhat_slc_1se <span class="ot">&lt;-</span> yhat_slc[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se]</span>
<span id="cb60-421"><a href="#cb60-421" aria-hidden="true" tabindex="-1"></a>yhat_slc_best <span class="ot">&lt;-</span> yhat_slc[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>ibest]</span>
<span id="cb60-422"><a href="#cb60-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-423"><a href="#cb60-423" aria-hidden="true" tabindex="-1"></a>yhat_trac_1se <span class="ot">&lt;-</span> yhat_trac[[<span class="dv">1</span>]][, cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se]</span>
<span id="cb60-424"><a href="#cb60-424" aria-hidden="true" tabindex="-1"></a>yhat_trac_best <span class="ot">&lt;-</span> yhat_trac[[<span class="dv">1</span>]][, cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest]</span>
<span id="cb60-425"><a href="#cb60-425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-426"><a href="#cb60-426" aria-hidden="true" tabindex="-1"></a>yhat_weighted_trac_1se <span class="ot">&lt;-</span> yhat_weighted_trac[[<span class="dv">1</span>]][, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se]</span>
<span id="cb60-427"><a href="#cb60-427" aria-hidden="true" tabindex="-1"></a>yhat_weighted_trac_best <span class="ot">&lt;-</span> yhat_weighted_trac[[<span class="dv">1</span>]][, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest]</span>
<span id="cb60-428"><a href="#cb60-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-429"><a href="#cb60-429" aria-hidden="true" tabindex="-1"></a>method_col <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"SLC (1SE)"</span>,  <span class="st">"SLC (best)"</span>, <span class="st">"trac (1SE)"</span>, <span class="st">"trac (best)"</span>,</span>
<span id="cb60-430"><a href="#cb60-430" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Weighted trac (1SE)"</span>, <span class="st">"Weighted trac (best)"</span>)</span>
<span id="cb60-431"><a href="#cb60-431" aria-hidden="true" tabindex="-1"></a>RMSE <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">sqrt</span>(<span class="fu">mean</span>((yte <span class="sc">-</span> yhat_slc_1se)<span class="sc">^</span><span class="dv">2</span>)),</span>
<span id="cb60-432"><a href="#cb60-432" aria-hidden="true" tabindex="-1"></a>          <span class="fu">sqrt</span>(<span class="fu">mean</span>((yte <span class="sc">-</span> yhat_slc_best)<span class="sc">^</span><span class="dv">2</span>)),</span>
<span id="cb60-433"><a href="#cb60-433" aria-hidden="true" tabindex="-1"></a>          <span class="fu">sqrt</span>(<span class="fu">mean</span>((yte <span class="sc">-</span> yhat_trac_1se)<span class="sc">^</span><span class="dv">2</span>)),</span>
<span id="cb60-434"><a href="#cb60-434" aria-hidden="true" tabindex="-1"></a>          <span class="fu">sqrt</span>(<span class="fu">mean</span>((yte <span class="sc">-</span> yhat_trac_best)<span class="sc">^</span><span class="dv">2</span>)),</span>
<span id="cb60-435"><a href="#cb60-435" aria-hidden="true" tabindex="-1"></a>          <span class="fu">sqrt</span>(<span class="fu">mean</span>((yte <span class="sc">-</span> yhat_weighted_trac_1se)<span class="sc">^</span><span class="dv">2</span>)),</span>
<span id="cb60-436"><a href="#cb60-436" aria-hidden="true" tabindex="-1"></a>          <span class="fu">sqrt</span>(<span class="fu">mean</span>((yte <span class="sc">-</span> yhat_weighted_trac_best)<span class="sc">^</span><span class="dv">2</span>)))</span>
<span id="cb60-437"><a href="#cb60-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-438"><a href="#cb60-438" aria-hidden="true" tabindex="-1"></a>NZ <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">sum</span>(fit_log_contrast<span class="sc">$</span>beta[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se] <span class="sc">!=</span> <span class="dv">0</span>),</span>
<span id="cb60-439"><a href="#cb60-439" aria-hidden="true" tabindex="-1"></a>        <span class="fu">sum</span>(fit_log_contrast<span class="sc">$</span>beta[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>ibest] <span class="sc">!=</span> <span class="dv">0</span>),</span>
<span id="cb60-440"><a href="#cb60-440" aria-hidden="true" tabindex="-1"></a>        <span class="fu">sum</span>(fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">!=</span> <span class="dv">0</span>),</span>
<span id="cb60-441"><a href="#cb60-441" aria-hidden="true" tabindex="-1"></a>        <span class="fu">sum</span>(fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest] <span class="sc">!=</span> <span class="dv">0</span>),</span>
<span id="cb60-442"><a href="#cb60-442" aria-hidden="true" tabindex="-1"></a>        <span class="fu">sum</span>(fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">!=</span> <span class="dv">0</span>),</span>
<span id="cb60-443"><a href="#cb60-443" aria-hidden="true" tabindex="-1"></a>        <span class="fu">sum</span>(fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest] <span class="sc">!=</span> <span class="dv">0</span>))</span>
<span id="cb60-444"><a href="#cb60-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-445"><a href="#cb60-445" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(method_col, NZ, RMSE) <span class="sc">%&gt;%</span></span>
<span id="cb60-446"><a href="#cb60-446" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">Method =</span> method_col) <span class="sc">%&gt;%</span></span>
<span id="cb60-447"><a href="#cb60-447" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">RMSE =</span> <span class="fu">round</span>(RMSE, <span class="at">digits =</span> <span class="dv">3</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb60-448"><a href="#cb60-448" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kbl</span>()</span>
<span id="cb60-449"><a href="#cb60-449" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-450"><a href="#cb60-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-451"><a href="#cb60-451" aria-hidden="true" tabindex="-1"></a>Let's plot the out of sample prediction vs the ground truth.</span>
<span id="cb60-452"><a href="#cb60-452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-455"><a href="#cb60-455" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-456"><a href="#cb60-456" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: out-of-sample-plot</span></span>
<span id="cb60-457"><a href="#cb60-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-458"><a href="#cb60-458" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(<span class="at">yte =</span> yte,</span>
<span id="cb60-459"><a href="#cb60-459" aria-hidden="true" tabindex="-1"></a>           <span class="st">"SLC (1se)"</span> <span class="ot">=</span> yhat_slc_1se,</span>
<span id="cb60-460"><a href="#cb60-460" aria-hidden="true" tabindex="-1"></a>           <span class="st">"SLC (best)"</span> <span class="ot">=</span> yhat_slc_best,</span>
<span id="cb60-461"><a href="#cb60-461" aria-hidden="true" tabindex="-1"></a>           <span class="st">"trac (1se)"</span> <span class="ot">=</span> yhat_trac_1se,</span>
<span id="cb60-462"><a href="#cb60-462" aria-hidden="true" tabindex="-1"></a>           <span class="st">"trac (best)"</span> <span class="ot">=</span> yhat_trac_best,</span>
<span id="cb60-463"><a href="#cb60-463" aria-hidden="true" tabindex="-1"></a>           <span class="st">"Weighted trac (1se)"</span> <span class="ot">=</span> yhat_weighted_trac_1se,</span>
<span id="cb60-464"><a href="#cb60-464" aria-hidden="true" tabindex="-1"></a>           <span class="st">"Weighted trac (best)"</span> <span class="ot">=</span> yhat_weighted_trac_best) <span class="sc">%&gt;%</span></span>
<span id="cb60-465"><a href="#cb60-465" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>yte) <span class="sc">%&gt;%</span></span>
<span id="cb60-466"><a href="#cb60-466" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> yte, <span class="at">y =</span> value)) <span class="sc">+</span></span>
<span id="cb60-467"><a href="#cb60-467" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb60-468"><a href="#cb60-468" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb60-469"><a href="#cb60-469" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>() <span class="sc">+</span></span>
<span id="cb60-470"><a href="#cb60-470" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>name, <span class="at">ncol =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb60-471"><a href="#cb60-471" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Predicted pH"</span>) <span class="sc">+</span></span>
<span id="cb60-472"><a href="#cb60-472" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"True pH"</span>)</span>
<span id="cb60-473"><a href="#cb60-473" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-474"><a href="#cb60-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-475"><a href="#cb60-475" aria-hidden="true" tabindex="-1"></a>Finally, plot out of sample error vs the different number of non-zero coefficients. for SLC, trac, and weighted trac.</span>
<span id="cb60-476"><a href="#cb60-476" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-479"><a href="#cb60-479" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-480"><a href="#cb60-480" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: plot-slc-sparsity-vs-mse</span></span>
<span id="cb60-481"><a href="#cb60-481" aria-hidden="true" tabindex="-1"></a>plot_nzz_mse <span class="ot">&lt;-</span> <span class="cf">function</span>(yhat, y, coefs, i1se, ibest) {</span>
<span id="cb60-482"><a href="#cb60-482" aria-hidden="true" tabindex="-1"></a>  mse <span class="ot">&lt;-</span> <span class="fu">colMeans</span>((yhat <span class="sc">-</span> y)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb60-483"><a href="#cb60-483" aria-hidden="true" tabindex="-1"></a>  nnz <span class="ot">&lt;-</span> <span class="fu">colSums</span>(coefs <span class="sc">!=</span> <span class="dv">0</span>)</span>
<span id="cb60-484"><a href="#cb60-484" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(mse, nnz) <span class="sc">%&gt;%</span></span>
<span id="cb60-485"><a href="#cb60-485" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> nnz, <span class="at">y =</span> mse)) <span class="sc">+</span></span>
<span id="cb60-486"><a href="#cb60-486" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb60-487"><a href="#cb60-487" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> nnz[i1se], <span class="at">linetype =</span> <span class="st">"longdash"</span>) <span class="sc">+</span> </span>
<span id="cb60-488"><a href="#cb60-488" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> nnz[ibest])</span>
<span id="cb60-489"><a href="#cb60-489" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-490"><a href="#cb60-490" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-491"><a href="#cb60-491" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_nzz_mse</span>(<span class="at">yhat =</span> yhat_slc, <span class="at">y =</span> yte, <span class="at">coefs =</span> fit_log_contrast<span class="sc">$</span>beta,</span>
<span id="cb60-492"><a href="#cb60-492" aria-hidden="true" tabindex="-1"></a>             <span class="at">i1se =</span> cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se,</span>
<span id="cb60-493"><a href="#cb60-493" aria-hidden="true" tabindex="-1"></a>             <span class="at">ibest =</span> cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>ibest)</span>
<span id="cb60-494"><a href="#cb60-494" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-495"><a href="#cb60-495" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-498"><a href="#cb60-498" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-499"><a href="#cb60-499" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: plot-trac-sparsity-vs-mse</span></span>
<span id="cb60-500"><a href="#cb60-500" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_nzz_mse</span>(<span class="at">yhat =</span> yhat_trac[[<span class="dv">1</span>]], <span class="at">y =</span> yte, <span class="at">coefs =</span> fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha,</span>
<span id="cb60-501"><a href="#cb60-501" aria-hidden="true" tabindex="-1"></a>             <span class="at">i1se =</span> cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se,</span>
<span id="cb60-502"><a href="#cb60-502" aria-hidden="true" tabindex="-1"></a>             <span class="at">ibest =</span> cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest)</span>
<span id="cb60-503"><a href="#cb60-503" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-504"><a href="#cb60-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-507"><a href="#cb60-507" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-508"><a href="#cb60-508" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: plot-weighted-trac-sparsity-vs-mse</span></span>
<span id="cb60-509"><a href="#cb60-509" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_nzz_mse</span>(<span class="at">yhat =</span> yhat_weighted_trac[[<span class="dv">1</span>]], </span>
<span id="cb60-510"><a href="#cb60-510" aria-hidden="true" tabindex="-1"></a>             <span class="at">y =</span> yte, <span class="at">coefs =</span> fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha,</span>
<span id="cb60-511"><a href="#cb60-511" aria-hidden="true" tabindex="-1"></a>             <span class="at">i1se =</span> cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se,</span>
<span id="cb60-512"><a href="#cb60-512" aria-hidden="true" tabindex="-1"></a>             <span class="at">ibest =</span> cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>ibest)</span>
<span id="cb60-513"><a href="#cb60-513" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-514"><a href="#cb60-514" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-515"><a href="#cb60-515" aria-hidden="true" tabindex="-1"></a><span class="fu">## Plot coefficients and trac results on a tree</span></span>
<span id="cb60-516"><a href="#cb60-516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-517"><a href="#cb60-517" aria-hidden="true" tabindex="-1"></a><span class="fu">### SLC</span></span>
<span id="cb60-518"><a href="#cb60-518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-519"><a href="#cb60-519" aria-hidden="true" tabindex="-1"></a>Start with the $\beta$ coefficients of the 1SE solution. The first plot shows the last known taxonomic annotation for the corresponding OTU.</span>
<span id="cb60-520"><a href="#cb60-520" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-523"><a href="#cb60-523" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-524"><a href="#cb60-524" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: plot-slc-coef-last-known</span></span>
<span id="cb60-525"><a href="#cb60-525" aria-hidden="true" tabindex="-1"></a><span class="co"># get which coefficients are not zero</span></span>
<span id="cb60-526"><a href="#cb60-526" aria-hidden="true" tabindex="-1"></a>nzz <span class="ot">&lt;-</span> <span class="cf">function</span>(coefs) {</span>
<span id="cb60-527"><a href="#cb60-527" aria-hidden="true" tabindex="-1"></a>  coefs[coefs <span class="sc">!=</span> <span class="dv">0</span>]</span>
<span id="cb60-528"><a href="#cb60-528" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-529"><a href="#cb60-529" aria-hidden="true" tabindex="-1"></a><span class="co"># for slc 1se</span></span>
<span id="cb60-530"><a href="#cb60-530" aria-hidden="true" tabindex="-1"></a>slc_names <span class="ot">&lt;-</span> fit_log_contrast<span class="sc">$</span>beta[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb60-531"><a href="#cb60-531" aria-hidden="true" tabindex="-1"></a>  nzz </span>
<span id="cb60-532"><a href="#cb60-532" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-533"><a href="#cb60-533" aria-hidden="true" tabindex="-1"></a><span class="co"># get the last known name for slc</span></span>
<span id="cb60-534"><a href="#cb60-534" aria-hidden="true" tabindex="-1"></a>slc_names_last_known <span class="ot">&lt;-</span> soil<span class="sc">@</span>tax_table<span class="sc">@</span>.Data <span class="sc">%&gt;%</span></span>
<span id="cb60-535"><a href="#cb60-535" aria-hidden="true" tabindex="-1"></a>  as.data.frame <span class="sc">%&gt;%</span></span>
<span id="cb60-536"><a href="#cb60-536" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>() <span class="sc">%&gt;%</span></span>
<span id="cb60-537"><a href="#cb60-537" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(rowname <span class="sc">%in%</span> <span class="fu">names</span>(slc_names))</span>
<span id="cb60-538"><a href="#cb60-538" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-539"><a href="#cb60-539" aria-hidden="true" tabindex="-1"></a>last_known_tax <span class="ot">&lt;-</span> <span class="fu">apply</span>(slc_names_last_known, <span class="at">MARGIN =</span> <span class="dv">2</span>, </span>
<span id="cb60-540"><a href="#cb60-540" aria-hidden="true" tabindex="-1"></a>      <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">str_length</span>(x) <span class="sc">==</span> <span class="dv">3</span>) </span>
<span id="cb60-541"><a href="#cb60-541" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-542"><a href="#cb60-542" aria-hidden="true" tabindex="-1"></a>last_known_slc <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb60-543"><a href="#cb60-543" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(slc_names_last_known)) {</span>
<span id="cb60-544"><a href="#cb60-544" aria-hidden="true" tabindex="-1"></a>  last_known_slc <span class="ot">&lt;-</span> <span class="fu">append</span>(last_known_slc,</span>
<span id="cb60-545"><a href="#cb60-545" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(slc_names_last_known[i, (<span class="dv">8</span> <span class="sc">-</span> <span class="fu">rowSums</span>(last_known_tax)[i])], <span class="st">"::"</span>,</span>
<span id="cb60-546"><a href="#cb60-546" aria-hidden="true" tabindex="-1"></a>         slc_names_last_known[i, <span class="dv">1</span>]))</span>
<span id="cb60-547"><a href="#cb60-547" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-548"><a href="#cb60-548" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-549"><a href="#cb60-549" aria-hidden="true" tabindex="-1"></a>last_known_slc <span class="ot">&lt;-</span> </span>
<span id="cb60-550"><a href="#cb60-550" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(<span class="at">name =</span> slc_names_last_known<span class="sc">$</span>rowname, <span class="at">last_name =</span> last_known_slc)</span>
<span id="cb60-551"><a href="#cb60-551" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-552"><a href="#cb60-552" aria-hidden="true" tabindex="-1"></a><span class="co"># start plotting</span></span>
<span id="cb60-553"><a href="#cb60-553" aria-hidden="true" tabindex="-1"></a>fit_log_contrast<span class="sc">$</span>beta[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb60-554"><a href="#cb60-554" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb60-555"><a href="#cb60-555" aria-hidden="true" tabindex="-1"></a>  <span class="fu">enframe</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-556"><a href="#cb60-556" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(last_known_slc) <span class="sc">%&gt;%</span></span>
<span id="cb60-557"><a href="#cb60-557" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">fct_reorder</span>(last_name, value), <span class="at">y =</span> value)) <span class="sc">+</span> </span>
<span id="cb60-558"><a href="#cb60-558" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb60-559"><a href="#cb60-559" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">fct_reorder</span>(last_name, value), </span>
<span id="cb60-560"><a href="#cb60-560" aria-hidden="true" tabindex="-1"></a>                   <span class="at">xend=</span> <span class="fu">fct_reorder</span>(last_name, value),</span>
<span id="cb60-561"><a href="#cb60-561" aria-hidden="true" tabindex="-1"></a>                   <span class="at">y=</span><span class="dv">0</span>, <span class="at">yend=</span>value)) <span class="sc">+</span></span>
<span id="cb60-562"><a href="#cb60-562" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()<span class="sc">+</span> </span>
<span id="cb60-563"><a href="#cb60-563" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb60-564"><a href="#cb60-564" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">label=</span><span class="cf">function</span>(x) <span class="fu">str_split_i</span>(x, <span class="at">pattern =</span> <span class="st">"::"</span>, <span class="at">i =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb60-565"><a href="#cb60-565" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"OTUs (as lowest known taxonomic rank)"</span>) <span class="sc">+</span></span>
<span id="cb60-566"><a href="#cb60-566" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(beta))</span>
<span id="cb60-567"><a href="#cb60-567" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-568"><a href="#cb60-568" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-569"><a href="#cb60-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-570"><a href="#cb60-570" aria-hidden="true" tabindex="-1"></a>Once again only with the the OTU ids.</span>
<span id="cb60-571"><a href="#cb60-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-574"><a href="#cb60-574" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-575"><a href="#cb60-575" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: plot-slc-coef-otu-name</span></span>
<span id="cb60-576"><a href="#cb60-576" aria-hidden="true" tabindex="-1"></a>fit_log_contrast<span class="sc">$</span>beta[, cvfit_log_contrast<span class="sc">$</span>cv<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb60-577"><a href="#cb60-577" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb60-578"><a href="#cb60-578" aria-hidden="true" tabindex="-1"></a>  <span class="fu">enframe</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb60-579"><a href="#cb60-579" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">fct_reorder</span>(name, value), <span class="at">y =</span> value)) <span class="sc">+</span> </span>
<span id="cb60-580"><a href="#cb60-580" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb60-581"><a href="#cb60-581" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">fct_reorder</span>(name, value), </span>
<span id="cb60-582"><a href="#cb60-582" aria-hidden="true" tabindex="-1"></a>                   <span class="at">xend=</span> <span class="fu">fct_reorder</span>(name, value),</span>
<span id="cb60-583"><a href="#cb60-583" aria-hidden="true" tabindex="-1"></a>                   <span class="at">y=</span><span class="dv">0</span>, <span class="at">yend=</span>value)) <span class="sc">+</span></span>
<span id="cb60-584"><a href="#cb60-584" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()<span class="sc">+</span> </span>
<span id="cb60-585"><a href="#cb60-585" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb60-586"><a href="#cb60-586" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">label=</span><span class="cf">function</span>(x) <span class="fu">str_split_i</span>(x, <span class="at">pattern =</span> <span class="st">"::"</span>, <span class="at">i =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb60-587"><a href="#cb60-587" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"OTUs"</span>) <span class="sc">+</span></span>
<span id="cb60-588"><a href="#cb60-588" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(beta))</span>
<span id="cb60-589"><a href="#cb60-589" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-590"><a href="#cb60-590" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-591"><a href="#cb60-591" aria-hidden="true" tabindex="-1"></a>Sparse log-contrast tree</span>
<span id="cb60-592"><a href="#cb60-592" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-595"><a href="#cb60-595" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-596"><a href="#cb60-596" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: plot-slc-coef-tree</span></span>
<span id="cb60-597"><a href="#cb60-597" aria-hidden="true" tabindex="-1"></a>slc_selected <span class="ot">&lt;-</span> tax <span class="sc">%&gt;%</span></span>
<span id="cb60-598"><a href="#cb60-598" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>() <span class="sc">%&gt;%</span></span>
<span id="cb60-599"><a href="#cb60-599" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(rowname <span class="sc">%in%</span> <span class="fu">names</span>(slc_names)) <span class="sc">%&gt;%</span></span>
<span id="cb60-600"><a href="#cb60-600" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(OTU) <span class="sc">%&gt;%</span></span>
<span id="cb60-601"><a href="#cb60-601" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">OTU =</span> <span class="fu">as.character</span>(OTU),</span>
<span id="cb60-602"><a href="#cb60-602" aria-hidden="true" tabindex="-1"></a>         <span class="at">SLC =</span> <span class="cn">TRUE</span>)</span>
<span id="cb60-603"><a href="#cb60-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-604"><a href="#cb60-604" aria-hidden="true" tabindex="-1"></a>tree_slc <span class="ot">&lt;-</span> tree <span class="sc">%&lt;+%</span> slc_selected</span>
<span id="cb60-605"><a href="#cb60-605" aria-hidden="true" tabindex="-1"></a>tree_slc <span class="sc">+</span></span>
<span id="cb60-606"><a href="#cb60-606" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point2</span>(<span class="fu">aes</span>(<span class="at">subset =</span> SLC <span class="sc">==</span> <span class="cn">TRUE</span>), <span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>)</span>
<span id="cb60-607"><a href="#cb60-607" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-608"><a href="#cb60-608" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-609"><a href="#cb60-609" aria-hidden="true" tabindex="-1"></a><span class="fu">### trac</span></span>
<span id="cb60-610"><a href="#cb60-610" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-611"><a href="#cb60-611" aria-hidden="true" tabindex="-1"></a>A possible reason for the behavior of trac could be because of the filter (OTUs present at least $100$ times). The included OTUs could be biased towards a specific Phylum.</span>
<span id="cb60-612"><a href="#cb60-612" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-615"><a href="#cb60-615" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-616"><a href="#cb60-616" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: trac-coef</span></span>
<span id="cb60-617"><a href="#cb60-617" aria-hidden="true" tabindex="-1"></a>fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb60-618"><a href="#cb60-618" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb60-619"><a href="#cb60-619" aria-hidden="true" tabindex="-1"></a>  <span class="fu">enframe</span>() <span class="sc">%&gt;%</span></span>
<span id="cb60-620"><a href="#cb60-620" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(value)) <span class="sc">%&gt;%</span></span>
<span id="cb60-621"><a href="#cb60-621" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kbl</span>()</span>
<span id="cb60-622"><a href="#cb60-622" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-623"><a href="#cb60-623" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-626"><a href="#cb60-626" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-627"><a href="#cb60-627" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: plot-trac-tree</span></span>
<span id="cb60-628"><a href="#cb60-628" aria-hidden="true" tabindex="-1"></a>tree_df <span class="ot">&lt;-</span> <span class="fu">enframe</span>(nodes_tree) <span class="sc">%&gt;%</span></span>
<span id="cb60-629"><a href="#cb60-629" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>name)</span>
<span id="cb60-630"><a href="#cb60-630" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-631"><a href="#cb60-631" aria-hidden="true" tabindex="-1"></a>trac_selected <span class="ot">&lt;-</span> fit_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb60-632"><a href="#cb60-632" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb60-633"><a href="#cb60-633" aria-hidden="true" tabindex="-1"></a>  names <span class="sc">%&gt;%</span></span>
<span id="cb60-634"><a href="#cb60-634" aria-hidden="true" tabindex="-1"></a>  enframe <span class="sc">%&gt;%</span></span>
<span id="cb60-635"><a href="#cb60-635" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">selected_trac =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb60-636"><a href="#cb60-636" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>name) <span class="sc">%&gt;%</span></span>
<span id="cb60-637"><a href="#cb60-637" aria-hidden="true" tabindex="-1"></a>  <span class="fu">right_join</span>(tree_df)</span>
<span id="cb60-638"><a href="#cb60-638" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-639"><a href="#cb60-639" aria-hidden="true" tabindex="-1"></a>tree_trac <span class="ot">&lt;-</span> tree <span class="sc">%&lt;+%</span> trac_selected</span>
<span id="cb60-640"><a href="#cb60-640" aria-hidden="true" tabindex="-1"></a>tree_trac <span class="sc">+</span></span>
<span id="cb60-641"><a href="#cb60-641" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point2</span>(<span class="fu">aes</span>(<span class="at">subset =</span> selected_trac <span class="sc">==</span> <span class="cn">TRUE</span>), <span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>)</span>
<span id="cb60-642"><a href="#cb60-642" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-643"><a href="#cb60-643" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-644"><a href="#cb60-644" aria-hidden="true" tabindex="-1"></a><span class="fu">### Weighted trac</span></span>
<span id="cb60-645"><a href="#cb60-645" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-646"><a href="#cb60-646" aria-hidden="true" tabindex="-1"></a>Therefore, using weighted trac could be better here.</span>
<span id="cb60-647"><a href="#cb60-647" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-650"><a href="#cb60-650" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-651"><a href="#cb60-651" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: weighted-trac-coefs</span></span>
<span id="cb60-652"><a href="#cb60-652" aria-hidden="true" tabindex="-1"></a>fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb60-653"><a href="#cb60-653" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb60-654"><a href="#cb60-654" aria-hidden="true" tabindex="-1"></a>  <span class="fu">enframe</span>() <span class="sc">%&gt;%</span></span>
<span id="cb60-655"><a href="#cb60-655" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(value)) <span class="sc">%&gt;%</span></span>
<span id="cb60-656"><a href="#cb60-656" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kbl</span>()</span>
<span id="cb60-657"><a href="#cb60-657" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-658"><a href="#cb60-658" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-661"><a href="#cb60-661" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-662"><a href="#cb60-662" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: plot-weighted-trac</span></span>
<span id="cb60-663"><a href="#cb60-663" aria-hidden="true" tabindex="-1"></a>fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb60-664"><a href="#cb60-664" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb60-665"><a href="#cb60-665" aria-hidden="true" tabindex="-1"></a>  <span class="fu">enframe</span>() <span class="sc">%&gt;%</span></span>
<span id="cb60-666"><a href="#cb60-666" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(value)) <span class="sc">%&gt;%</span></span>
<span id="cb60-667"><a href="#cb60-667" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kbl</span>()</span>
<span id="cb60-668"><a href="#cb60-668" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-669"><a href="#cb60-669" aria-hidden="true" tabindex="-1"></a>weighted_trac_selected <span class="ot">&lt;-</span> </span>
<span id="cb60-670"><a href="#cb60-670" aria-hidden="true" tabindex="-1"></a>  fit_weighted_trac[[<span class="dv">1</span>]]<span class="sc">$</span>alpha[, cvfit_weighted_trac<span class="sc">$</span>cv[[<span class="dv">1</span>]]<span class="sc">$</span>i1se] <span class="sc">%&gt;%</span></span>
<span id="cb60-671"><a href="#cb60-671" aria-hidden="true" tabindex="-1"></a>  nzz <span class="sc">%&gt;%</span></span>
<span id="cb60-672"><a href="#cb60-672" aria-hidden="true" tabindex="-1"></a>  names <span class="sc">%&gt;%</span></span>
<span id="cb60-673"><a href="#cb60-673" aria-hidden="true" tabindex="-1"></a>  enframe <span class="sc">%&gt;%</span></span>
<span id="cb60-674"><a href="#cb60-674" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">selected_weighted_trac =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb60-675"><a href="#cb60-675" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>name) <span class="sc">%&gt;%</span></span>
<span id="cb60-676"><a href="#cb60-676" aria-hidden="true" tabindex="-1"></a>  <span class="fu">right_join</span>(tree_df)</span>
<span id="cb60-677"><a href="#cb60-677" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-678"><a href="#cb60-678" aria-hidden="true" tabindex="-1"></a>tree_weighted_trac <span class="ot">&lt;-</span> tree <span class="sc">%&lt;+%</span> weighted_trac_selected</span>
<span id="cb60-679"><a href="#cb60-679" aria-hidden="true" tabindex="-1"></a>tree_weighted_trac <span class="sc">+</span></span>
<span id="cb60-680"><a href="#cb60-680" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point2</span>(<span class="fu">aes</span>(<span class="at">subset =</span> selected_weighted_trac <span class="sc">==</span> <span class="cn">TRUE</span>), </span>
<span id="cb60-681"><a href="#cb60-681" aria-hidden="true" tabindex="-1"></a>              <span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>)</span>
<span id="cb60-682"><a href="#cb60-682" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb60-683"><a href="#cb60-683" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-684"><a href="#cb60-684" aria-hidden="true" tabindex="-1"></a><span class="fu">## Computational environment</span></span>
<span id="cb60-685"><a href="#cb60-685" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-688"><a href="#cb60-688" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb60-689"><a href="#cb60-689" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: comp-envir</span></span>
<span id="cb60-690"><a href="#cb60-690" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb60-691"><a href="#cb60-691" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>